{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline to analyze data from methylome\n",
    "\n",
    "## Data from *Arabidopsis*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Condensed code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Importing libraries\n",
    "import os\n",
    "import pandas \n",
    "import numpy \n",
    "%matplotlib notebook\n",
    "import matplotlib.pyplot \n",
    "matplotlib.style.use('ggplot')\n",
    "import sys\n",
    "sys.path.insert(0, '/Users/luciehuyghe/Desktop/Pipeline/scripts') #needed to be able to import a script in another folder.\n",
    "#Here my methyl script with the graph functions is not in the same folder as my python general script.\n",
    "import methyl as m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Defining parameters\n",
    "nbchr = 5 #numbre of chromosoms\n",
    "numchr = 'all' #chromosome(s) we want to study : either a number [x], a serie of numbers [x,y] or 'all'\n",
    "nrows = 2000000#number of rows to read or 'all' if you wish to read the entire file\n",
    "nbecotype= 60 #number of ecotypes\n",
    "window= 1000 #size of the window in basepair\n",
    "max_size=40000000 #size of the longest chromosome\n",
    "min_size =0 # 0 if we want all the chromosome, other if you want a specific portion : position start\n",
    "list_ecotypes=os.path.join('ecotypes_list.txt') #list of the ecotypes that will be used. Put the ref FIRST\n",
    "path_to_file = '/Users/luciehuyghe/Desktop/Pipeline/Eckers_data/Methylome'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Only the first 60 ecotypes will be read\n",
      "Only the first 2000000 rows will be read\n",
      "I read ecotype 1/60 : Col_0\n",
      "I read ecotype 2/60 : Fr_2\n",
      "I read ecotype 3/60 : Kro_0\n",
      "I read ecotype 4/60 : Bsch_0\n",
      "I read ecotype 5/60 : En_D\n",
      "I read ecotype 6/60 : Kelsterbach_4\n",
      "I read ecotype 7/60 : Aa_0\n",
      "I read ecotype 8/60 : Abd_0\n",
      "I read ecotype 9/60 : Ag_0\n",
      "I read ecotype 10/60 : Ak_1\n",
      "I read ecotype 11/60 : Altai_5\n",
      "I read ecotype 12/60 : Amel_1\n",
      "I read ecotype 13/60 : An_1\n",
      "I read ecotype 14/60 : Anholt_1\n",
      "I read ecotype 15/60 : Ann_1\n",
      "I read ecotype 16/60 : Anz_0\n",
      "I read ecotype 17/60 : Appt_1\n",
      "I read ecotype 18/60 : Ba_1\n",
      "I read ecotype 19/60 : Baa_1\n",
      "I read ecotype 20/60 : Benk_1\n",
      "I read ecotype 21/60 : Bik_1\n",
      "I read ecotype 22/60 : Bl_1\n",
      "I read ecotype 23/60 : Bla_1\n",
      "I read ecotype 24/60 : Boot_1\n",
      "I read ecotype 25/60 : Bor_4\n",
      "I read ecotype 26/60 : Bs_1\n",
      "I read ecotype 27/60 : Bu_0\n",
      "I read ecotype 28/60 : Ca_0\n",
      "I read ecotype 29/60 : Cal_0\n",
      "I read ecotype 30/60 : Chat_1\n",
      "I read ecotype 31/60 : Chi_0\n",
      "I read ecotype 32/60 : Cnt_1\n",
      "I read ecotype 33/60 : Com_1\n",
      "I read ecotype 34/60 : Cvi_0\n",
      "I read ecotype 35/60 : Da_1_12\n",
      "I read ecotype 36/60 : Db_1\n",
      "I read ecotype 37/60 : Di_G\n",
      "I read ecotype 38/60 : Dja_1\n",
      "I read ecotype 39/60 : Ei_2\n",
      "I read ecotype 40/60 : El_0\n",
      "I read ecotype 41/60 : Ema_1\n",
      "I read ecotype 42/60 : Es_0\n",
      "I read ecotype 43/60 : Est\n",
      "I read ecotype 44/60 : Et_0\n",
      "I read ecotype 45/60 : Fi_0\n",
      "I read ecotype 46/60 : Ga_0\n",
      "I read ecotype 47/60 : Gel_1\n",
      "I read ecotype 48/60 : Gie_0\n",
      "I read ecotype 49/60 : Gifu_2\n",
      "I read ecotype 50/60 : Got_7\n",
      "I read ecotype 51/60 : Gr_1\n",
      "I read ecotype 52/60 : Gre_0\n",
      "I read ecotype 53/60 : Gu_0\n",
      "I read ecotype 54/60 : Gy_0\n",
      "I read ecotype 55/60 : Ha_0\n",
      "I read ecotype 56/60 : Hey_1\n",
      "I read ecotype 57/60 : Hs_0\n",
      "I read ecotype 58/60 : In_0\n",
      "I read ecotype 59/60 : Is_0\n",
      "I read ecotype 60/60 : Je_0\n"
     ]
    }
   ],
   "source": [
    "imp=open(list_ecotypes, 'r')\n",
    "list_ecotype=[]\n",
    "names=[]\n",
    "for line in imp:\n",
    "    for i in line.split():\n",
    "        list_ecotype.append(i)\n",
    "        names.append(i[20:-4])\n",
    "imp.close()\n",
    "for i in range (1, nbchr+1):\n",
    "    exec('methylome_file%d = []' % (i))\n",
    "    exec('onlymethyl%d = []' % (i))\n",
    "    exec('CGs%d = []' % (i))\n",
    "    exec('CHGs%d = []' % (i))\n",
    "    exec('CHHs%d = []' % (i))\n",
    "    \n",
    "if nbecotype == 'all':\n",
    "    print('No number of ecotype specified, reading all the ecotypes')\n",
    "    if nrows == 'all' :\n",
    "        print('No number of rows specified, reading all the file')\n",
    "        for i in range (len(list_ecotype)) :    \n",
    "            methylome_files=(pandas.read_csv(os.path.join(path_to_files, list_ecotype[i]), sep='\\t'))\n",
    "            if numchr == 'all' :\n",
    "                for x in range (1,nbchr+1):\n",
    "                    exec('onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (x,x))\n",
    "                    exec('methylome_file%d.append(methylome_files[methylome_files.chrom == %d])'% (x,x))\n",
    "            else :\n",
    "                for c in numchr :\n",
    "                    exec('onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (c,c))\n",
    "                    exec('methylome_file%d.append(methylome_files[methylome_files.chrom == %d])'% (c,c))\n",
    "            methylome_files=None\n",
    "            print('I read ecotype %d/%d : %s' %(i+1,len(list_ecotype), names[i]))\n",
    "    else :\n",
    "        print('Only the first %d rows will be read' % (nrows))\n",
    "        for i in range (len(list_ecotype)) :    \n",
    "            methylome_files=(pandas.read_csv(os.path.join(path_to_file,list_ecotype[i]), sep='\\t', nrows=nrows))\n",
    "            if numchr == 'all' :\n",
    "                for x in range (1,nbchr+1):\n",
    "                    exec('onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (x,x))\n",
    "                    exec('methylome_file%d.append(methylome_files[methylome_files.chrom == %d])'% (x,x))\n",
    "            else :\n",
    "                for c in numchr :\n",
    "                    exec('onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (c,c))\n",
    "                    exec('methylome_file%d.append(methylome_files[methylome_files.chrom == %d])'% (c,c))\n",
    "            methylome_files=None\n",
    "            print('I read ecotype %d/%d : %s' %(i+1,len(list_ecotype), names[i]))\n",
    "else :\n",
    "    print('Only the first %d ecotypes will be read' % (nbecotype))\n",
    "    if nrows == 'all':\n",
    "        print('No number of rows specified, reading all the rows')\n",
    "        for i in range (nbecotype) :\n",
    "            methylome_files=(pandas.read_csv(os.path.join(path_to_file,list_ecotype[i]), sep='\\t'))\n",
    "            if numchr == 'all' :\n",
    "                for x in range (1,nbchr+1):\n",
    "                    exec('onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (x,x))\n",
    "                    exec('methylome_file%d.append(methylome_files[methylome_files.chrom == %d])'% (x,x))\n",
    "            else :\n",
    "                for c in numchr :\n",
    "                    exec('onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (c,c))\n",
    "                    exec('methylome_file%d.append(methylome_files[methylome_files.chrom == %d])'% (c,c))\n",
    "            methylome_files=None    \n",
    "            print('I read ecotype %d/%d : %s' %(i+1,nbecotype, names[i]))\n",
    "    else:\n",
    "        print('Only the first %d rows will be read' % (nrows))\n",
    "        for i in range (nbecotype) :\n",
    "            methylome_files=(pandas.read_csv(os.path.join(path_to_file,list_ecotype[i]), sep='\\t', nrows=nrows))\n",
    "            if numchr == 'all' :\n",
    "                for x in range (1,nbchr+1):\n",
    "                    exec('onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (x,x))\n",
    "                    exec('methylome_file%d.append(methylome_files[methylome_files.chrom == %d])'% (x,x))\n",
    "            else :\n",
    "                for c in numchr :\n",
    "                    exec('onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (c,c))\n",
    "                    exec('methylome_file%d.append(methylome_files[methylome_files.chrom == %d])'% (c,c))\n",
    "            methylome_files=None    \n",
    "            print('I read ecotype %d/%d : %s' %(i+1,nbecotype, names[i]))\n",
    "\n",
    "for a in range (1,nbchr+1) :\n",
    "    for i in range (nbecotype) :\n",
    "        exec('CHGs%d.append(onlymethyl%d[i][(onlymethyl%d[i].mc_class==\"CAG\") | (onlymethyl%d[i].mc_class==\"CCG\") |(onlymethyl%d[i].mc_class==\"CTG\")])' % (a,a,a,a,a))\n",
    "        exec('CGs%d.append(onlymethyl%d[i][(onlymethyl%d[i].mc_class==\"CGA\")|(onlymethyl%d[i].mc_class==\"CGT\")|(onlymethyl%d[i].mc_class==\"CGC\")|(onlymethyl%d[i].mc_class==\"CGG\")])' % (a,a,a,a,a,a))\n",
    "        exec('CHHs%d.append(onlymethyl%d[i][(onlymethyl%d[i].mc_class==\"CAA\")|(onlymethyl%d[i].mc_class==\"CAT\")|(onlymethyl%d[i].mc_class==\"CAC\")|(onlymethyl%d[i].mc_class==\"CCA\")|(onlymethyl%d[i].mc_class==\"CCC\")|(onlymethyl%d[i].mc_class==\"CCT\")|(onlymethyl%d[i].mc_class==\"CTA\")|(onlymethyl%d[i].mc_class==\"CTC\")|(onlymethyl%d[i].mc_class==\"CTT\")])'% (a,a,a,a,a,a,a,a,a,a,a))\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>mc_class</th>\n",
       "      <th>methylated_bases</th>\n",
       "      <th>total_bases</th>\n",
       "      <th>methylation_call</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>-</td>\n",
       "      <td>CAG</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>75</td>\n",
       "      <td>+</td>\n",
       "      <td>CCA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>76</td>\n",
       "      <td>+</td>\n",
       "      <td>CAT</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>79</td>\n",
       "      <td>-</td>\n",
       "      <td>CAT</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>83</td>\n",
       "      <td>+</td>\n",
       "      <td>CCC</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>84</td>\n",
       "      <td>+</td>\n",
       "      <td>CCT</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>85</td>\n",
       "      <td>+</td>\n",
       "      <td>CTA</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>92</td>\n",
       "      <td>+</td>\n",
       "      <td>CCT</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>+</td>\n",
       "      <td>CTA</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>99</td>\n",
       "      <td>+</td>\n",
       "      <td>CCC</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>CCT</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>101</td>\n",
       "      <td>+</td>\n",
       "      <td>CTA</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>106</td>\n",
       "      <td>+</td>\n",
       "      <td>CCC</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>107</td>\n",
       "      <td>+</td>\n",
       "      <td>CCG</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>108</td>\n",
       "      <td>+</td>\n",
       "      <td>CGA</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>-</td>\n",
       "      <td>CGG</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>+</td>\n",
       "      <td>CCG</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>114</td>\n",
       "      <td>+</td>\n",
       "      <td>CGG</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>115</td>\n",
       "      <td>-</td>\n",
       "      <td>CGG</td>\n",
       "      <td>28</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>116</td>\n",
       "      <td>-</td>\n",
       "      <td>CCG</td>\n",
       "      <td>7</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>+</td>\n",
       "      <td>CTC</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>+</td>\n",
       "      <td>CTG</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>-</td>\n",
       "      <td>CAG</td>\n",
       "      <td>23</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>-</td>\n",
       "      <td>CCA</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>128</td>\n",
       "      <td>-</td>\n",
       "      <td>CAA</td>\n",
       "      <td>7</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>134</td>\n",
       "      <td>+</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>138</td>\n",
       "      <td>-</td>\n",
       "      <td>CAA</td>\n",
       "      <td>16</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>-</td>\n",
       "      <td>CAC</td>\n",
       "      <td>17</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>149</td>\n",
       "      <td>-</td>\n",
       "      <td>CAT</td>\n",
       "      <td>17</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>CGT</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999970</th>\n",
       "      <td>1</td>\n",
       "      <td>7814512</td>\n",
       "      <td>+</td>\n",
       "      <td>CTT</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999971</th>\n",
       "      <td>1</td>\n",
       "      <td>7814515</td>\n",
       "      <td>+</td>\n",
       "      <td>CTT</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999972</th>\n",
       "      <td>1</td>\n",
       "      <td>7814522</td>\n",
       "      <td>+</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999973</th>\n",
       "      <td>1</td>\n",
       "      <td>7814525</td>\n",
       "      <td>+</td>\n",
       "      <td>CAA</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999974</th>\n",
       "      <td>1</td>\n",
       "      <td>7814531</td>\n",
       "      <td>+</td>\n",
       "      <td>CTT</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999975</th>\n",
       "      <td>1</td>\n",
       "      <td>7814535</td>\n",
       "      <td>-</td>\n",
       "      <td>CAA</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999976</th>\n",
       "      <td>1</td>\n",
       "      <td>7814536</td>\n",
       "      <td>-</td>\n",
       "      <td>CCA</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999977</th>\n",
       "      <td>1</td>\n",
       "      <td>7814542</td>\n",
       "      <td>+</td>\n",
       "      <td>CGG</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999978</th>\n",
       "      <td>1</td>\n",
       "      <td>7814543</td>\n",
       "      <td>-</td>\n",
       "      <td>CGT</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999979</th>\n",
       "      <td>1</td>\n",
       "      <td>7814544</td>\n",
       "      <td>-</td>\n",
       "      <td>CCG</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999980</th>\n",
       "      <td>1</td>\n",
       "      <td>7814546</td>\n",
       "      <td>-</td>\n",
       "      <td>CAC</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999981</th>\n",
       "      <td>1</td>\n",
       "      <td>7814549</td>\n",
       "      <td>-</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999982</th>\n",
       "      <td>1</td>\n",
       "      <td>7814551</td>\n",
       "      <td>-</td>\n",
       "      <td>CTC</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999983</th>\n",
       "      <td>1</td>\n",
       "      <td>7814556</td>\n",
       "      <td>+</td>\n",
       "      <td>CGA</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999984</th>\n",
       "      <td>1</td>\n",
       "      <td>7814557</td>\n",
       "      <td>-</td>\n",
       "      <td>CGT</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999985</th>\n",
       "      <td>1</td>\n",
       "      <td>7814559</td>\n",
       "      <td>-</td>\n",
       "      <td>CTC</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999986</th>\n",
       "      <td>1</td>\n",
       "      <td>7814563</td>\n",
       "      <td>+</td>\n",
       "      <td>CTC</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999987</th>\n",
       "      <td>1</td>\n",
       "      <td>7814565</td>\n",
       "      <td>+</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999988</th>\n",
       "      <td>1</td>\n",
       "      <td>7814570</td>\n",
       "      <td>-</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999989</th>\n",
       "      <td>1</td>\n",
       "      <td>7814571</td>\n",
       "      <td>+</td>\n",
       "      <td>CTG</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999990</th>\n",
       "      <td>1</td>\n",
       "      <td>7814573</td>\n",
       "      <td>-</td>\n",
       "      <td>CAG</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999991</th>\n",
       "      <td>1</td>\n",
       "      <td>7814579</td>\n",
       "      <td>-</td>\n",
       "      <td>CTT</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999992</th>\n",
       "      <td>1</td>\n",
       "      <td>7814580</td>\n",
       "      <td>+</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999993</th>\n",
       "      <td>1</td>\n",
       "      <td>7814583</td>\n",
       "      <td>+</td>\n",
       "      <td>CAA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999994</th>\n",
       "      <td>1</td>\n",
       "      <td>7814587</td>\n",
       "      <td>-</td>\n",
       "      <td>CTT</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999995</th>\n",
       "      <td>1</td>\n",
       "      <td>7814589</td>\n",
       "      <td>+</td>\n",
       "      <td>CGC</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999996</th>\n",
       "      <td>1</td>\n",
       "      <td>7814590</td>\n",
       "      <td>-</td>\n",
       "      <td>CGT</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999997</th>\n",
       "      <td>1</td>\n",
       "      <td>7814591</td>\n",
       "      <td>+</td>\n",
       "      <td>CTA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999998</th>\n",
       "      <td>1</td>\n",
       "      <td>7814595</td>\n",
       "      <td>+</td>\n",
       "      <td>CTT</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999999</th>\n",
       "      <td>1</td>\n",
       "      <td>7814602</td>\n",
       "      <td>-</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000000 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chrom      pos strand mc_class  methylated_bases  total_bases  \\\n",
       "0            1       33      -      CAG                 3            9   \n",
       "1            1       75      +      CCA                 0            1   \n",
       "2            1       76      +      CAT                 1            1   \n",
       "3            1       79      -      CAT                 2           29   \n",
       "4            1       83      +      CCC                 0            3   \n",
       "5            1       84      +      CCT                 2            4   \n",
       "6            1       85      +      CTA                 2            4   \n",
       "7            1       92      +      CCT                 1            4   \n",
       "8            1       93      +      CTA                 2            4   \n",
       "9            1       99      +      CCC                 2            5   \n",
       "10           1      100      +      CCT                 2            5   \n",
       "11           1      101      +      CTA                 2            5   \n",
       "12           1      106      +      CCC                 0            5   \n",
       "13           1      107      +      CCG                 4            5   \n",
       "14           1      108      +      CGA                 5            5   \n",
       "15           1      109      -      CGG                29           29   \n",
       "16           1      113      +      CCG                 3            6   \n",
       "17           1      114      +      CGG                 5            6   \n",
       "18           1      115      -      CGG                28           29   \n",
       "19           1      116      -      CCG                 7           29   \n",
       "20           1      120      +      CTC                 0            7   \n",
       "21           1      122      +      CTG                 7            7   \n",
       "22           1      124      -      CAG                23           29   \n",
       "23           1      125      -      CCA                10           28   \n",
       "24           1      128      -      CAA                 7           29   \n",
       "25           1      134      +      CAT                 0            7   \n",
       "26           1      138      -      CAA                16           30   \n",
       "27           1      140      -      CAC                17           29   \n",
       "28           1      149      -      CAT                17           25   \n",
       "29           1      160      +      CGT                 9           11   \n",
       "...        ...      ...    ...      ...               ...          ...   \n",
       "1999970      1  7814512      +      CTT                 0            3   \n",
       "1999971      1  7814515      +      CTT                 0            3   \n",
       "1999972      1  7814522      +      CAT                 0            3   \n",
       "1999973      1  7814525      +      CAA                 0            3   \n",
       "1999974      1  7814531      +      CTT                 0            3   \n",
       "1999975      1  7814535      -      CAA                 0            2   \n",
       "1999976      1  7814536      -      CCA                 0            2   \n",
       "1999977      1  7814542      +      CGG                 0            3   \n",
       "1999978      1  7814543      -      CGT                 0            2   \n",
       "1999979      1  7814544      -      CCG                 0            2   \n",
       "1999980      1  7814546      -      CAC                 0            2   \n",
       "1999981      1  7814549      -      CAT                 0            2   \n",
       "1999982      1  7814551      -      CTC                 0            2   \n",
       "1999983      1  7814556      +      CGA                 0            2   \n",
       "1999984      1  7814557      -      CGT                 0            2   \n",
       "1999985      1  7814559      -      CTC                 0            2   \n",
       "1999986      1  7814563      +      CTC                 0            2   \n",
       "1999987      1  7814565      +      CAT                 0            2   \n",
       "1999988      1  7814570      -      CAT                 0            1   \n",
       "1999989      1  7814571      +      CTG                 0            2   \n",
       "1999990      1  7814573      -      CAG                 0            1   \n",
       "1999991      1  7814579      -      CTT                 0            1   \n",
       "1999992      1  7814580      +      CAT                 0            2   \n",
       "1999993      1  7814583      +      CAA                 0            1   \n",
       "1999994      1  7814587      -      CTT                 0            1   \n",
       "1999995      1  7814589      +      CGC                 0            1   \n",
       "1999996      1  7814590      -      CGT                 0            1   \n",
       "1999997      1  7814591      +      CTA                 0            1   \n",
       "1999998      1  7814595      +      CTT                 0            1   \n",
       "1999999      1  7814602      -      CAT                 0            2   \n",
       "\n",
       "         methylation_call  \n",
       "0                       1  \n",
       "1                       0  \n",
       "2                       0  \n",
       "3                       0  \n",
       "4                       0  \n",
       "5                       1  \n",
       "6                       1  \n",
       "7                       0  \n",
       "8                       1  \n",
       "9                       0  \n",
       "10                      0  \n",
       "11                      0  \n",
       "12                      0  \n",
       "13                      1  \n",
       "14                      1  \n",
       "15                      1  \n",
       "16                      1  \n",
       "17                      1  \n",
       "18                      1  \n",
       "19                      1  \n",
       "20                      0  \n",
       "21                      1  \n",
       "22                      1  \n",
       "23                      1  \n",
       "24                      1  \n",
       "25                      0  \n",
       "26                      1  \n",
       "27                      1  \n",
       "28                      1  \n",
       "29                      1  \n",
       "...                   ...  \n",
       "1999970                 0  \n",
       "1999971                 0  \n",
       "1999972                 0  \n",
       "1999973                 0  \n",
       "1999974                 0  \n",
       "1999975                 0  \n",
       "1999976                 0  \n",
       "1999977                 0  \n",
       "1999978                 0  \n",
       "1999979                 0  \n",
       "1999980                 0  \n",
       "1999981                 0  \n",
       "1999982                 0  \n",
       "1999983                 0  \n",
       "1999984                 0  \n",
       "1999985                 0  \n",
       "1999986                 0  \n",
       "1999987                 0  \n",
       "1999988                 0  \n",
       "1999989                 0  \n",
       "1999990                 0  \n",
       "1999991                 0  \n",
       "1999992                 0  \n",
       "1999993                 0  \n",
       "1999994                 0  \n",
       "1999995                 0  \n",
       "1999996                 0  \n",
       "1999997                 0  \n",
       "1999998                 0  \n",
       "1999999                 0  \n",
       "\n",
       "[2000000 rows x 7 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "methylome_file1[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "/* Put everything inside the global mpl namespace */\n",
       "window.mpl = {};\n",
       "\n",
       "mpl.get_websocket_type = function() {\n",
       "    if (typeof(WebSocket) !== 'undefined') {\n",
       "        return WebSocket;\n",
       "    } else if (typeof(MozWebSocket) !== 'undefined') {\n",
       "        return MozWebSocket;\n",
       "    } else {\n",
       "        alert('Your browser does not have WebSocket support.' +\n",
       "              'Please try Chrome, Safari or Firefox ≥ 6. ' +\n",
       "              'Firefox 4 and 5 are also supported but you ' +\n",
       "              'have to enable WebSockets in about:config.');\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure = function(figure_id, websocket, ondownload, parent_element) {\n",
       "    this.id = figure_id;\n",
       "\n",
       "    this.ws = websocket;\n",
       "\n",
       "    this.supports_binary = (this.ws.binaryType != undefined);\n",
       "\n",
       "    if (!this.supports_binary) {\n",
       "        var warnings = document.getElementById(\"mpl-warnings\");\n",
       "        if (warnings) {\n",
       "            warnings.style.display = 'block';\n",
       "            warnings.textContent = (\n",
       "                \"This browser does not support binary websocket messages. \" +\n",
       "                    \"Performance may be slow.\");\n",
       "        }\n",
       "    }\n",
       "\n",
       "    this.imageObj = new Image();\n",
       "\n",
       "    this.context = undefined;\n",
       "    this.message = undefined;\n",
       "    this.canvas = undefined;\n",
       "    this.rubberband_canvas = undefined;\n",
       "    this.rubberband_context = undefined;\n",
       "    this.format_dropdown = undefined;\n",
       "\n",
       "    this.image_mode = 'full';\n",
       "\n",
       "    this.root = $('<div/>');\n",
       "    this._root_extra_style(this.root)\n",
       "    this.root.attr('style', 'display: inline-block');\n",
       "\n",
       "    $(parent_element).append(this.root);\n",
       "\n",
       "    this._init_header(this);\n",
       "    this._init_canvas(this);\n",
       "    this._init_toolbar(this);\n",
       "\n",
       "    var fig = this;\n",
       "\n",
       "    this.waiting = false;\n",
       "\n",
       "    this.ws.onopen =  function () {\n",
       "            fig.send_message(\"supports_binary\", {value: fig.supports_binary});\n",
       "            fig.send_message(\"send_image_mode\", {});\n",
       "            fig.send_message(\"refresh\", {});\n",
       "        }\n",
       "\n",
       "    this.imageObj.onload = function() {\n",
       "            if (fig.image_mode == 'full') {\n",
       "                // Full images could contain transparency (where diff images\n",
       "                // almost always do), so we need to clear the canvas so that\n",
       "                // there is no ghosting.\n",
       "                fig.context.clearRect(0, 0, fig.canvas.width, fig.canvas.height);\n",
       "            }\n",
       "            fig.context.drawImage(fig.imageObj, 0, 0);\n",
       "        };\n",
       "\n",
       "    this.imageObj.onunload = function() {\n",
       "        this.ws.close();\n",
       "    }\n",
       "\n",
       "    this.ws.onmessage = this._make_on_message_function(this);\n",
       "\n",
       "    this.ondownload = ondownload;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_header = function() {\n",
       "    var titlebar = $(\n",
       "        '<div class=\"ui-dialog-titlebar ui-widget-header ui-corner-all ' +\n",
       "        'ui-helper-clearfix\"/>');\n",
       "    var titletext = $(\n",
       "        '<div class=\"ui-dialog-title\" style=\"width: 100%; ' +\n",
       "        'text-align: center; padding: 3px;\"/>');\n",
       "    titlebar.append(titletext)\n",
       "    this.root.append(titlebar);\n",
       "    this.header = titletext[0];\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_canvas = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var canvas_div = $('<div/>');\n",
       "\n",
       "    canvas_div.attr('style', 'position: relative; clear: both; outline: 0');\n",
       "\n",
       "    function canvas_keyboard_event(event) {\n",
       "        return fig.key_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    canvas_div.keydown('key_press', canvas_keyboard_event);\n",
       "    canvas_div.keyup('key_release', canvas_keyboard_event);\n",
       "    this.canvas_div = canvas_div\n",
       "    this._canvas_extra_style(canvas_div)\n",
       "    this.root.append(canvas_div);\n",
       "\n",
       "    var canvas = $('<canvas/>');\n",
       "    canvas.addClass('mpl-canvas');\n",
       "    canvas.attr('style', \"left: 0; top: 0; z-index: 0; outline: 0\")\n",
       "\n",
       "    this.canvas = canvas[0];\n",
       "    this.context = canvas[0].getContext(\"2d\");\n",
       "\n",
       "    var rubberband = $('<canvas/>');\n",
       "    rubberband.attr('style', \"position: absolute; left: 0; top: 0; z-index: 1;\")\n",
       "\n",
       "    var pass_mouse_events = true;\n",
       "\n",
       "    canvas_div.resizable({\n",
       "        start: function(event, ui) {\n",
       "            pass_mouse_events = false;\n",
       "        },\n",
       "        resize: function(event, ui) {\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "        stop: function(event, ui) {\n",
       "            pass_mouse_events = true;\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "    });\n",
       "\n",
       "    function mouse_event_fn(event) {\n",
       "        if (pass_mouse_events)\n",
       "            return fig.mouse_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    rubberband.mousedown('button_press', mouse_event_fn);\n",
       "    rubberband.mouseup('button_release', mouse_event_fn);\n",
       "    // Throttle sequential mouse events to 1 every 20ms.\n",
       "    rubberband.mousemove('motion_notify', mouse_event_fn);\n",
       "\n",
       "    rubberband.mouseenter('figure_enter', mouse_event_fn);\n",
       "    rubberband.mouseleave('figure_leave', mouse_event_fn);\n",
       "\n",
       "    canvas_div.on(\"wheel\", function (event) {\n",
       "        event = event.originalEvent;\n",
       "        event['data'] = 'scroll'\n",
       "        if (event.deltaY < 0) {\n",
       "            event.step = 1;\n",
       "        } else {\n",
       "            event.step = -1;\n",
       "        }\n",
       "        mouse_event_fn(event);\n",
       "    });\n",
       "\n",
       "    canvas_div.append(canvas);\n",
       "    canvas_div.append(rubberband);\n",
       "\n",
       "    this.rubberband = rubberband;\n",
       "    this.rubberband_canvas = rubberband[0];\n",
       "    this.rubberband_context = rubberband[0].getContext(\"2d\");\n",
       "    this.rubberband_context.strokeStyle = \"#000000\";\n",
       "\n",
       "    this._resize_canvas = function(width, height) {\n",
       "        // Keep the size of the canvas, canvas container, and rubber band\n",
       "        // canvas in synch.\n",
       "        canvas_div.css('width', width)\n",
       "        canvas_div.css('height', height)\n",
       "\n",
       "        canvas.attr('width', width);\n",
       "        canvas.attr('height', height);\n",
       "\n",
       "        rubberband.attr('width', width);\n",
       "        rubberband.attr('height', height);\n",
       "    }\n",
       "\n",
       "    // Set the figure to an initial 600x600px, this will subsequently be updated\n",
       "    // upon first draw.\n",
       "    this._resize_canvas(600, 600);\n",
       "\n",
       "    // Disable right mouse context menu.\n",
       "    $(this.rubberband_canvas).bind(\"contextmenu\",function(e){\n",
       "        return false;\n",
       "    });\n",
       "\n",
       "    function set_focus () {\n",
       "        canvas.focus();\n",
       "        canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    window.setTimeout(set_focus, 100);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>')\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items) {\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) {\n",
       "            // put a spacer in here.\n",
       "            continue;\n",
       "        }\n",
       "        var button = $('<button/>');\n",
       "        button.addClass('ui-button ui-widget ui-state-default ui-corner-all ' +\n",
       "                        'ui-button-icon-only');\n",
       "        button.attr('role', 'button');\n",
       "        button.attr('aria-disabled', 'false');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "\n",
       "        var icon_img = $('<span/>');\n",
       "        icon_img.addClass('ui-button-icon-primary ui-icon');\n",
       "        icon_img.addClass(image);\n",
       "        icon_img.addClass('ui-corner-all');\n",
       "\n",
       "        var tooltip_span = $('<span/>');\n",
       "        tooltip_span.addClass('ui-button-text');\n",
       "        tooltip_span.html(tooltip);\n",
       "\n",
       "        button.append(icon_img);\n",
       "        button.append(tooltip_span);\n",
       "\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    var fmt_picker_span = $('<span/>');\n",
       "\n",
       "    var fmt_picker = $('<select/>');\n",
       "    fmt_picker.addClass('mpl-toolbar-option ui-widget ui-widget-content');\n",
       "    fmt_picker_span.append(fmt_picker);\n",
       "    nav_element.append(fmt_picker_span);\n",
       "    this.format_dropdown = fmt_picker[0];\n",
       "\n",
       "    for (var ind in mpl.extensions) {\n",
       "        var fmt = mpl.extensions[ind];\n",
       "        var option = $(\n",
       "            '<option/>', {selected: fmt === mpl.default_extension}).html(fmt);\n",
       "        fmt_picker.append(option)\n",
       "    }\n",
       "\n",
       "    // Add hover states to the ui-buttons\n",
       "    $( \".ui-button\" ).hover(\n",
       "        function() { $(this).addClass(\"ui-state-hover\");},\n",
       "        function() { $(this).removeClass(\"ui-state-hover\");}\n",
       "    );\n",
       "\n",
       "    var status_bar = $('<span class=\"mpl-message\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.request_resize = function(x_pixels, y_pixels) {\n",
       "    // Request matplotlib to resize the figure. Matplotlib will then trigger a resize in the client,\n",
       "    // which will in turn request a refresh of the image.\n",
       "    this.send_message('resize', {'width': x_pixels, 'height': y_pixels});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_message = function(type, properties) {\n",
       "    properties['type'] = type;\n",
       "    properties['figure_id'] = this.id;\n",
       "    this.ws.send(JSON.stringify(properties));\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_draw_message = function() {\n",
       "    if (!this.waiting) {\n",
       "        this.waiting = true;\n",
       "        this.ws.send(JSON.stringify({type: \"draw\", figure_id: this.id}));\n",
       "    }\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    var format_dropdown = fig.format_dropdown;\n",
       "    var format = format_dropdown.options[format_dropdown.selectedIndex].value;\n",
       "    fig.ondownload(fig, format);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_resize = function(fig, msg) {\n",
       "    var size = msg['size'];\n",
       "    if (size[0] != fig.canvas.width || size[1] != fig.canvas.height) {\n",
       "        fig._resize_canvas(size[0], size[1]);\n",
       "        fig.send_message(\"refresh\", {});\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_rubberband = function(fig, msg) {\n",
       "    var x0 = msg['x0'];\n",
       "    var y0 = fig.canvas.height - msg['y0'];\n",
       "    var x1 = msg['x1'];\n",
       "    var y1 = fig.canvas.height - msg['y1'];\n",
       "    x0 = Math.floor(x0) + 0.5;\n",
       "    y0 = Math.floor(y0) + 0.5;\n",
       "    x1 = Math.floor(x1) + 0.5;\n",
       "    y1 = Math.floor(y1) + 0.5;\n",
       "    var min_x = Math.min(x0, x1);\n",
       "    var min_y = Math.min(y0, y1);\n",
       "    var width = Math.abs(x1 - x0);\n",
       "    var height = Math.abs(y1 - y0);\n",
       "\n",
       "    fig.rubberband_context.clearRect(\n",
       "        0, 0, fig.canvas.width, fig.canvas.height);\n",
       "\n",
       "    fig.rubberband_context.strokeRect(min_x, min_y, width, height);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_figure_label = function(fig, msg) {\n",
       "    // Updates the figure title.\n",
       "    fig.header.textContent = msg['label'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_cursor = function(fig, msg) {\n",
       "    var cursor = msg['cursor'];\n",
       "    switch(cursor)\n",
       "    {\n",
       "    case 0:\n",
       "        cursor = 'pointer';\n",
       "        break;\n",
       "    case 1:\n",
       "        cursor = 'default';\n",
       "        break;\n",
       "    case 2:\n",
       "        cursor = 'crosshair';\n",
       "        break;\n",
       "    case 3:\n",
       "        cursor = 'move';\n",
       "        break;\n",
       "    }\n",
       "    fig.rubberband_canvas.style.cursor = cursor;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_message = function(fig, msg) {\n",
       "    fig.message.textContent = msg['message'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_draw = function(fig, msg) {\n",
       "    // Request the server to send over a new figure.\n",
       "    fig.send_draw_message();\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_image_mode = function(fig, msg) {\n",
       "    fig.image_mode = msg['mode'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Called whenever the canvas gets updated.\n",
       "    this.send_message(\"ack\", {});\n",
       "}\n",
       "\n",
       "// A function to construct a web socket function for onmessage handling.\n",
       "// Called in the figure constructor.\n",
       "mpl.figure.prototype._make_on_message_function = function(fig) {\n",
       "    return function socket_on_message(evt) {\n",
       "        if (evt.data instanceof Blob) {\n",
       "            /* FIXME: We get \"Resource interpreted as Image but\n",
       "             * transferred with MIME type text/plain:\" errors on\n",
       "             * Chrome.  But how to set the MIME type?  It doesn't seem\n",
       "             * to be part of the websocket stream */\n",
       "            evt.data.type = \"image/png\";\n",
       "\n",
       "            /* Free the memory for the previous frames */\n",
       "            if (fig.imageObj.src) {\n",
       "                (window.URL || window.webkitURL).revokeObjectURL(\n",
       "                    fig.imageObj.src);\n",
       "            }\n",
       "\n",
       "            fig.imageObj.src = (window.URL || window.webkitURL).createObjectURL(\n",
       "                evt.data);\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "        else if (typeof evt.data === 'string' && evt.data.slice(0, 21) == \"data:image/png;base64\") {\n",
       "            fig.imageObj.src = evt.data;\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        var msg = JSON.parse(evt.data);\n",
       "        var msg_type = msg['type'];\n",
       "\n",
       "        // Call the  \"handle_{type}\" callback, which takes\n",
       "        // the figure and JSON message as its only arguments.\n",
       "        try {\n",
       "            var callback = fig[\"handle_\" + msg_type];\n",
       "        } catch (e) {\n",
       "            console.log(\"No handler for the '\" + msg_type + \"' message type: \", msg);\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        if (callback) {\n",
       "            try {\n",
       "                // console.log(\"Handling '\" + msg_type + \"' message: \", msg);\n",
       "                callback(fig, msg);\n",
       "            } catch (e) {\n",
       "                console.log(\"Exception inside the 'handler_\" + msg_type + \"' callback:\", e, e.stack, msg);\n",
       "            }\n",
       "        }\n",
       "    };\n",
       "}\n",
       "\n",
       "// from http://stackoverflow.com/questions/1114465/getting-mouse-location-in-canvas\n",
       "mpl.findpos = function(e) {\n",
       "    //this section is from http://www.quirksmode.org/js/events_properties.html\n",
       "    var targ;\n",
       "    if (!e)\n",
       "        e = window.event;\n",
       "    if (e.target)\n",
       "        targ = e.target;\n",
       "    else if (e.srcElement)\n",
       "        targ = e.srcElement;\n",
       "    if (targ.nodeType == 3) // defeat Safari bug\n",
       "        targ = targ.parentNode;\n",
       "\n",
       "    // jQuery normalizes the pageX and pageY\n",
       "    // pageX,Y are the mouse positions relative to the document\n",
       "    // offset() returns the position of the element relative to the document\n",
       "    var x = e.pageX - $(targ).offset().left;\n",
       "    var y = e.pageY - $(targ).offset().top;\n",
       "\n",
       "    return {\"x\": x, \"y\": y};\n",
       "};\n",
       "\n",
       "/*\n",
       " * return a copy of an object with only non-object keys\n",
       " * we need this to avoid circular references\n",
       " * http://stackoverflow.com/a/24161582/3208463\n",
       " */\n",
       "function simpleKeys (original) {\n",
       "  return Object.keys(original).reduce(function (obj, key) {\n",
       "    if (typeof original[key] !== 'object')\n",
       "        obj[key] = original[key]\n",
       "    return obj;\n",
       "  }, {});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.mouse_event = function(event, name) {\n",
       "    var canvas_pos = mpl.findpos(event)\n",
       "\n",
       "    if (name === 'button_press')\n",
       "    {\n",
       "        this.canvas.focus();\n",
       "        this.canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    var x = canvas_pos.x;\n",
       "    var y = canvas_pos.y;\n",
       "\n",
       "    this.send_message(name, {x: x, y: y, button: event.button,\n",
       "                             step: event.step,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "\n",
       "    /* This prevents the web browser from automatically changing to\n",
       "     * the text insertion cursor when the button is pressed.  We want\n",
       "     * to control all of the cursor setting manually through the\n",
       "     * 'cursor' event from matplotlib */\n",
       "    event.preventDefault();\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    // Handle any extra behaviour associated with a key event\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.key_event = function(event, name) {\n",
       "\n",
       "    // Prevent repeat events\n",
       "    if (name == 'key_press')\n",
       "    {\n",
       "        if (event.which === this._key)\n",
       "            return;\n",
       "        else\n",
       "            this._key = event.which;\n",
       "    }\n",
       "    if (name == 'key_release')\n",
       "        this._key = null;\n",
       "\n",
       "    var value = '';\n",
       "    if (event.ctrlKey && event.which != 17)\n",
       "        value += \"ctrl+\";\n",
       "    if (event.altKey && event.which != 18)\n",
       "        value += \"alt+\";\n",
       "    if (event.shiftKey && event.which != 16)\n",
       "        value += \"shift+\";\n",
       "\n",
       "    value += 'k';\n",
       "    value += event.which.toString();\n",
       "\n",
       "    this._key_event_extra(event, name);\n",
       "\n",
       "    this.send_message(name, {key: value,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onclick = function(name) {\n",
       "    if (name == 'download') {\n",
       "        this.handle_save(this, null);\n",
       "    } else {\n",
       "        this.send_message(\"toolbar_button\", {name: name});\n",
       "    }\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onmouseover = function(tooltip) {\n",
       "    this.message.textContent = tooltip;\n",
       "};\n",
       "mpl.toolbar_items = [[\"Home\", \"Reset original view\", \"fa fa-home icon-home\", \"home\"], [\"Back\", \"Back to  previous view\", \"fa fa-arrow-left icon-arrow-left\", \"back\"], [\"Forward\", \"Forward to next view\", \"fa fa-arrow-right icon-arrow-right\", \"forward\"], [\"\", \"\", \"\", \"\"], [\"Pan\", \"Pan axes with left mouse, zoom with right\", \"fa fa-arrows icon-move\", \"pan\"], [\"Zoom\", \"Zoom to rectangle\", \"fa fa-square-o icon-check-empty\", \"zoom\"], [\"\", \"\", \"\", \"\"], [\"Download\", \"Download plot\", \"fa fa-floppy-o icon-save\", \"download\"]];\n",
       "\n",
       "mpl.extensions = [\"eps\", \"jpeg\", \"pdf\", \"png\", \"ps\", \"raw\", \"svg\", \"tif\"];\n",
       "\n",
       "mpl.default_extension = \"png\";var comm_websocket_adapter = function(comm) {\n",
       "    // Create a \"websocket\"-like object which calls the given IPython comm\n",
       "    // object with the appropriate methods. Currently this is a non binary\n",
       "    // socket, so there is still some room for performance tuning.\n",
       "    var ws = {};\n",
       "\n",
       "    ws.close = function() {\n",
       "        comm.close()\n",
       "    };\n",
       "    ws.send = function(m) {\n",
       "        //console.log('sending', m);\n",
       "        comm.send(m);\n",
       "    };\n",
       "    // Register the callback with on_msg.\n",
       "    comm.on_msg(function(msg) {\n",
       "        //console.log('receiving', msg['content']['data'], msg);\n",
       "        // Pass the mpl event to the overriden (by mpl) onmessage function.\n",
       "        ws.onmessage(msg['content']['data'])\n",
       "    });\n",
       "    return ws;\n",
       "}\n",
       "\n",
       "mpl.mpl_figure_comm = function(comm, msg) {\n",
       "    // This is the function which gets called when the mpl process\n",
       "    // starts-up an IPython Comm through the \"matplotlib\" channel.\n",
       "\n",
       "    var id = msg.content.data.id;\n",
       "    // Get hold of the div created by the display call when the Comm\n",
       "    // socket was opened in Python.\n",
       "    var element = $(\"#\" + id);\n",
       "    var ws_proxy = comm_websocket_adapter(comm)\n",
       "\n",
       "    function ondownload(figure, format) {\n",
       "        window.open(figure.imageObj.src);\n",
       "    }\n",
       "\n",
       "    var fig = new mpl.figure(id, ws_proxy,\n",
       "                           ondownload,\n",
       "                           element.get(0));\n",
       "\n",
       "    // Call onopen now - mpl needs it, as it is assuming we've passed it a real\n",
       "    // web socket which is closed, not our websocket->open comm proxy.\n",
       "    ws_proxy.onopen();\n",
       "\n",
       "    fig.parent_element = element.get(0);\n",
       "    fig.cell_info = mpl.find_output_cell(\"<div id='\" + id + \"'></div>\");\n",
       "    if (!fig.cell_info) {\n",
       "        console.error(\"Failed to find cell for figure\", id, fig);\n",
       "        return;\n",
       "    }\n",
       "\n",
       "    var output_index = fig.cell_info[2]\n",
       "    var cell = fig.cell_info[0];\n",
       "\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.handle_close = function(fig, msg) {\n",
       "    fig.root.unbind('remove')\n",
       "\n",
       "    // Update the output cell to use the data from the current canvas.\n",
       "    fig.push_to_output();\n",
       "    var dataURL = fig.canvas.toDataURL();\n",
       "    // Re-enable the keyboard manager in IPython - without this line, in FF,\n",
       "    // the notebook keyboard shortcuts fail.\n",
       "    IPython.keyboard_manager.enable()\n",
       "    $(fig.parent_element).html('<img src=\"' + dataURL + '\">');\n",
       "    fig.close_ws(fig, msg);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.close_ws = function(fig, msg){\n",
       "    fig.send_message('closing', msg);\n",
       "    // fig.ws.close()\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.push_to_output = function(remove_interactive) {\n",
       "    // Turn the data on the canvas into data in the output cell.\n",
       "    var dataURL = this.canvas.toDataURL();\n",
       "    this.cell_info[1]['text/html'] = '<img src=\"' + dataURL + '\">';\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Tell IPython that the notebook contents must change.\n",
       "    IPython.notebook.set_dirty(true);\n",
       "    this.send_message(\"ack\", {});\n",
       "    var fig = this;\n",
       "    // Wait a second, then push the new image to the DOM so\n",
       "    // that it is saved nicely (might be nice to debounce this).\n",
       "    setTimeout(function () { fig.push_to_output() }, 1000);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>')\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items){\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) { continue; };\n",
       "\n",
       "        var button = $('<button class=\"btn btn-default\" href=\"#\" title=\"' + name + '\"><i class=\"fa ' + image + ' fa-lg\"></i></button>');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    // Add the status bar.\n",
       "    var status_bar = $('<span class=\"mpl-message\" style=\"text-align:right; float: right;\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "\n",
       "    // Add the close button to the window.\n",
       "    var buttongrp = $('<div class=\"btn-group inline pull-right\"></div>');\n",
       "    var button = $('<button class=\"btn btn-mini btn-primary\" href=\"#\" title=\"Stop Interaction\"><i class=\"fa fa-power-off icon-remove icon-large\"></i></button>');\n",
       "    button.click(function (evt) { fig.handle_close(fig, {}); } );\n",
       "    button.mouseover('Stop Interaction', toolbar_mouse_event);\n",
       "    buttongrp.append(button);\n",
       "    var titlebar = this.root.find($('.ui-dialog-titlebar'));\n",
       "    titlebar.prepend(buttongrp);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(el){\n",
       "    var fig = this\n",
       "    el.on(\"remove\", function(){\n",
       "\tfig.close_ws(fig, {});\n",
       "    });\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(el){\n",
       "    // this is important to make the div 'focusable\n",
       "    el.attr('tabindex', 0)\n",
       "    // reach out to IPython and tell the keyboard manager to turn it's self\n",
       "    // off when our div gets focus\n",
       "\n",
       "    // location in version 3\n",
       "    if (IPython.notebook.keyboard_manager) {\n",
       "        IPython.notebook.keyboard_manager.register_events(el);\n",
       "    }\n",
       "    else {\n",
       "        // location in version 2\n",
       "        IPython.keyboard_manager.register_events(el);\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    var manager = IPython.notebook.keyboard_manager;\n",
       "    if (!manager)\n",
       "        manager = IPython.keyboard_manager;\n",
       "\n",
       "    // Check for shift+enter\n",
       "    if (event.shiftKey && event.which == 13) {\n",
       "        this.canvas_div.blur();\n",
       "        event.shiftKey = false;\n",
       "        // Send a \"J\" for go to next cell\n",
       "        event.which = 74;\n",
       "        event.keyCode = 74;\n",
       "        manager.command_mode();\n",
       "        manager.handle_keydown(event);\n",
       "    }\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    fig.ondownload(fig, null);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.find_output_cell = function(html_output) {\n",
       "    // Return the cell and output element which can be found *uniquely* in the notebook.\n",
       "    // Note - this is a bit hacky, but it is done because the \"notebook_saving.Notebook\"\n",
       "    // IPython event is triggered only after the cells have been serialised, which for\n",
       "    // our purposes (turning an active figure into a static one), is too late.\n",
       "    var cells = IPython.notebook.get_cells();\n",
       "    var ncells = cells.length;\n",
       "    for (var i=0; i<ncells; i++) {\n",
       "        var cell = cells[i];\n",
       "        if (cell.cell_type === 'code'){\n",
       "            for (var j=0; j<cell.output_area.outputs.length; j++) {\n",
       "                var data = cell.output_area.outputs[j];\n",
       "                if (data.data) {\n",
       "                    // IPython >= 3 moved mimebundle to data attribute of output\n",
       "                    data = data.data;\n",
       "                }\n",
       "                if (data['text/html'] == html_output) {\n",
       "                    return [cell, data, j];\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "}\n",
       "\n",
       "// Register the function which deals with the matplotlib target/channel.\n",
       "// The kernel may be null if the page has been refreshed.\n",
       "if (IPython.notebook.kernel != null) {\n",
       "    IPython.notebook.kernel.comm_manager.register_target('matplotlib', mpl.mpl_figure_comm);\n",
       "}\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<img src=\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAoAAAAHgCAYAAAA10dzkAAAgAElEQVR4XuydCZhU1Z3231tV3TS90DSIiqiAglFRYxRjEgEXSFAUF+hbjUl0NI6SGSbLlzhmNJ8ZklmeL4njuMRMiGtMBLpuS8aoLCoaAaO4K7LvKIuyNPQCTXdX3e+5RVdb3VR33apzb9U5dd5+nvnyhbrn3P953///nl/OuYsB/lEBKkAFqAAVoAJUgApopYCh1Wg5WCpABagAFaACVIAKUAEQAJkEVIAKUAEqQAWoABXQTAECoGaGc7hUgApQASpABagAFSAAMgeoABWgAlSAClABKqCZAgRAzQzncKkAFaACVIAKUAEqQABkDlABKkAFqAAVoAJUQDMFCICaGc7hUgEqQAWoABWgAlSAAMgcoAJUgApQASpABaiAZgoQADUznMOlAlSAClABKkAFqAABkDlABagAFaACVIAKUAHNFCAAamY4h0sFqAAVoAJUgApQAQIgc4AKUAEqQAWoABWgApopQADUzHAOlwpQASpABagAFaACBEDmABWgAlSAClABKkAFNFOAAKiZ4RwuFaACVIAKUAEqQAUIgMwBKkAFqAAVoAJUgApopgABUDPDOVwqQAWoABWgAlSAChAAmQNUgApQASpABagAFdBMAQKgZoZzuFSAClABKkAFqAAVIAAyB6gAFaACVIAKUAEqoJkCBEDNDOdwqQAVoAJUgApQASpAAGQOUAEqQAWoABWgAlRAMwUIgJoZzuFSASpABagAFaACVIAAyBygAlSAClABKkAFqIBmChAANTOcw6UCVIAKUAEqQAWoAAGQOUAFqAAVoAJUgApQAc0UIABqZjiHSwWoABWgAlSAClABAiBzgApQASpABagAFaACmilAANTMcA6XClABKkAFqAAVoAIEQOYAFaACVIAKUAEqQAU0U4AAqJnhHC4VoAJUgApQASpABQiAzAEqQAWoABWgAlSACmimAAFQM8M5XCpABagAFaACVIAKEACZA1SAClABKkAFqAAV0EwBAqBmhnO4VIAKUAEqQAWoABUgADIHqAAVoAJUgApQASqgmQIEQM0M53CpABWgAlSAClABKkAAZA5QASpABagAFaACVEAzBQiAmhnO4VIBKkAFqAAVoAJUgADIHKACVIAKUAEqQAWogGYKEAA1M5zDpQJUgApQASpABagAAZA5QAWoABWgAlSAClABzRQoKAA0TfN+AFcbhjE0FoudW1dX92HCz3A4PNO27esBHAaw27Ks8c5vpmn2BfAogAsARAOBwE9ra2uf1iwPOFwqQAWoABWgAlRAIwUKDQDHhEKhTe3t7cts2742AYCmaf4AwFgANZZlRadMmXLsvHnzPnN8DofDdwMYHolEvjNt2rRhsVhseSgUOn327Nn1bvKgpqZmYm1t7SI3xxbSMRx3IbmZfiz0O71GhXQE/S4kN9OPRVe/0ytT2EcUFAAmrDJNc7Nt29ckADAcDn9s2/allmVt6G5nOBz+yLbt71iW9WbHiuBcwzBeiEQij7mx3jTNey3L+pGbYwvpGI67kNxMPxb6nV6jQjqCfheSm+nHoqvf6ZUp7CMKHgCvueaaiuLi4r2GYdwFoNq2bdswjP+ORCKRjhXAhmAwOHLOnDmfdvz3XwI4FIlEZrqxXtfC4bjdZEfhHEO/C8dLNyOh325UKpxjdPW7cBzMbiQFD4Df/OY3q9rb2/fatn23ZVn/MXXq1KGBQOBvAC63LGtFOBzOGgAdlrz99tt/d8899/xzdvKr22r69On/OWvWLAeqtfrjuLWyG/SbfuuggK55DqACwA7DMGwdfO4+xoIHQGfApmkeCAaDX5w7d+6WjlU+Z/VvobPNa5rmCgC3JG0B1xqGsainLWDnXolYLDbR6efkk08uu+eee27TMXE4ZipABagAFaACqitw++23/37btm3NzjgCgcAine7p1wIAw+Hw72zb/sCyrP8xTXOAYRjvxmKx6rq6urdN0/xX56lh5yEQ0zSHA3gdwJmWZe1Ll9i2bfcDcGDXrl2wbb3+B0RFRQUaGxvTSVRwv3PcBWdprwOi3/RbBwV0zHPDMHD88cc79lYahtGgg88FvQLogB6AK23bPg6Ac99fYyQSOc2BPgCPAzgFgENqD1mWNcsRY/LkyaUlJSXOAx+jAbRn8hqYBADu3LlTOwDs168fGhr0qxmOW6/LJP2m3zoooGOeOwA4ePBgAqAOCe7HGAmABEA/8krGPnWcIBwfOG4Zs9G/mOi3f9rK1jMBECjILeBcJRoBkACYq1zL93k4Mebbgdyen37nVu98n01HvwmABEChuiMAEgCFEkihxjpOEFwBZH0rVKJCoepY3wRAAqBQ0RAAOUEIJZBCjXWcIAiArG+FSlQoVB3rmwBIABQqGgIgJwihBFKosY4TBAGQ9Z2PEnWeyHXgJJd/wWAQ0Wg0l6fM2bmcN3SkemMFAZAAKJSEBEBOEEIJpFBjAqBCZnkQKv32QMQsu9BV+yzlStusJz0JgATAtMnT2wEEQAKgUAIp1FjXSYnjVihJPQhVBr9liMEDKaXpggDYsxW5XWeWJiW8CYQASAD0JpPk70XXSYnjlj83vYxQBr9liMFLTfPdFwGQAOhLDhIACYC+JJaEneo6KXHcEiajjyHJ4LcMMfgocc67JgASAH1JOgIgAdCXxJKwU10nJY5bwmT0MSQZ/JYhBh8lznnXBEACoC9JRwAkAPqSWBJ2quukxHFLmIw+hiSD3zLEkKnEq1evxoMPPog33ngDzc3NGDBgAM4//3x897vfxVlnnYVDhw7ht7/9LebPn4+tW7eirKwMxx57LCZMmIBbbrkFxxxzTKandH08AZAA6DpZMjmQAEgAzCRfVD5WxUnJC705bi9UVKcPGfyWIYZMHPvb3/6GG264If5/t956K4YMGRJ/7cqCBQuwefNmfP/738d1110H51UzM2fOjANh3759sWXLFsybNw+nn346Jk2alMkpMzqWAEgAzChh3B5MACQAus0V1Y9TbVLySm+O2ysl1ehHBr9liCETt8aNG4fzzjsP9913X8pm999/Px555BEsW7YMlZWVPXbd1taGn/3sZ1i4cGF8xbCqqgrTp0/HTTfdlEk4Rx1LACQACiVQT40JgARAXxJLwk5Vm5S8kpDj9kpJNfqRwe90MTgvNkbLIW8FLemb1cunnRW+sWPHYs6cOfH/TPV3zTXXYOjQoXjggQd6jXn27Nl48sknMXfuXPTv3x979uzBrl274iuGIn8EQAKgSP702JYASAD0JbEk7DTdpCRhyJ6ExHF7IqMyncjgd7oY7EMHEfv+NE81DTwwF0bf0oz7fOuttzBlyhS88sorGDFiRMr2Y8aMwZVXXok777yz83enzZo1a+Cs+jnbw7/61a9gWVYcEu+55574/YOhUCjjeFI1IAASAD1JpO6dEAAJgL4kloSdppuUJAzZk5A4bk9kVKYTGfxOF4OKK4Ann3xy/CGR7n/f+973UFRUhHvvvRexWAwPP/wwnn32WWzcuBGjR4/GHXfcgbPPPlsofwiABEChBOqpMQGQAOhLYknYabpJScKQPQmJ4/ZERmU6kcFvGWLIxDA39wA++uijWLJkSXxrN/kvGQCT/925B9BZFXRg8O23384knKOOJQASAIUSiAB4tAKqXaS8SgCO2ysl1eiHfqvhk1dRyuC3DDFkoufrr78efwL4xhtvjL/SxXkKuKmpKf4wx6ZNm+BAnnMfYHFxMe6++26cc8458aeAt23bFn9C+JRTTomvAL722muoqKjAGWecEb8f0XmopK6uLv5qGZE/AiABUCR/emyr8wrgs7X7ccnlFaioDPqiraydqnZx9kpHjtsrJdXoh37nzycVtXfu53Pu33NeCeOs3jnvAXS2cJ33AI4aNarzPYDPPfccPv74Y5SWluL444/H+PHjcfPNN8ffCfjMM8/E+/jkk0/i9/85IOg8FewAo8gfAZAAKJI/BMAUCjgAeN5XSjFkaLEv2sraqYoXZy+05Li9UFGdPuh3/rzSVXu/FCcAEgB9yS3dVwAJgL6klZSd6jopcdxSpqNvQcngtwwx+CZwHjomABIAfUk7AiBXAH1JLAk71XVS4rglTEYfQ5LBbxli8FHinHdNACQA+pJ0BEACoC+JJWGnuk5KHLeEyehjSDL4LUMMPkqc864JgARAX5KOAEgA9CWxJOxU10mJ45YwGX0MSQa/ZYjBR4lz3jUBkADoS9IRAAmAviSWhJ3qOilx3BImo48hyeC3DDH4KHHOuyYAEgB9SToCIAHQl8SSsFNdJyWOW8Jk9DEkGfyWIQYfJc551wRAAqAvSUcAJAD6klgSdqrrpMRxS5iMPoYkg98yxOCjxDnvmgBIAPQl6QiABEBfEkvCTnWdlDhuCZPRx5Bk8FuGGHyUOOddEwAJgL4kHQGQAOhLYknYqa6TEsctYTL6GJIMfssQg48S57xrAiAB0JekIwASAH1JLAk71XVS4rglTEYfQ5LBbxli8FHinHdNACQA+pJ0BMCeAdD5VNylkypQXlFY3wrW9eLMcftyCZG2U/qdP2tU1H716tV48MEH8cYbb6C5uTn+LeDzzz8//i3gVatW4Ve/+hXefvvtLqK+/vrrME0T27ZtQyAQiP/mfEf4t7/9LebPn4+tW7eirKws/p3gCRMm4JZbbsExxxyTsTEEQAJgxknjpgEBsHcAvGBMGY4fUuRGSmWOUfHi7IW4HLcXKqrTB/3On1eqaf+3v/0NN9xwQ/z/br31VgwZMgSNjY1YsGABNm/ejOHDh+PXv/413nrrraMAMBwOx0HPAUAH/q677joEg0HMnDkTZ511Fvr27YstW7Zg3rx5OP300zFp0qSMjSEAEgAzTho3DQiABEA3eVIIx6g2KXmlOcftlZJq9COD3zLEkIlb48aNw3nnnYf77rsvZbNIJOIKAO+//3488sgjWLZsGSorK3sMoa2tDT/72c+wcOHCODRWVVVh+vTpuOmmm1K2IQASADPJZ9fHEgAJgK6TRfEDVZuUvJKb4/ZKSTX6kcHvdDHYto1D7TFPBe0bCsAwjIz7dFb4xo4dizlz5sT/M9VfTwDorBzW1NR0rgBec801GDp0KB544IFe45g9ezaefPJJzJ07F/3798eePXuwa9eu+Iphqj8CIAEw48R204AASAB0kyeFcEy6SakQxpjJ5FGo402Mi37nz+F02h9si+L6yHpPA5wTHonSoszv13a2dadMmYJXXnkFI0aM6BEAf/zjH8MZV/Kfs5LnrOAltoDHjBmDK6+8EnfeeWfnYU7fa9asgXOssz3s3EtoWVYcEu+55574fYahUKhXLQiABEBPiyXRGQGQAOhLYknYabpJScKQPQmJ4/ZERmU6kcHvdDEUygqg8xBI8j2AzgrgySefHH+YpPvf9773PRQVFeHee+9FLBbDww8/jGeffRYbN27E6NGjcccdd+Dss8/mCmCGlZb5mm+GJyjkwwmABMBCzu/ksaWblApVB467UJ1NPS4Z/JYhhkxc9/IewEcffRRLliyJb+0m/yUDYPK/OyuIzqqgA4PdnzJOt5rtbHkPHjzYOazSMIyGTMZcKMcSAAWc1B0Av/SVUpw4tDilgs5rYPgUsEBySdZUtUnJK/k4bq+UVKMfGfyWIYZM3HJW8pwngG+88cb4q1qcp4CbmpriD2ls2rQJw4YNc/UQyMGDB+PbvM5K3913341zzjkn/hSw85qY73//+zjllFPiK4CvvfYaKioqcMYZZ8TvW3QePqmrq4u/gibVH7eAe3aTAJhJpnc7lgBIABRIH6WaqjYpeSUux+2Vkmr0I4PfMsSQqVvOfXrOfXnOgx3OqpzzHkBna9Z5D+DKlStdAaBzzsR7AJ977jl8/PHHKC0txfHHH4/x48fj5ptvjr8T8Jlnnomf65NPPonf/+eAoPNUsAOMBMDMnCMAZqZXl6MJgARAgfRRqqmKk5IXAnPcXqioTh8y+C1DDOo4lj5SrgBqsgJomub9AK42DGNoLBY7t66u7sPkoZumeRmAFwzD+FEkEok/a26aZl8AjwK4AEA0EAj8tLa29un0aQUQAAmAbvKkEI7RdVLiuAshe92PQQa/ZYjBvWLyH0kA1AcAx4RCoU3t7e3LbNu+NhkAv/3tb/c7fPjwiwB2GYaxOAGA4XD4bgDDI5HId6ZNmzYsFostD4VCp8+ePbs+XWoTAAmA6XKkUH7XdVLiuAslg92NQwa/ZYjBnVpqHEUA1AQAE8M0TXOzbdvXJAOgaZpP2rYdMQxjqmEY7yUB4Ee2bX/Hsqw3O1YE5xqG8UIkEnksXXoTAAmA6XKkUH7XdVLiuAslg92NQwa/ZYjBnVpqHEUA1BwAa2pqpsZisassy7rZNM3HuwFgQzAYHDlnzpxPHZnC4fAvnXtRI5HIzHTpTQAkAKbLkUL5XddJieMulAx2Nw4Z/JYhBndqqXEUAVBjALz++uuPa29vX9inT5+L//SnPzUQAL0pWuc1L3wNjDdaqtCLrpMSx61CdnoXowx+yxCDd4rmvycCoMYAWFNTMykWizkPeRwC4Dz1fIxhGIcB/E8kErk7HA533wKuNQxjUU9bwDU1NRNjsdhER9KqqqriWbNmzXDeeaTb31MPb8PXLh2I4SPKUg7d+f3ibxyDE4eWFpQ0xcXFaG1tLagxuRkMx+1GpcI5hn7nz8tgMIj6+rS3oOcvQMXOXFVVhWg0mjLq8vJyTJ8+/aH6+vr4RT0QCCyqra1dpNgQsw63IF8Dk+oewIRC3VcATdP8V+epYechENM0hwN4HcCZlmXtS6cqt4C5BZwuRwrld11XJTjuQslgd+OQwW8ZYnCnlhpHcQVQkxXAcDj8OwBX2rZ9HIC9hmE0RiKR05KHHw6HnYc73k88BDJ58uTSkpIS599GA2jna2DcFTW3gN3pVChH6TopcdyFksHuxiGD3zLE4E4tNY4iAGoCgLlOR64AcgUw1zmXr/PpOilx3PnKuPycVwa/ZYghP+r7c1YCIAHQl8wiABIAfUksCTvVdVLiuCVMRh9DksFvGWLIVOLVq1fjwQcfjH+Pt7m5Of4puPPPPz/+KbhVq1bhV7/6Fd5+++0u3TrfEDZNM/6t30AggEgk4uq4TGMjABIAM80ZV8cTAAmArhKlAA5ScVLyQnaO2wsV1elDBr9liCETx5zv/95www3x/7v11lsxZMgQNDY2YsGCBdi8eTOGDx/u6lvADgD++te/xltvvXUUKIbDYWzdujUOipn+EQAJgJnmjKvjCYAEQFeJUgAHqTYpeSU5x+2Vkmr0I4PfMsSQiVvjxo3Deeedh/vuuy9lM7dg5/Y4Z0XxZz/7WXxl0TAMDB06FL/5zW9wyimnpDw/AZAAmEk+uz6WAEgAdJ0sih+o2qTkldwct1dKqtGPDH7LEINbt5wVvrFjx2LOnDnx/0z11xPYOSuHNTU1nSt7bo+79tprcckll+AHP/gBbNvGmjVrcNxxx2HgwIEEQLfGdRxXkK+ByVCDrA8nABIAs04exRqqNCl5KS3H7aWa8vclg9/pYnCgp73dWy1DIcRX0zL9c7Zrp0yZgldeeQUjRozoEQB//OMfwxlX8l9bWxsOHTrUBQDdHOdsBw8bNgwzZsyIr/6l++MKIFcA0+VIVr8TAAmAWSWOgo3STUoKDslVyBy3K5kK5iAZ/E4XQ1ubjYXzDniq+eVTKlFUlDkAiqwAOg+BJN/b53YLePv27bj//vvx6quvIhaLYdKkSfjJT36C0tLUHx0gABIAPS2WRGcEQAKgL4klYafpJiUJQ/YkJI7bExmV6UQGv9PFINMKoGNsru8BTE4m58GQm2++GZdffjnuuOOOlHlGACQA+nIB0h4ALyzFicOKU2rrvCj6gjFlOH5IkS/a56vTdBfnfMXl93k5br8Vlqt/+p0/P1TT3lnJc54AvvHGG3HLLbfEnwJ2Po+6cOFCbNq0Kb5d6+bpXrcrgM5xzv2GgwcPxp49e3D99dfjiiuuwI9+9CMCYIZpm/mab4YnKOTDCYAEwELO7+SxqTYpeeULx+2Vkmr0I4PfMsSQqVvOgxgPPPAAnAc7nPv6nPcAjh49Ov4ewJUrV3oKgD/84Q+xdOnS+KtmKioqMHHixPhTwSUlJQTADI0jAGYoWPLhBEACoED6KNVUxUnJC4E5bi9UVKcPGfyWIQZ1HEsfKbeAuQWcPkuyOIIASADMIm2UbKLrpMRxK5muWQctg98yxJC1gBI2JAASAH1JSwIgAdCXxJKwU10nJY5bwmT0MSQZ/JYhBh8lznnXBEACoC9JRwAkAPqSWBJ2quukxHFLmIw+hiSD3zLE4KPEOe+aAEgA9CXpCIAEQF8SS8JOdZ2UOG4Jk9HHkGTwW4YYfJQ4510TAAmAviRdAgC3btmOomK9nqdxXvPyJb4Gxpe8krFTXScljlvGbPQvJhn8liEG/xTOfc8EQAKgL1mXAMB5c9bjKxeX+XIOWTslAMrqjD9x6Topcdz+5JOsvcrgtwwxyOpPNnERAAmA2eRN2jYJAJzz+DpcPLE87fGFdAABsJDcTD8WXScljjt9bhTSETL4LUMMOnjqfPvYeZk0gErDMBoKacxux6LXvqVbVVweRwDkPYAuU0X5w3SdlDhu5VM3owHI4LcMMWQkmuQHcwWQK4C+pCgBkADoS2JJ2KmukxLHLWEy+hiSDH47X7dwVqdy+RcMBhGNRnN5ypydy/l2svPVkO5/XAEEcptlObM8NyciABIAc5Np+T+LDBNjPlTguPOhev7OSb/zp32uz0wAJAAK5RwBkAAolEAKNebEqJBZHoRKvz0QUaEudPSbAEgAFCpRAiABUCiBFGqs4wTh2MNxK5SkHoRKvz0QUZEuCIAEQKFUJQASAIUSSKHGnBgVMsuDUOm3ByIq1IWOfhMACYBCJUoAJAAKJZBCjXWcILgCqN+bMZjnCl2UBEMlABIAhVKIAEgAFEoghRpzYlTILA9Cpd8eiKhQFzr6TQAkAAqVKAGQACiUQAo11nGC4AogVwAVKlGhUHWsbwIgAVCoaAiA6QFw4KAgDh+2UV4RFNJalsY6XigJQgQhWerP7zhY334rLE//BEACoFA2EgDTA+CWDYexe1c7Jtf0F9JalsacIGRxIjdx0O/c6CzLWei3LE74HwcBkAAolGWdAPjYWlx8eYVQX6o1dvst4LUrDqHhQIwAqJrB3eLlxKi4gRmGT78zFEzxw3X0mwBIABQqWwJg+hVAAqBQiknTWMcJglvf3PqWpgB9DkTH+iYAEgCFyooASAAUSiCFGus4QRAACYAKlahQqDrWNwGQAChUNARAAqBQAinUWMcJggBIAFSoRIVC1bG+CYAEQKGiIQASAIUSSKHGOk4QBEACoEIlKhSqjvVNACQAChUNAZAAKJRACjXWcYIgABIAFSpRoVB1rG8CIAFQqGgIgARAoQRSqLGOEwQBkACoUIkKhapjfRMACYBCRUMAJAAKJZBCjXWcIAiABECFSlQoVB3rmwBIABQqGgJgegB8b3kz2tvA9wAKZVr+G+s4QRAACYD5r7zcRKBjfRMACYBC1UUATA+Aby1rjmvML4EIpVreG+s4QRAACYB5L7wcBaBjfRMACYBC5UUAJAAKJZBCjXWcIAiABECFSlQoVB3rmwBYYABomub9AK42DGNoLBY7t66u7kOnKsLh8GO2bV8E4CCAJtu2/09dXd3bzm+mafYF8CiACwBEA4HAT2tra592U00EQAKgmzwphGN0nCAIgATAQqhdN2PQsb4JgIUHgGNCodCm9vb2ZbZtX5sAwOrq6qvOOuus+TNnzoyZpnklgN9YljW8Aw7vBjA8Eol8Z9q0acNisdjyUCh0+uzZs+vTFQ4BkACYLkcK5XcdJwgCIAGwUOo33Th0rG8CYIEBYCLJTdPcbNv2NQkATE7+6667bmAoFNoxatSovg4QhsPhj2zb/o5lWW92rAjONQzjhUgk8li6oiEAEgDT5Uih/K7jBEEAJAAWSv2mG4eO9U0A1BAAw+Hwv9m2fZZlWdd1rAA2BIPBkXPmzPm047//EsChSCQyM13REAAJgOlypFB+13GCIAASAAulftONQ8f6JgBqBoDhcPjbAH5q2/Y4y7J2EwDTXRZ6/v3Z2v340oUEwOwVVKuljhMEAZAAqFaVZh+tjvVNANQIAE3TrAHwCwCXWZa1PWm7eAWAW5K2gGsNw1jU0xZwTU3NxFgsNtFpX1VVVTxr1qwZc59Yj6uqB2dffQq2fOrhbfjapQMxfERZyuid3y/+xjF49YU98d+/devJCo7y6JCLi4vR2tpaEGPJZBAcdyZqqX8s/Vbfw0xGoKvf5eXlmD59+kP19fXxi3ogEFhUW1u7KBPtVD7WUDn4nmLvfg9gOBwOA/j3QCAwfu7cuR8ntzNN81+dp4adh0BM03QeDHkdwJmWZe1Lpw23gMVWANvbbISK1EpBHf+XMlfCuBKW7lpYKL+zvgvFyfTj4Apgga0AhsPh3wG40rbt4wDsNQyjMRKJnGaapkP3Ow3D2GvbtmEYhh0KhcY7T/pOnjy5tKSkxHngYzSAdr4GJn3hOEeIbgHv39eOpS82KfeCaE4Q7vKjUI6i34XipLtx0G93OhXCUQTAAgPAXCclVwCzXwH8dEcb3lzaTADMddJmeT5OjFkKp2gz+q2ocVmGraPfBEACYJblcqRZIQHgu28045SRfdB/YMiVJqIrgARAVzJLc5COEwS3vrn1LU0B+hyIjvVNACQACpVVIQGgA3QnDS/GuV8udaUJAdCVTAVzkI4TBAGQAFgwBZxmIDrWNwGQAChU3wRAbgELJZBCjXWcIAiABECFSlQoVB3rmwBIABQqGgIgAVAogRRqrOMEQQAkACpUokKh6ljfBEACoFDREAAJgEIJpFBjHScIAiABUKESFQpVx/omABIAhYqGANg7ACaLO7mmfxet+RCIUNZVqr0AACAASURBVOrlvLGOEwQBkACY80LL0wl1rG8CIAFQqNwIgARAoQRSqLGOEwQBkACoUIkKhapjfRMACYBCRUMAJAAKJZBCjXWcIAiABECFSlQoVB3rmwBIABQqmgQAzn5sLS65vEKor3w39uM1MNwCzrer3p1fxwmCAEgA9K6C5O5Jx/omABIAhaqSAMgVQKEEUqixjhMEAZAAqFCJCoWqY30TAAmAQkVDACQACiWQQo11nCAIgARAhUpUKFQd65sASAAUKhoCIAFQKIEUaqzjBEEAJAAqVKJCoepY3wRAAqBQ0fgJgAfq27Fh9WGc/7UyoRjdNuY9gO6U0vFCSRAiCLmrDvWPYn2r76HbERAACYBucyXlcX4+Bfzh2wexdWMrur8/TyjgXhoTAN0pywnCnU6FchT9LhQn3Y2DfrvTqRCOIgASAIXymADILWChBFKoMSdGhczyIFT67YGICnWho98EQAKgUIkSAAmAQgmkUGMdJwhufXPrW6ESFQpVx/omABIAhYpGdwA898JSnDSsOKWGzpZy8h8/BSeUanlvrOMEQQAkAOa98HIUgI71TQAkAAqVl+4A6IjX0z2KBECh1JKusY4TBAGQAChdIfoUkI71TQAkAAqVEwGQACiUQAo11nGCIAASABUqUaFQdaxvAiABUKhoCIAEQKEEUqixjhMEAZAAqFCJCoWqY30TAAmAQkVDACQACiWQQo11nCAIgARAhUpUKFQd65sASAAUKhqVAbC9zUaoyOgcfzbvAeQ9gELpo1RjHScIAiABUKkiFQhWx/omABIABUoG8PNLIH6+CLqxIYq/Lmjs8gAHAdBdKuh4oSQIEYTcVYf6R7G+1ffQ7QgIgARAt7mS8jhVAXDv7nb87eUmAmAW7nOCyEI0hZvQb4XNyyJ0+p2FaIo2IQASAIVSNxcAWFkVxEXjyxEMfr5dKxQ0AAJg9gpygsheOxVb0m8VXcs+ZvqdvXaqtSQAEgCFcjYXAOgEOGFyP/QtDQjFmty4RwAcVgzn5c5u/hLv+eN7AN2opf4xnBjV9zCTEdDvTNRS/1gd/SYAEgCFKpcAyKeAhRJIocY6ThCOPRy3QknqQaj02wMRFemCAEgAFEpVPwFwxTsHsWVDazw+rgAK2eRpY04QnsopfWf0W3qLPA2Qfnsqp9SdEQAJgEIJSgDkCqBQAinUmBOjQmZ5ECr99kBEhbrQ0W8CIAFQqEQJgAIAuLMNby5p7vFbwkLG+NhYxwslt0L5GhgfS0qqrlnfUtnhazAEQAKgUIIRAAmAQgmkUGNOjAqZ5UGo9NsDERXqQke/CYAEQKESLTQA7Nc/gIsn9nOlSbZPAbe329j5SRuK+xhcAXSltBwH6ThBcOWTK59yVJ//UehY3wRAAqBQZRUaADpi9PRal+5CZQuAm9Ydxsr3DuHL48oIgELZl9vGOk4QBEACYG6rLH9n07G+CYAEQKGKIwBmvgVMABRKubw11nGCIAASAPNWcDk+sY71TQAkAAqVGQGQACiUQAo11nGCIAASABUqUaFQdaxvAiABUKhoOgHw0bW45IoKob66N87HewC5BZzeQh0vlAQhglD6yiiMI1jfheGjm1EQAAmAbvKkx2MIgFwBFEoghRpzYlTILA9Cpd8eiKhQFzr6TQAkAAqVKAGQACiUQAo11nGC4MonVz4VKlGhUHWsbwIgAVCoaHIFgOOv6ofSsoDrWPfva0f/AaEej9+7ux1/e7mpyxO/6Z7q7d5ZuuMTvyfaJZ4u5kMgrm2U6kAdJwgCIAFQqiL0MRgd65sAWGAAaJrm/QCuNgxjaCwWO7euru5Dp2ZM0xwE4EnDME61bbvFtu0ZdXV1Szt+6wvgUQAXAIgGAoGf1tbWPu2m1mQEwIPNMSx+rgETr+2H4j6poZEA6Mbd1MfoeKEkCBGEsq8YtVqyvtXySyRaAmDhAeCYUCi0qb29fZlt29cmAaADeFsty/pFdXX16EAg8GfbtodZlhUNh8N3AxgeiUS+M23atGGxWGx5KBQ6ffbs2fXpkktGAGxqjOKV+Y34xjX90KeEAJjOw0x/5wSRqWJqH0+/1fYv0+jpd6aKqXs8AbDAADCRiqZpbrZt+5okAGyMRqOnzps37zPnmHA4/IZt23dZlvVyOBz+yLbt71iW9WbHiuBcwzBeiEQij6VLbVUBcN/udrz2chOuClfCKQLnL92WLreAjyjACSJdVRTW7/S7sPxMNxr6nU6hwvmdAKgBAJqmOQDAdsuynK3e+J9pmrUAFliW9UQ4HG4IBoMj58yZ82kHHP4SwKFIJDIzXaoTAPkQSLocKZTfOTEWipPuxkG/3elUKEfp6DcBkABIAOQKYEbXcB0vlFz55D2AGRWJwgezvhU2L8PQCYAaAGDHil+XLWDTNJcDuNPZAjZNcwWAW5K2gGsNw1jU0xZwTU3NxFgsNtHpt6qqqnjWrFkz5j6+HleZgzNMv94Pf+u1fVi3qil+0LXXn4Cy8p6f6k3uqeFAG56N7MTUbw9BSd9gypN8tuswXnz2U3zz70/q3AJ+6uFt8WO/devJrsaR7vjE74nOEv2u+agB77y+H5dePgivLNzt+nyugsrBQcXFxWhtbc3BmeQ6Bcctlx9+R0O//VZYrv519bu8vBzTp09/qL6+Pn5RDwQCi2praxfJ5Y5/0Ry5AazA/rrfAxgOhx+zbdt5COTnNTU1F9i2PS/xEIhpmv/qPDXsPARimuZwAK8DONOyrH3pZOEWMLeA0+VIofzOlZFCcdLdOOi3O50K5Sgd/eYKYIGtAIbD4d8BuNK27eMA7DUMozESiZw2ZcqUY0Oh0B+dp31t2z4cCARm1NbWLnGKd/LkyaUlJSXOAx+jAbSr/hoYN08B8yGQ7C/bOl4oHbU47uxzRsWW9FtF17KPWUe/CYAFBoDZp392LbkCmMUK4NoWrHy/BV8eV4Y3lzR3eRl1di7ktpWOF0oCIO8BzG2V5e9srO/8aZ/rMxMACYBCOecnAH707kFsXn/kXrNMvgQi/QogAVAo5/LVmBNjvpTPz3npd350z9dZdfSbAEgAFKo3AiBXAIUSSKHGOk4QXPnkyqdCJSoUqo71TQAkAAoVjbIAuKcdry3O04uguQIolHP5aqzjBEEAJADmq95yfV4d65sASAAUqjMZAbC5MYqX03wKbh8BMGvfdbxQEoQIQlkXjGINWd+KGSYQLgGQACiQPgABkFvAQgmkUGNOjAqZ5UGo9NsDERXqQke/CYAEQKESJQASAIUSSKHGOk4QXPnkyqdCJSoUqo71TQAkAAoVDQGQACiUQAo11nGCIAASABUqUaFQdaxvAiABUKhoCIDiAHjV6F0wTj1dyIdcNtbxQkkQIgjlssbyeS7Wdz7Vz+25CYAEQKGMIwCKA+Ckl25E8OG/CPmQy8acIHKpdv7PRb/z70EuI6DfuVQ7v+ciABIAhTKQAEgAFEoghRpzYlTILA9Cpd8eiKhQFzr6TQAkAAqVKAEwCwBcdxgr3zvU+Sk4rgAKpWDOGus4QXDrm1vfOSuwPJ9Ix/omABIAhcpOeQA0K2EEjLgGz9buj//n5Jr+rjRJd3zi90RniX43dQDghePKsHxJMwiAruTO+0E6ThAEQAJg3gsvRwHoWN8EQAKgUHkRALNfAUwGwHf/bg5OP7sEFZVBIT9y0VjHCyVBiCCUi9qS4RysbxlcyE0MBEACoFCmEQC9AcD5E57EiDP64Ixz+gr5kYvGnCByobI856Df8niRi0jody5UluMcBEACoFAmSgmATVG8/HwjvnFNP/QpCaQcX+en4CTZAiYACqVhThpzYsyJzNKchH5LY0VOAtHRbwIgAVCouAiAWawArm3ByvdbkLwFTAAUSsOcNNZxguDWN7e+c1JcEpxEx/omABIAhUqPAEgAFEoghRrrOEEQAAmACpWoUKg61jcBkAAoVDS5A8AKlJa5e0CiWfYt4G4rgMfufg+fDfoS7wEUykT/G+s4QRAACYD+V5YcZ9CxvgmABECh6iMAiq8AJgzgQyBCqeh7Yx0nCAIgAdD3wpLkBDrWNwGQAChUfgRAAqBQAinUWMcJggBIAFSoRIVC1bG+CYAEQKGiIQASAIUSSKHGOk4QBEACoEIlKhSqjvVNACQAChVN7gCwH0rLUr/SpfsAMrkH8EqzEoFcfwmk2z2A3AIWSsGcNdZxgiAAEgBzVmB5PpGO9U0AJAAKlR0BkCuAQgmkUGMdJwgCIAFQoRIVClXH+iYAEgCFioYASAAUSiCFGus4QRAACYAKlahQqDrWNwGQAChUNDID4Nev7oeSvke2jZsao1j53iFcOK48/t8TXwLhFnDm9ut4oSQIEYQyrxQ1W7C+1fQtm6gJgATAbPKms40qALh+VQvWrGjB5Jr+BEAhxwFOEIICKtacfitmmGC49FtQQIWaEwAJgELpSgDkFrBQAinUmBOjQmZ5ECr99kBEhbrQ0W8CIAFQqEQJgARAoQRSqLGOEwS3vrn1rVCJCoWqY30TAAmAQkXjKwC+dwib1x2Oxzf+qsxfA5N8DyC3gIVs7tJYxwslQYgg5F0Fyd0T61tuf7yMjgBIABTKJ2UAcHUL1nz4+T2A9XvasWxxE/gQSOb2c4LIXDOVW9Bvld3LPHb6nblmqrYgABIAhXKXAMgtYKEEUqgxJ0aFzPIgVPrtgYgKdaGj3wRAAqBQiXoFgLs/bUPf0gDKK4Kd8XzUZQu4AqVln//WW9CJL4F02QKWaAVw49oWrHq/BReOK8PyJc2dQxlxRh+ccU5fIT9y0VjHCyW3gLkFnIvakuEcrG8ZXMhNDARAAqBQpnkFgM/W7kffUgMTJlf6AoAbVrdgtSRbwARAoZTLW2NOjHmTPi8npt95kT1vJ9XRbwIgAVCo4JIBcOw3ymEACASd/zezvzgAlgUw4ap+BMDMpMv50TpeKLkCyBXAnBdank7I+s6T8Hk4LQGQACiUdskA2Noai2/jjv16RcZ9EgABbgFnnDY5bcCJMady5/1k9DvvFuQ0AB39JgASAIWKLBkAGxui8b4SX9vIpONUAOh8um1T52tgxO4BlGkLeNPaFqzkPYCZpIcUx+o4QXDlkyufUhRfDoLQsb4JgARAodKSEQAPNkWx+PlGJD8EogQAnt4HZ3yRD4EIJaSPjXWcIAiABEAfS0qqrnWsbwIgAVCoCAmAHr4GhgAolIt+N9ZxgiAAEgD9ritZ+texvgmABECh+lMWAPe2Y9lLmb0Iuq01hsOH7c5X1Tjb1r1teSd+Twic2BrvcQuYACiUi3431nGCIAASAP2uK1n617G+CYCaAWBNTc2kWCz2b87DugCcF+vdY1nWk6ZpDgLwpGEYp9q23WLb9oy6urql6YpTeQCsrux8ajkd0L3xahN272rvvMcx3fEEwHTZo9bvOk4QBEACoFpVmn20OtY3AVAzADRNc69t2+Pq6upWTp06dWggEFhTUlJyzOHDhx+0bXuLZVm/qK6uHh0IBP5s2/Ywy7KOPNnRw59OAPjqwgY0HIgRAPv1Q0MDJ8bspxq1Wuo4MRJ8Wd9qVWl20RIA9QPA3QCusyxrWXV19TmBQOD5/v37n1JfX78vGo2eOm/evM+cVAqHw2/Ytn2XZVkvEwCPKOA5AF5chuWvJn0JhFvA2V3FctSKIJQjoSU5Df2WxIgchaGj3wRAzQCwurp6vGEYcw3DaLZtuz+AKQDeB7DdsqzOR1BN06wFsMCyrCfcAOBTj6xFU6Pga2C6fQkk29fApHwKeE0LVn/Q0rl6V5+4BzCDLWAC4JFM0PFCyXFzRShHHJL307C+825BzgIgAGoEgKZpBg3DeMkwjP9bW1v7WsdW71/a29vPDYVC2yKRSEki81QAwGjUxvatrTj5lD5dCkZVACzpa+DrV3/+KbycXQUyPBEniAwFU/xw+q24gRmGT78zFEzhwwmAegHg+QCesizr9CTQezMQCNwVi8X+nLwFbJrmcgB3ptoCrqmpmRiLxSY6fVRVVRXPmjVrxpzH1qPhQFu822/denLGJfHUw9tQVh7EtdcP6Wz7zhv1WLOiMf7fr5l2AsorQl363bKhGa+9sveo8zU1tOOZ2h2Y8q0h6FvqPOcCrPqgAe+9ub/z2D2fHcaiZz7FtO+chGDHp+ucGHqL//mnd2L/vrbOPtIdn/g9EXRClzUrGvDOG/tx2RWD8PICZ0f+879stMtYbMEGxcXFaG1tFexFveYct3qeiURMv0XUU6+trn6Xl5dj+vTpD9XX18cv6oFAYFFtbe0i9RzMLuLMP1yb3Xny3mrKlCnHBoPB9bFY7MKnn356jWmaIwzDeKNjBfAXtm1vtSzr5zU1NRfYtj0vk4dAvNoCvnTSkW8BO1DWZQv4ygqUlh+BObvlINCnL3Zsa8O7bxw86ssj+VgBbG+zESrqmkqZPgXsjC2br6jkOrG4QpBrxfN7PvqdX/1zfXb6nWvF83c+rgBqtALopJlpmjUAfgogahhGwLbt/7Qsq9aBw1Ao9EcAw23bPhwIBGbU1tYuSZeaiaeAvQLAQMdq3GWT+vUIgNFbr4Zxy4+wc/DXcguAixrQsP/op4AnVVdift0BjL+qH0rLnLfrHPkjAKbLHrV+58Soll+i0dJvUQXVaq+j3wRAzQDQ65L0GgAPHbTjITorYT2tAMYBcPL12PnFqakBsDmKxc91+xScFw+B9ACAV0ytxIKnD+DiiRXo1//IKmVvALhxbQtWpfgWMFcAvc5Ob/vTcYJwFOS4vc0j2Xuj37I75F18BEACoFA25QsA113+MzQddwZ2bW/DF84qwWmjOp9fwUHJAXDD8+9hddNwXDiuDMuXfP4aGAKgUCr63pgTo+8SS3UC+i2VHb4Ho6PfBEACoFBh+QqA7x/CprWH4/GNT7oH0FkBnD/hyS5xJ987lwoAN65pwaoUr4Fxtm8TD4Gk+7LHqx6tAG74t4ew+rRvEQCFMi/3jXWcILgCyNff5L7S8nNGHeubAEgAFKo2XQFw/752LH2xKa5dplvABEChlMtbYx0nCAIgATBvBZfjE+tY3wRAAqBQmekKgMkPeBAAhVJImcY6ThAEQAKgMgUqGKiO9U0AJAAKlQ0BkCuAQgmkUGMdJwgCIAFQoRIVClXH+iYAEgCFikYvAGxEw/5o/Anl3lYAnd+cexYXP3/kJdaJv8R9itwCFkq5vDXWcYIgABIA81ZwOT6xjvVNACQACpUZAfDoFUAHAHt7wpcAKJRyeWus4wRBACQA5q3gcnxiHeubAEgAFCqzBADOfnQtGhui8b4y+ZrFgfp2xKLAssVN6FtqoMt7AJOeAj77vL4YNvLIN3/z9xSw+xXAVAB4xZTK+GtrWv74MJ8CFsq6/DTWcYIgABIA81NtuT+rjvVNACQAClWa6Apg8lZqbwCYDJZeAOD+ve1Y+lITMnsNjBgAnvnFkviraM5Y99QRALy4DMtf5XsAhRIwh411nCAIgATAHJZYXk+lY30TAAmAQkUnMwBOmNQXfSuOrBp2fw8gATB723W8UBKECELZV4xaLVnfavklEi0BkAAokj+QEwBjWPxcAy7b8GuU/fQ/CIBCDh/dmBOEx4JK3h39ltwgj8Oj3x4LKnF3BEACoFB6Sg2AS76PsoeOfDGEK4BCNndpzAnCOy1V6Il+q+CSdzHSb++0lL0nAiABUChHVQXA+r3tWJanewATgvMeQKHUy3ljTow5lzyvJ6TfeZU/5yfX0W8CIAFQqNAIgO5fA5N4CIQAKJRyeWus4wThiM1x5y3l8nJi+p0X2fNyUgIgAVAo8QiAnwPguG9UoLIqGH9JdKrXwAwbUYwtG1o79eYKoFDq5bwxJ8acS57XE9LvvMqf85Pr6DcBkAAoVGh+AuCq9w9h49rDnfEl3i+Y/jUwHQ+BJN8DuLYFq95v6XxHoR9bwOkAsLvQBECh1Mt5Yx0nCK4A8unnnBdank6oY30TAAmAQuVGAAQmTa3E/KcPgAAolErSN9ZxgiAAEgClL0yPAtSxvgmABECh8lEVAL18DyABUCiFlGms4wRBACQAKlOggoHqWN8EQAKgUNkQALkCKJRACjXWcYIgABIAFSpRoVB1rG8CIAFQqGiUB8CplQiGjLgGic/S9fQt4yUvNOJAfTR+H2HyJ+y4AiiUQso01nGCIAASAJUpUMFAdaxvAiABUKhsCIBcARRKIIUa6zhBEAAJgAqVqFCoOtY3AZAAKFQ0UgPg0h+g7Dd/iI9vY7engDvvAeQKYMb+63ihJAgRhDIuFEUbsL4VNS6LsAmABMAs0ubzJr4C4AeHsHGNwGtg8giAXx5XhjeXNPeqLV8DI5R6OW/MiTHnkuf1hPQ7r/Ln/OQ6+k0AJAAKFRoBMHkLuByVVaH4/YFuAPArF5fhjVe7QmJP9x8KmeRxYx0vlFwB5Aqgx2UkbXesb2mt8TwwAiABUCipCIAEQKEEUqgxJ0aFzPIgVPrtgYgKdaGj3wRAAqBQiRIAgUnVlZhf57wImiuAQskkeWMdJwiufHLlU/Ky9Cw8HeubAEgAFCogLwGwpNRAy0E7Ho+zFbrKz3sA97Vj6YtN8a94CL8GhgAolEOqNNZxgiAAEgBVqU/ROHWsbwIgAVCobgiAXAEUSiCFGus4QRAACYAKlahQqDrWNwGQAChUNARAAqBQAinUWMcJggBIAFSoRIVC1bG+CYAEQKGikREAX/9rE/Z82o7Lkl4Ds2ltC1a+3xLfWnb+9ndsAV8xtRIh0S+BcAtYKIdUaazjBEEAJACqUp+icepY3wRAAqBQ3cgIgInPtOUVAMeW4c2lvb8HkK+BEUq9nDfWcYIgABIAc15oeTqhjvVNACQACpWbagB47pf74qThffxfASQACuWVjI11nCAIgARAGWvRj5h0rG8CIAFQqJZkBsBLl/4A5R2fgktsATuDdbaBfd8CJgAK5ZWMjXWcIAiABEAZa9GPmHSsbwIgAVColgiAPTwEQgAUyisZG+s4QRAACYAy1qIfMelY3wRAAqBQLREACYBCCaRQYx0nCAIgAVChEhUKVcf6JgASAIWKxk8AXP3BIWxYc7gzvsQTvNFbr8b8CU92iTv5G7qJh0C4BSxkbY+NdbxQEoQIQv5Uk3y9sr7l88SviAiABECh3FIGANcdxsr3DsXH6vU9gBWVATQeiHX9FBy3gIXySsbGnBhldMW/mOi3f9rK2LOOfhMACYBCtegpAPY10HLo80/BeboC6CMAJgTs8i1gAqBQXsnYWMcJgiufXPmUsRb9iEnH+iYAEgCFaokA+Ll8BEChVJK+sY4TBAGQACh9YXoUoI71TQDUDABN0ywG8F+GYUwEcMi27Q8sy7rRNM1BAJ40DONU27ZbbNueUVdXtzRdbekEgEtfbMT+fdH4FnLiPsNkfcZ+vRz9B4Tiv32ZK4DpUke533WcIAiABEDlCjXLgHWsbwKgZgAYDof/27btgGVZP3DqZMqUKcfOmzfvM9M0HwWw1bKsX1RXV48OBAJ/tm17mGVZ0d7qiQD4uToEwCyvvIo003GCIAASABUpT+EwdaxvAqBGADh58uTSkpKSnQCGWJbVlFwxpmk2RqPRUx0YdP49HA6/Ydv2XZZlvUwAPKIAVwCP6KDjhZLjJggJE4YiHbC+FTHKgzAJgBoBoGmaZwP4C4C5hmFMsG37IICfA3gfwHbLsvomcso0zVoACyzLeiJvAPjhIWxY7dFrYDx4CIQASABsaCAIeTDvKNEFQUgJmzwLUke/CYAaAWB1dfWXAoHAO7Zt32BZ1lOmaZ4L4IVoNHpWKBTaFolESrwEQPuTLUDLQRgjzuyxSJPvpSspNdByMOkpYGEA/CHKf/MEYstfxeayL2HVylg8jmxfA5MLAHTuHezXP4i+pQHPLmxed6TjhZIrgARfr+tI1v5Y37I6431cBECNAPC6664bWFRUtCsSiTgPgsRJyzTNN23bvtMwjP9N3gI2TXM5gDtTbQHX1NRMjMVizkMkqKqqKp41a9aM2Y+uQ2NDezxDv3XryfH/rK+5JP6fVbV/7TFzn3p4W+dvpWVBHGw+csuh08d7b+7Hqg8+n3iS++3+IujEb07bRJ+XLv0hTnhyXjyOT676KT5s+UJn3/v2tGLBn3eh5qYTESo6AluJdsl9JQe+8H93Ye/u1nhsyXEnjrniuuMx4Jji+G+XTByEvy7a3WvFXjbpWLw8P77j3uXvuMF9MOGq47yvdo96LC4uRmtrq0e9qdMNx62OV15ESr+9UFGdPnT1u7y8HNOnT3+ovr4+flEPBAKLamtrF6njnFikhlhztVqbprnQtu376+rqFpimORzA8mg0+sVQKPQftm07D4H8vKam5gLbtueJPgTifLEjnlC/fwY4fAhGSelRYuViBdCJY8vl/4JV7UdWImVeAXTiO+bYEL56abm0icUVAmmt8SUw+u2LrNJ2Sr+ltcbzwLgCqNEKoJM9HdDnPPF7DIBoIBD4eW1t7f86TwOHQqE/Ahhu2/bhQCAwo7a2dkm6jOvtKeAEABrf+i7sp36H4MPO7Ydd/wiAzUdpQgBMl3X5+Z0TY350z9dZ6Xe+lM/PeXX0mwCoGQB6XVquAHD8ZNiLn5UTAKdUIlR0ZBE4AaPJ3xVO1isX9wByBdDrDPWuPx0nCEc9jtu7HFKhJ/qtgkvexEgAJAAKZVI6AFw9choGntwPxy7+fXoA7P4pOI8eAul1C7gDANetasHaFS1xLdIB4AknFWHHx21H6ebFewAJgELp6GtjToy+yitd5/RbOkt8DUhHvwmABEChokoA4J8eXovmpiMPcCQAygEv52GNcvsAxi3+Xp4B8CdY1T6qM779+9qx9MUmXNEBgMlb0ekAsCfBCIBCqSR9Yx0nCK4A8uln6QvTowB1rG8CIAFQqHyyAUB7y3oYw0bGz9vlHkBfVwDVAcCBx4bwNT4EIpSXfjTWcYIgABIAqXtrTwAAIABJREFU/aglGfvUsb4JgARAoVrMFADtvZ8h9i9/j8B9T8EoqyAAvnr0QyAEQKGU9K2xjhMEAZAA6FtBSdaxjvVNACQACpVhxgD46Q7E/u93Ebj3jzAqKgmABECh/MtlYx0nCAIgATCXNZbPc+lY3wRAAqBQzaV7CKT7PYB2DgHwmD0f4qszxuHIQyDdtoC37MHS5SHX9wDOf3o/okfec93jn1f3AHIFUCglfWus4wRBACQA+lZQknWsY30TAAmAQmUoMwA6A3Me6EgJgO+vw9K1x7oGwOR7FXsSjAAolErSN9ZxgiAAEgClL0yPAtSxvgmABECh8nEHgA0Yt/if4k8B53IF0BnYVaN3Ifb/7jh6BbADAMdf1Q/OE8Hv/O1gpw6pngLOFABHfakvVr53qFdtv3JxGd7gFrBQ/uWysY4TBAGQAJjLGsvnuXSsbwIgAVCo5mQHwEkv3Rgf31FbwB0AOGRoEbZv7fpOPy8A0I2oBEA3KslzjI4TBAGQAChPBfobiY71TQAkAApVFQHwc/mSt4DdiEoAdKOSPMfoOEEQAAmA8lSgv5HoWN8EQAKgUFVlDYD/9SSMfv19fQrYGRhXAIXsTdlYxwslQYgg5H0lydkj61tOX/yIigBIABTKK1cAiAaMe6njHsDPdiD20+8iQAAEVwCFUi/njTkx5lzyvJ6QfudV/pyfXEe/CYAEQKFC6xUAb7sG88f/AeUyAuAH67F0zSDwHsDM7dfxQskVQK4AZl4parZgfavpWzZREwAJgNnkTWcbNwBYhgZcLNsKIAEwa985QWQtnZIN6beStmUdNP3OWjrlGhIACYBCSasuAK7D0jXHcgUwC/c5QWQhmsJN6LfC5mUROv3OQjRFmxAACYBCqdvrp+A6toBlWAHcevkdWNl+Vnyszmte9n8gMQAOCuJrl1UI+eJnY04QfqorX9/0Wz5P/IyIfvuprlx9EwAJgEIZ6ScArllxCOtXHe6ML/F+PufLHs4n5pL/kt/dl/zS5sRTwD0B4IlDi/CJbO8BJAAK5aRfjTkx+qWsnP3Sbzl98SsqHf0mABIAheqpdwC8FvPHP4GuK4A7EfvpdAT+6w8w+lV1fQ0MWtCCks5VulwA4JCdy7B98JgeYTLxQ6ZfAnEjao9PARMA3ciX82N0nCAckTnunKdaXk9Iv/Mqf05PTgAkAAolnKcA2LofLcX9cwOAH67D0tXH4oSdr2HH4IsIgBlkASeIDMQqgEPpdwGYmMEQ6HcGYil+KAGQACiUwqkA8KpwJdatbEHF7F/inS/+EGVoxMUvzTjyLeDPelkB1AwAv3hBX3zw1tHfCx7IFUChnPSrMSdGv5SVs1/6LacvfkWlo98EQAKgUD2lAsBJ1ZWYX3egs1+ZATCfW8A9CU8AFEpJ3xrrOEFwC5jvP/StoCTrWMf6JgASAIXK0C0AXvTy/0Hx7yxg964e7wHs03oAh4srtdkCJgAKpV7OG+s4QRAACYA5L7Q8nVDH+iYAEgCFys0tADajAqeN6oPTjt2fJwD8Z6xsP7sTLvdLcA9gT8IPGBTERXwNjFBe+tFYxwmCAEgA9KOWZOxTx/omABIAhWoxEwA8/sQijB7ZQABMozgBUCglfWus4wRBACQA+lZQknWsY30TAAmAQmWYMQCe1ojYXbchcM8fYFR2ew1M94dAlu3A+u2lnfGJvAdwy4njser0v0taAVyPpasHwdN7ACeUo//AUJdX22QjLgEwG9X8b6PjBEEAJAD6X1lynEHH+iYAEgCFqs9PAFw9eyk2BI9s2zp/BEAhqzxrrOOFkiBEEPKsgCTviPUtuUEehkcAJAAKpVMqALyyuhLPd3sK2LkHML4FnMEKYD4B0N63GwiG4quUzp+rF0FzBVAol2RvzIlRdoe8jY9+e6un7L3p6DcBkAAoVJcZAeCQIoz+gvst4HwCoPO5OZSWI3j/bAJgtwzR8ULJFUCuAApdKBVqzPpWyCzBUAmABEChFPIVAOcsw4bAWXnZAo4DIBB/eTVXALumCCcIoZJRrjH9Vs4yoYDpt5B8SjUmABIAhRJWFQDceuJ4rMzgIRACYM9pwQlCqGSUa0y/lbNMKGD6LSSfUo0JgARAoYT1EwDXzFmK9QFvHgIhAArZ3KUxJwjvtFShJ/qtgkvexUi/vdNS9p4IgARAoRxNCYBmJZ63un4KLv4QSPI9gL9+Akb/AV0erijp/hoYPwFwxQYsXXUMhuxYhu0njOmigfO0MVcAuQLYXQFOjEKXCuUa02/lLBMKWEe/CYAEQKGi8RIA+7QdwOGipE/BzVmG9R7dA9h9BfDAig1Y4jEAjplQjiq+B1Aon2RurOME4fjBccucld7HRr+911TWHgmABECh3Ez5HsBrSjD/mZbOfsvQCDcrgEXtB9EWOvLiZ2cVbg0BUMgbvxpzgvBLWTn7pd9y+uJXVPTbL2Xl65cASAAUyspUAHj54puxcPzjGQNgciAEwAohX/xszAnCT3Xl65t+y+eJnxHRbz/VlatvAiABUCgjCYCfy8ctYKFUkr4xJ0bpLfI0QPrtqZzSd6aj3wRAAqBQYboBwMQJ0j0E4usK4JDLsPKMmzq3l3kPYPa263ihdNTiuLPPGRVb0m8VXcs+Zh39JgASALOvGACpAfAmLBz/xFH9SgmAO5dh+2BvngLmCqBQKknfWMcJguDLL6BIX5geBahjfRMACYBC5ZMAwD/+fi0ONkfjfV2+ODUAHjckhAtOb0bszlsRSPEamORAxn69HLsWvZHyKeDmGTfg5XEPdonbuWcw8Zf83d5JL90Y/+etPawAnrDzNewYfNFRfWXzGpiLxgRRFWrEc3898iBLtn8DBgVx0WW8BzBb/fxqp+MEQQAkAPpVT7L1q2N9EwA1BEDTNG82DONR27avtSzrL6ZpDgLwpGEYp9q23WLb9oy6urqlbgo0AYC/++0qGIePtHAHgI8DRgDPLTR6PM3I2EcpAfD52Z8hFiz2BABhx+JxJP9l+x7Ar+2ejf4fLMT8CU+6ka7HYwYcE8RF4wmAQiL60FjHCYIASAD0oZSk7FLH+iYAagaAU6dOHRoMBmc7FWjb9i87APBRZ5HMsqxfVFdXjw4EAn+2bXuYZVlHlvR6+esEwIdWwWjNEABjMTz3YlHGAJi8wpdonO0KYKqTZw2AW2ah/4bXCIDpkkbR33WcIAiABEBFyzXjsHWsbwKgXgBomKb5gm3bdwQCgXtt2/7vDgBsjEajp86bN+8zp2rC4fAbtm3fZVnWy+mqKBMADIaAKy4+3LEF/DjgIQBeOXAJAhOujoebyRZwOgB0fg/c/d947vWB6aTA17b8Hv03LCMAplVKzQN0nCAIgARANas186h1rG8CoEYAaJrmjwGUOSt94XD4FQcAASwDsN2yrL6JkjFNsxbAAsuyjn6So1tdZQKATtOrxvsDgJM2/juCd93jCwAaN/4Tnt/x5bRXFM8AcGAQF03gFnBawXN8gI4TBAGQAJjjMsvb6XSsbwKgJgBYXV09yjCMhwGMdbZ28w2A9i8fx47tNj5YFep5C/jgm1hf+jl4JbZ5U20BOwAYuKoG6D8Qz70+oLPPr79yGz475ktoD5Vg5Rk3x//d6SfxGhg3K4DGhKvxPKrTXpgIgGklUvoAHScIAiABUOmizSB4HeubAKgJAIbD4e/atn03AOdRDefJi+MBHDAMY6Zt279O3gI2TXM5gDt72gKuqamZGIvFJjq1VVVVVTxr1qwZv3NxD6Bz/LRritDwvetRf9dcvP5mx02DPRTpyI3zsP7UKZ2/fuvWk+P//6ce3nZUi8mb/xPRjWvi/578EMbIjU9j/alTMWr1450A6PSz593VWPROWcozO7/X11zS5Tc3D3aM2fYI+q1bIrwFPOi4YnzjasceOf+Ki4vR2tq7d3JGLhYVxy2mn2qt6bdqjonFq6vf5eXlmD59+kP19fXxi3ogEFhUW1u7SExNdVr3/BiqOmPIOFJnBdAwjHtra2ufDYfDj9m27TwE8vOampoLbNue58dDIE6QV01oRexf/h47v/8I3lvV9Une7oNIwFvi33tdAVz/C2DrBlcAmOkKYHeo7ElsrgBmnIZKNdBxhcAxiONWKk2Fg6XfwhIq0wFXADVZAeyekaZpOg943Oc8BDJlypRjQ6HQHwEMt237cCAQmFFbW7vETRZnfA9gBwDuOO4reP/sf+z1FPoCYAAXTejnRv68HMMJIi+y5+2k9Dtv0uflxPQ7L7Ln5aQEQE0B0KtsyxgAOx4CIQD27MCAgQRAr/LTy344MXqppvx90W/5PfIyQh39JgASAIVqSBoAXPdzYNtG11vAO9/fgLfXHpNy7MnvAUwc4OYeQK+2gKsGBjCGK4BCeelHYx0nCG4B8yEQP2pJxj51rG8CIAFQqBblAcCZwLZNrgHw3Tc2YPtWbwHwq1seRtWGpcIPgRAAhVLSt8Y6ThAEQAKgbwUlWcc61jcBkAAoVIb+AmDXp4CdB0iMgcd2edFzIvhJy78HNB44CgBP21iHdadW46zVj+OjpNfAvLt8A7Zv8RYAz9z1PIZ9VEsAFMooeRvrOEEQAAmA8lakt5HpWN8EQAKgUBXlEgAnvXQjgg//JTUAvnRj5ziSt2t7BEAfVgCdAJwY3WwX9yY6VwCFUtK3xjpOEARAAqBvBSVZxzrWNwGQAChUhhkDYEZPAXddASQAClnlWWMdL5QEIYKQZwUkeUesb8kN8jA8AiABUCidVARAByTfP/5C7DhrRsqxX1m6EPZfZnf5ze2qHlcAhdJJ6sacGKW2x/Pg6LfnkkrdoY5+EwAJgEJF6S8AHvmKR+LPqxVAp58Pjr8Q23sAwEnv3Q7s/Sx/ADgggDFf53sAhRLTh8Y6ThBc+eTKpw+lJGWXOtY3AZAAKFSMmQLglcM+gP3If8HdewAz3wJuKD8Jy77yH51jOm2DhXUjzC6fgjt3xW+x27B7BsCk+wkTHeV0BZAAKJSTfjXWcYIgABIA/aon2frVsb4JgARAoTrsBMDffASjLRDv6/LFN2Hh+CdS9uusvjl/fgFgd1BLAOCpm/+CjcOv7oxpyEcPEQCzdF7HCyVBiCCUZbko14z1rZxlWQdMACQAZp08TkMCYFf5PLkHkCuAQjnpV2NOjH4pK2e/9FtOX/yKSke/CYAEQKF6IgASAIUSSKHGOk4QXPnkyqdCJSoUqo71TQAkAAoVjb8AmPlDIK63gFf8BtvP/qdet6mTf+Q9gJ+roeOFkiBEEBK6UCrUmPWtkFmCoRIACYBCKZQaAG/GwvGP9wpX2d4DaEz5OzzfMP6ovhP3FvYMgM9g4/BrOtsNIQBm7TsniKylU7Ih/VbStqyDpt9ZS6dcQwIgAVAoaXMNgE6wqVbj0gNgt4dACIBZ+84JImvplGxIv5W0Leug6XfW0inXkABIABRKWlUAcMD2V7BvyKVcARRy+0hjThAeiKhQF/RbIbM8CJV+eyCiIl0QAAmAQqmqCgC+F2vClwLlBEAhtwmADQ28F86DFFKiC4KQEjZ5FqSOfhMACYBCBZQ9AF6I989O/Sm2REAjNx79EEi2W8AEQCGbuzTW8ULJlU+Cr3cVJHdPrG+5/fEyOgIgAVAonxIA+NhDa9DWGov39cUV/4MPzv6HlP0m7tX7ZPAYfDjqtl7PPXLj0V8CIQAK2eVJY04QnsioTCf0WxmrPAmUfnsioxKdEAAJgEKJmgoAT9j5GnYMvqhXANxy4nisOv3vCIApFKjii6CFctKvxpwY/VJWzn7pt5y++BWVjn4TAAmAQvWUCgCP3/k6dg3+qrIA+LXl/xqPvX/j5s4x5PI9gP2rDIz9RqWQL3421vFC6ejJcfuZVfL1Tb/l88TPiHT0mwBIABSqqVQAGKpfhfaqM6UCwHdjTTivy0MgD2L72d/rdeyJ7WrnIALg51LpeKEkAPIeQKELpUKNWd8KmSUYKgGQACiUQtkC4NYTx2Nl2i1g7x4C6Q6AJ654EJ9ICoBVVcCYb/QX8sXPxpwg/FRXvr7pt3ye+BkR/fZTXbn6JgASAIUyMiUA7luJ9gGjel0BzB0ARrBuRBgqAWD/Axsw9rbRQr742ZgThJ/qytc3/ZbPEz8jot9+qitX3wRAAqBQRvoLgF48BawgAO5fj7HTLxDyxc/GnCD8VFe+vum3fJ74GRH99lNdufomABIAhTKSANhVPue+Qbf3C/YkfH8CoFBO+tWYE6NfysrZL/2W0xe/otLRbwIgAVConrIGwCGXYeUZN/V6bm9eBN3TCuAD+OTs7/d6/rw9BEIAFMpJvxrrOEE4WnLcfmWUnP3Sbzl98SMqAiABUCivCIB+rABuwNjpvAdQKDF9aMyJ0QdRJe6Sfktsjg+h6eg3AZAAKFRKqQHwI7QPOCtlv4lVta05WgH8wvoI1o5M9RCIzCuABEChpPSpsY4TBFcA+fobn8pJum51rG8CIAFQqBD9BUDvHgJ5J9qI84MVnWM9cQUBMFvjdbxQEoQIQtnWi2rtWN+qOZZ9vARAAmD22QMgFQAG932EqCcrgP4BYMXud9A46HxJ7wHkCqBQUvrUmBOjT8JK2i39ltQYn8LS0W8CIAFQqJz8BUAvXgR95CGQ7iuAiLUDgRABMAv3dbxQcgWQK4BZlIqSTVjfStqWVdAEQAJgVomTaJQaAFcgOuBsKe4BPG3DEQB8O9qI0UlbwG4A8EsfPoAB9WvRp63R9atdvHkNDFcAhZLSp8acGH0SVtJu6bekxvgUlo5+EwAJgELllD0AXoqVZ9zc67m9eA1M4iGQngBwdewgzgiU9hjHidv/inNWP0YATFJIxwslVwC5Aih0oVSoMetbIbMEQyUAEgCFUshfABS/BzAdAK6INePsQFmPGpy0/a84O9cAyE/BCeWkX405MfqlrJz90m85ffErKh39JgASAIXqKSUA7l2B6MDet4C3DbkUH6VdAcw/AOZlBZAAKJSTfjXWcYLgyidXPv2qJ9n61bG+CYAEQKE6zB4AL8FHZ3wn71vA6VYAMwXAK166EQsmPCmkaX8CoJB+fjXWcYIgABIA/aon2frVsb4JgARAoTpUFQCNWBvsQBEIgJnbr+OFkiBEEMq8UtRswfpW07dsoiYAEgCzyZvONtneA7htSK5WAGuxdmTNUU8BuwfAV3HO6kddPwTCFUChdJK6MSdGqe3xPDj67bmkUneoo98EQAKgUFGmXgH8ENGB56TsN/EpuG0nXIKPzky3BezFPYAEQCGDUzTW8ULJFUCuAHpdR7L2x/qW1Rnv4yIAagSAV1xxRZ/y8vK5AM4AcMgwjM8CgcA/zp07d6NpmoMAPGkYxqm2bbfYtj2jrq5uabqUSwWAgb0fIiY7AEZbYQeLXWwBZ7oC+HdYMOEP6WTr9ffKAxsx7rbev1IidALBxpwgBAVUrDn9VswwwXDpt6CACjUnAOoHgJdalrXQydFwODwDQHUkErk0HA4/Ztv2FsuyflFdXT06EAj82bbtYZZlRXvLZ38BUPxLIF9Y38MKoFsA3LEE56x6JIMtYAKgQte/jELlxJiRXMofTL+VtzCjAejoNwFQIwDsXg2maTrLTJZlWaeYptkYjUZPnTdv3mcdcPiGbdt3WZb1ciEAYPcxGATAjC6OyQfreKHkFjC3gLMuGMUasr4VM0wgXAKg3gDovK9kD4B/B7Ddsqy+iVwyTbMWwALLsp5QGwDnYu3IaUcNIQGAH8aacU4vL4IesmMJvsgVwC76cYIQuOIq2JR+K2iaQMj0W0A8xZoSADUFwHA4fBeAK0tLS8c3Nzc730LzEAA/QGzgF1OWQmYPgXixBUwA9PqaxAnCa0Xl7o9+y+2P19HRb68Vlbc/AqCGABgOh2+3bTvc2to6/plnnml00rP7FrBpmssB3JlqC7impmZiLBab6LSrqqoqnjVr1ozHHlqDttZYPNMDe2UCwCP3AB69BXwYdrAPZF0BvOr2S6W9ahQXF6O1tVXa+PwKjOP2S1k5+6XfcvriV1S6+l1eXo7p06c/VF9fH7+oBwKBRbW1tYv80lm2fg3ZAvIzHtM0fwTgmwDGW5Z1IHGujodAtlqW9fOampoLbNuel+1DIMG97yM68NxeVwA/PuFirDjzll6HOnKjFyuAogC4FF9c9bDrh0AuX3wzFo5/XMhCPgUsJJ9vjbky4pu0UnZMv6W0xbegdPSbK4AarQCapjnEMIyPbdveaBhGo23bDvy2WJb11SlTphwbCoX+CGC4bduHA4HAjNra2iXpqi3VU8BGy17YJQM9AEDv3gPY0wrghtghjAh03vp4VMxDdmQGgEM/fhFbT/p6Otl6/Z0AKCSfb411nCAcMTlu31JKyo7pt5S2+BIUAVAjAPQjg1IBIKKHgWAfSQCwp3sAj2wBb421YGigpEdpMgXAqv3rUN//NCGpCYBC8vnWmBOjb9JK2TH9ltIW34LS0W8CIAFQqKBkB8Dypo/RVH7SUWM0okcAcEusBcM8BMD++9djf/+RQpoSAIXk862xjhMEVwD5+hvfCkqyjnWsbwIgAVCoDFMDYAsQTL2qlngK+OMTxmHFmX/f67lHbhTfAu7pBATA7G3X8UJJECIIZV8xarVkfavll0i0BEACoEj+ICUAtrcAocIAwIH7VuLCd3/p+iGQ/vs3YH//EUKaVjRsxiW3fkmoDz8bc4LwU135+qbf8nniZ0T020915eqbAEgAFMrIQgdARxxn1XL+BOed2en/+h/YgP2VYgB4qHU/wjcMS3+yPB3BCSJPwufptPQ7T8Ln6bT0O0/C5+G0BEACoFDapQbAQ0Ao9ZO1sm0Bb461YHgv9wBmCoDO/XsHKk8V0rSprRHXf/vo+xaFOvWwMScID8VUoCv6rYBJHoZIvz0UU/KuCIAEQKEU9RMAqz5ZjPoTx3fGl4DHVKtxvf2WaoCBaAtiwRIQADO3nxNE5pqp3IJ+q+xe5rHT78w1U7UFAZAAKJS7qQHwIBByvi539F8mK4Cx3e8gMOh8AqCQQ9435gThvaYy90i/ZXbH+9jot/eaytojAZAAKJSbWQPg4LFYMerWXs8d2/02AoNGuwLAge//El84sBV/u/i3rsYTaG9BLOTHCuAmHKg8xVUMPR3ELWAh+XxrzInRN2ml7Jh+S2mLb0Hp6DcBkAAoVFCyAOAj7bsw7tN3cdqQSa7GE2g/hFior+dbwP0aNqOh33BXMRAAhWTKeWMdJwhHZI4756mW1xPS77zKn9OTEwAJgEIJly0AfjJ4DD4cdZtnK4DZAuCmWAtOSfMQyBUv3YgFLp8CJgAKpZPUjTkxSm2P58HRb88llbpDHf0mABIAhYoyJQC2HQSKer8H0A0A2rvfhuFyCzh7ADyEU3r5FrAjTvFHD6L1rO+50skLAHRONLmmv6vz5eMgHS+UXAnji6DzUWv5OCfrOx+q5+ecBEACoFDm+bkCaO9+C8agCzrj6+1JXwcAx376Hr4w5ApX40lsAW+KpQfAPZvm4JhTrnfVb7+GLWjoJ/4OPwKgK7lzehAnxpzKnfeT0e+8W5DTAHT0mwBIABQqskwB8PR1s3HKtoVwswK478BaDKj8gq8AuDF2CKemWQHcu2kOBroFwMataKgYKqQpVwCF5fOlAx0nCK58cuXTl2KSsFMd65sASAAUKsXUW8DNQFFZj/06K3mfDL4IH46a3uu534k24vxghU8AeBCxUCkIgJnbr+OFkiBEEMq8UtRswfpW07dsoiYAEgCzyZvONtkC4Pbjv4YPzvquEgC4b+NsDDj1m6506scVQFc6qXgQJ0YVXcs+ZvqdvXYqttTRbwIgAVCoVlMDYBNQVN7rCmChAmBF41Y0cgtYKKdkbazjBMGVT658ylqPXselY30TAAmAQnWkLAC2HUSsyN0WcCYrgBWN29BYcbKQpk5jPgQiLKHnHeg4QRAACYCeF5KkHepY3wRAAqBQOWYPgF/FB2f9g6dbwGM+fR+nD7nc1XgCbc2IFZW5ugeQANhVUh0vlAQhgpCrC0sBHMT6LgATXQ6BAEgAdJkqqQ/LGgCP+yo+OJsA2JP4XAEUSktfGnNi9EVWaTul39Ja40tgOvpNACQAChVTagBsBIo+f3q3+wmcp4C3ewyAL0brMXT3ezjt+MtcjSfY1oxoURk2xA5hRJrXwDTueAkVJ0xw1W9508doKj/J1bG9HUQAFJbQ8w50nCC48smVT88LSdIOdaxvAiABUKgc/QTAd6MNOC/YrzO+3l4E7Rx04OAOVJae4Go8mQCgqw47Dipv+gRN5Sdm0iTlsQRAYQk970DHCYIASAD0vJAk7VDH+iYAEgCFyjFbANxx3Ffw/tn/2Ou5uwPgxMU3od0IYvFlj6ZslxkANiFaVO5qBTATgQiAmail1rE6ThAEQAKgWlWafbQ61jcBkACYfcUASAWAB2OtKA0U99ivs5L31vCrsPvUcEYAuPnjv2Dx4C/j70PHewaA62OHMDLNFnAmAhEAM1FLrWN1nCAIgARAtao0+2h1rG8CIAEw+4rpAQAb7SgqjGCvAPjyyDBahl6VEQC+sP89bCsf3AsA7kRl6WBX4wm2HVkB9B4At6OpfIirGHo7iFvAwhJ63oGOEwQBkADoeSFJ2qGO9U0AJAAKlWOqFUA3ALjowp8jWjE8IwB8cf972OoxAAoNPkXj8iYCoNeaytKfjhMEAZAAKEv9+R2HjvVNACQACtVVtgA4f8KTac/b/R7AF/e/i63lJ3i6Apg2iAwPKGvegeYydw+icAUwQ3HzfLiOEwQBkACY57LL2el1rG8CIAFQqMBSA2A7KoxQr1vA2QDgS/vfxRavALC1EdHinl9Vk60oBMBslZO/nY4TBAGQACh/ZXoToY71TQAkAApVTyoAbLDb0c8HAFxc/w42VwzpeQXw0E5U9nV5DyABMGvfdbxQEoQIQlkXjGINWd+KGSYQLgGQACj6lMpfAAAONUlEQVSQPqmfAtYZAEubd+JgmTsI5RawUOrlvDEnxpxLntcT0u+8yp/zk+voNwGQAChUaLlcAXy5/m1sqjixlxXAXajsm/oVMd0HGfRpBTDUsgftJccIaeo05lPAwhJ63oGOEwRXPrny6XkhSdqhjvVNACQACpVjKgA8YLej0octYG8BsAHR4s+/MiIkQlLjaGsDgh70SwD0yhHv+tFxgiAAEgC9qyC5e9KxvgmABEChqlQVAI22JthF5UJjT9U42taIYC/fQXZ7QgKgW6Vyd5yOEwQBkACYuwrL75l0rG8CIAFQqOr8BMB3og04P+lbwJGDG9BQXO7JFvBBO4rSXl5WLSSKB40JgB6I6HEXOk4QBEACoMdlJG13OtY3AZAAKFSQ2QDgFS/diAUu3gP4TrQR5wc/f1WLdXA9DhRXeAKAzXYUZQTArLzX8UJJECIIZVUsCjZifStoWpYhEwAJgFmmzpFm2bwI+pyVv8eHo25Le97uAPh08zrU9+nXCwB+isq+x6Xt1zmAK4CuZEp5ECeI7LVTsSX9VtG17GOm39lrp1pLAiABUChnUwHgbrsNg4yiHvs9Zctz2DSs9+8AO427A+Cfm9Zgb0l/TwCQK4DZ284JInvtVGxJv1V0LfuY6Xf22qnWkgBIABTK2WwAcOD2V7F3yMVpz9sdAP+3aTX2lFQRANMq5+8BnCD81Ve23um3bI74Gw/99ldfmXonABIAhfIxFQB+ZrfiWKO4x35b9q9BSf/T0563OwD+pXEVPus7gACYVjl/D+AE4a++svVOv2VzxN946Le/+srUOwGQACiUj9kA4KED69C38rS05021AhgrGYQpodQvWm5sa0KFy1e7cAs4rfw9HsAJInvtVGxJv1V0LfuY6Xf22qnWkgBIAOzMWdM0RxiG8Qfbth3C2h8MBm+aO3fu6t6SOhsAPNiwAaX9RqStle4A+Ej7rh5X/9J21u2AJjuKcj4FnKls8eM5QWQlm7KN6Ley1mUVOP3OSjYlGxEACYDJALgYwBOWZf2xpqZmaiwW+4llWV/2HgA3orTfqWkLhgCYVqK8HMAJIi+y5+2k9Dtv0uflxPQ7L7Ln5aQEQAJgPPFM0xwEYP2oUaMGzJw5M9bxbztjsdhFTz/99KaesjPVCmBeMrnATsoXQctnKCdG+TzxMyL67ae68vWto98EQAJgvBJramrOs237qUgkckaiNE3TXG7b9k/q6ur+SgDM7QWLAJhbvd2cTccJwtGF43aTHYVzDP0uHC/TjYQASAD0BAD/+Pt1aGuLLxzyzwMFDsY+86AXf7oIBAzEYrY/nUvcK8ctsTk+hEa/fRBV4i519LuoKIjp373IcaXSMAz9PvcDAmC8JDPZAq6pqZkYi8UmOu1OPvnksnvuuSf9Zz0kLnyGRgWoABWgAlRAVwVuv/3232/btq3ZGX8gEFhUW1u7SBctDF0Gmm6cpmm+7DwFHIlE/mCaZjWAO1w8BGLcfvvtv7vnnnv+OV3/hfb79OnT/3PWrFl3Fdq40o2H406nUGH9Tr8Ly890o6Hf6RQquN8rAOwwDEO/bR2uAH6ezNOmTTstGo0+AWAggAO2bd9cV1e3Ml26m6Z5r2VZP0p3XKH9znEXmqO9j4d+028dFGCe6+Ayx5hQgCuAgrnAC4aggIo1p9+KGSYYLv0WFFCx5vRbMcMYrpACBEAh+eJPEE/U6Z6BhFwct2DiKNacfitmmGC49FtQQMWa6+q3YjZ5Hi4B0HNJ2SEVoAJUgApQASpABeRWgAAotz+MjgpQASpABagAFaACnitAAPRcUnZIBagAFaACVIAKUAG5FSAAyu0Po6MCVIAKUAEqQAWogOcKEABdSGqa5gjnHYG2bR8DYH8wGLxp7ty5q7s3ra6uvioQCPzatu0AgBUAbrIsq8nFKaQ8xM24p06dOjQQCGw0DOND27YN531Ktm1PtSxrs5SDShOUaZr3A7jaMIyhsVjs3Lq6ug9TNSlAr9OOu9C8dny94oor+pSXl88F4HwG8pBhGJ8FAoF/nDt37sZCrm+34y5Ez8Ph8CLbto8D4Lz7zfkCxA8sy3q/kP12xuZm3IXod8JX0zRvNgzjUdu2r7Us6y+F7reb+ZcA6EIl0zQXA3jCsqw/1tTUTI3FYj/p/pLoG264oaylpWVjLBYb+/TTT683TfNBZ0KxLOsOF6eQ8hA34+64YLxnWdYAKQeRYVCmaY4JhUKb2tvblzkXilQAWKBepx13oXmdBICXWpa1sGOSnAGgOhKJXJqcOoXmeQcAph13IXr+7W9/u9+f/vSn+Ke/ampqro3FYjMtyzq3kP12xuZm3IXotzN2Z1zBYHC28/+3bfuX3QGw0Orb7bRHAEy/IjQIwPpRo0YNmDlzZvyDv6Zp7ozFYhc9/fTTm5L+10W1YRjfiUQik5x/mzZt2hmxWOyFSCRyklszZDrO7efxOgrr/UgkUiVT/KKxmKa52bbta1IBoPOlmELyOlmr3sZdqF53G//5ACzLsk7p9u8F63nHNS3luAvdc9M0bzIM4/uRSOQ8zfxOOe4C9dswTfMF27bvCAQC99q2/d/dAbCQr+m9zYUEwDSkUFNTc55t209FIhFniyj+Z5rmctu2f1JXV/fXpH9zvgYy0rKsf+g4pq/zgekzzzyzTwIcRaEkl+3djrvjfzGuS9oCfiYSifx7x/ZKLkP29FxpALCgvM4EAAOBQMF53W38TwLY0/3rPqZpFqznHderlOMu1PoOh8POLT3OKq9zy8qk7l99KlS/0427EP02TfPHAMosy/pFOBx+pQcALOj67mlyJAASAFMq4BYAb7vttqKmpqbK2bNn77n22mv7FxUVRQzDcFY+7/GUyHLcGQHw6HsfC9XrRGqFw2Hn29ZXlpaWjn/iiSdauoFhwU4QvY270D03TfMGZ8PGsqwrdfG7A/hTjrvQ/K6urh5lGMbDAMZalhUlAHadSAmAacDC7Vaos4QM4BbLsq7oKLAzASy0LOvkHLOLJ6dzO+7uJzNNcxqA6y3LusaTQPLUSbot4ELy2u0KYKF67YwrHA7fbtt2uLW1dfwzzzzTmGKsBVXfSdDb67gL2fOk3ZuDRUVFQ2bPnl2f9G8F6Xe3Wj9q3IXmdzgc/q5t23cDOAzg/7d3P6+NFnEcxz/zJOCPtUqRVcFf+IOCPXnwokgvHgVR2glhpSzon6CoN6NXL+pB1uBhVSjtMxIKnjxYetiDHvaiEkTEi0jBuq7Si23z9JFhUwgxdmehs51M37kVHp6Z7+vbJJ88yTzj8859kv6W9LZz7uPT1O9Jb6UEwICAYa3d8KuAy7L8dBj03hhfBGKtvUPSz41GY2F1dfWnTBaBhNR9dnt7++rm5ubA/7B8Zmbmc0n9siw7AbTJHnKdAJhdr0deCI/67WOuvfZX985Jes45598c/vPI9PkdUndWPbfW3lVV1e29Xm/LN3m4COTD8Q/qufX7BurOqt/jT+QjrgBm+5p+1JssATAggrTb7bmqqi5Kunv46cHf3qVvrX1H0m/Oua4/zeGtQSQ16rr+YW9v7/ykqwkBQyZxSEjd1tqXJL0raSCpKWljdnb29W63u59EETc4iVardcF/DTi8TcQVY8xOWZZzufc6pO7ceu3/Nay19xtjfq3r2t/KaMffykjSP865p3PueWjdufXcWuu/kXHGmFv9j/8k/S7pNefc95n3O6ju3Po94YrmhqT3/SKQnPsd+rZHAAyV4jgEEEAAAQQQQCATAQJgJo2kDAQQQAABBBBAIFSAABgqxXEIIIAAAggggEAmAgTATBpJGQgggAACCCCAQKgAATBUiuMQQAABBBBAAIFMBAiAmTSSMhBAAAEEEMhZwFr7gaQXjDEPHxwcPDlpq87R+q21fo/6r40xfsW33/bljKRH9vf371lfX/8rZ6uQ2giAIUocgwACCCCAAAInKmCtfbbZbP4yGAwu1XX94vUC4Phkh9vCLUz7RgXH1QQC4HFJch4EEEAAAQQQiC4wfqN+a+3j/v5+ks5KukVS1zn30YQA2C+K4s21tbUvo09yCgYgAE5Bk5giAggggAACCFwTGA2AnU6n6Pf73xZF8fJwF67bjDHf1HX9inPu8qHZ0tLSM0VRfDE/P/9Ap9M5wPLa3ng8EEAAAQQQQACBqRAYDYDtdvuJqqouG2N+HO7kI2PMnXVdd5xzfmvSw9D4iaQ/nHNvTUWRN2GSBMCbgMwQCCCAAAIIIHA8AqMB0Fo7b4z5qizLB//v7MvLy2d2d3e3iqJ4yl8lPJ5ZTP9ZCIDT30MqQAABBBBA4NQIjAXAhqTvJL3nnLvoEdrt9mNFUfy5srJy1f9trX3VGHO+LMuFU4MUUCgBMACJQxBAAAEEEEDgZAVardYFSc/XdX2vpCvGmJ2yLOcWFxcfLYrCLwJ5yBjTlLQ9GAzO9Xq9rWEAvDRcGPLZyVaQ1ugEwLT6wWwQQAABBBBAAIHoAgTA6MQMgAACCCCAAAIIpCVAAEyrH8wGAQQQQAABBBCILkAAjE7MAAgggAACCCCAQFoCBMC0+sFsEEAAAQQQQACB6AIEwOjEDIAAAggggAACCKQlQABMqx/MBgEEEEAAAQQQiC5AAIxOzAAIIIAAAggggEBaAgTAtPrBbBBAAAEEEEAAgegCBMDoxAyAAAIIIIAAAgikJUAATKsfzAYBBBBAAAEEEIguQACMTswACCCAAAIIIIBAWgIEwLT6wWwQQAABBBBAAIHoAgTA6MQMgAACCCCAAAIIpCVAAEyrH8wGAQQQQAABBBCILkAAjE7MAAgggAACCCCAQFoCBMC0+sFsEEAAAQQQQACB6AIEwOjEDIAAAggggAACCKQl8C/Oy8EXITZgGAAAAABJRU5ErkJggg==\">"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "m.plotclass(CGs1,CHGs1,CHHs1,1,0,max_size,window)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Ploting the number of methylated cytosines for a specified window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "/* Put everything inside the global mpl namespace */\n",
       "window.mpl = {};\n",
       "\n",
       "mpl.get_websocket_type = function() {\n",
       "    if (typeof(WebSocket) !== 'undefined') {\n",
       "        return WebSocket;\n",
       "    } else if (typeof(MozWebSocket) !== 'undefined') {\n",
       "        return MozWebSocket;\n",
       "    } else {\n",
       "        alert('Your browser does not have WebSocket support.' +\n",
       "              'Please try Chrome, Safari or Firefox ≥ 6. ' +\n",
       "              'Firefox 4 and 5 are also supported but you ' +\n",
       "              'have to enable WebSockets in about:config.');\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure = function(figure_id, websocket, ondownload, parent_element) {\n",
       "    this.id = figure_id;\n",
       "\n",
       "    this.ws = websocket;\n",
       "\n",
       "    this.supports_binary = (this.ws.binaryType != undefined);\n",
       "\n",
       "    if (!this.supports_binary) {\n",
       "        var warnings = document.getElementById(\"mpl-warnings\");\n",
       "        if (warnings) {\n",
       "            warnings.style.display = 'block';\n",
       "            warnings.textContent = (\n",
       "                \"This browser does not support binary websocket messages. \" +\n",
       "                    \"Performance may be slow.\");\n",
       "        }\n",
       "    }\n",
       "\n",
       "    this.imageObj = new Image();\n",
       "\n",
       "    this.context = undefined;\n",
       "    this.message = undefined;\n",
       "    this.canvas = undefined;\n",
       "    this.rubberband_canvas = undefined;\n",
       "    this.rubberband_context = undefined;\n",
       "    this.format_dropdown = undefined;\n",
       "\n",
       "    this.image_mode = 'full';\n",
       "\n",
       "    this.root = $('<div/>');\n",
       "    this._root_extra_style(this.root)\n",
       "    this.root.attr('style', 'display: inline-block');\n",
       "\n",
       "    $(parent_element).append(this.root);\n",
       "\n",
       "    this._init_header(this);\n",
       "    this._init_canvas(this);\n",
       "    this._init_toolbar(this);\n",
       "\n",
       "    var fig = this;\n",
       "\n",
       "    this.waiting = false;\n",
       "\n",
       "    this.ws.onopen =  function () {\n",
       "            fig.send_message(\"supports_binary\", {value: fig.supports_binary});\n",
       "            fig.send_message(\"send_image_mode\", {});\n",
       "            fig.send_message(\"refresh\", {});\n",
       "        }\n",
       "\n",
       "    this.imageObj.onload = function() {\n",
       "            if (fig.image_mode == 'full') {\n",
       "                // Full images could contain transparency (where diff images\n",
       "                // almost always do), so we need to clear the canvas so that\n",
       "                // there is no ghosting.\n",
       "                fig.context.clearRect(0, 0, fig.canvas.width, fig.canvas.height);\n",
       "            }\n",
       "            fig.context.drawImage(fig.imageObj, 0, 0);\n",
       "        };\n",
       "\n",
       "    this.imageObj.onunload = function() {\n",
       "        this.ws.close();\n",
       "    }\n",
       "\n",
       "    this.ws.onmessage = this._make_on_message_function(this);\n",
       "\n",
       "    this.ondownload = ondownload;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_header = function() {\n",
       "    var titlebar = $(\n",
       "        '<div class=\"ui-dialog-titlebar ui-widget-header ui-corner-all ' +\n",
       "        'ui-helper-clearfix\"/>');\n",
       "    var titletext = $(\n",
       "        '<div class=\"ui-dialog-title\" style=\"width: 100%; ' +\n",
       "        'text-align: center; padding: 3px;\"/>');\n",
       "    titlebar.append(titletext)\n",
       "    this.root.append(titlebar);\n",
       "    this.header = titletext[0];\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_canvas = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var canvas_div = $('<div/>');\n",
       "\n",
       "    canvas_div.attr('style', 'position: relative; clear: both; outline: 0');\n",
       "\n",
       "    function canvas_keyboard_event(event) {\n",
       "        return fig.key_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    canvas_div.keydown('key_press', canvas_keyboard_event);\n",
       "    canvas_div.keyup('key_release', canvas_keyboard_event);\n",
       "    this.canvas_div = canvas_div\n",
       "    this._canvas_extra_style(canvas_div)\n",
       "    this.root.append(canvas_div);\n",
       "\n",
       "    var canvas = $('<canvas/>');\n",
       "    canvas.addClass('mpl-canvas');\n",
       "    canvas.attr('style', \"left: 0; top: 0; z-index: 0; outline: 0\")\n",
       "\n",
       "    this.canvas = canvas[0];\n",
       "    this.context = canvas[0].getContext(\"2d\");\n",
       "\n",
       "    var rubberband = $('<canvas/>');\n",
       "    rubberband.attr('style', \"position: absolute; left: 0; top: 0; z-index: 1;\")\n",
       "\n",
       "    var pass_mouse_events = true;\n",
       "\n",
       "    canvas_div.resizable({\n",
       "        start: function(event, ui) {\n",
       "            pass_mouse_events = false;\n",
       "        },\n",
       "        resize: function(event, ui) {\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "        stop: function(event, ui) {\n",
       "            pass_mouse_events = true;\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "    });\n",
       "\n",
       "    function mouse_event_fn(event) {\n",
       "        if (pass_mouse_events)\n",
       "            return fig.mouse_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    rubberband.mousedown('button_press', mouse_event_fn);\n",
       "    rubberband.mouseup('button_release', mouse_event_fn);\n",
       "    // Throttle sequential mouse events to 1 every 20ms.\n",
       "    rubberband.mousemove('motion_notify', mouse_event_fn);\n",
       "\n",
       "    rubberband.mouseenter('figure_enter', mouse_event_fn);\n",
       "    rubberband.mouseleave('figure_leave', mouse_event_fn);\n",
       "\n",
       "    canvas_div.on(\"wheel\", function (event) {\n",
       "        event = event.originalEvent;\n",
       "        event['data'] = 'scroll'\n",
       "        if (event.deltaY < 0) {\n",
       "            event.step = 1;\n",
       "        } else {\n",
       "            event.step = -1;\n",
       "        }\n",
       "        mouse_event_fn(event);\n",
       "    });\n",
       "\n",
       "    canvas_div.append(canvas);\n",
       "    canvas_div.append(rubberband);\n",
       "\n",
       "    this.rubberband = rubberband;\n",
       "    this.rubberband_canvas = rubberband[0];\n",
       "    this.rubberband_context = rubberband[0].getContext(\"2d\");\n",
       "    this.rubberband_context.strokeStyle = \"#000000\";\n",
       "\n",
       "    this._resize_canvas = function(width, height) {\n",
       "        // Keep the size of the canvas, canvas container, and rubber band\n",
       "        // canvas in synch.\n",
       "        canvas_div.css('width', width)\n",
       "        canvas_div.css('height', height)\n",
       "\n",
       "        canvas.attr('width', width);\n",
       "        canvas.attr('height', height);\n",
       "\n",
       "        rubberband.attr('width', width);\n",
       "        rubberband.attr('height', height);\n",
       "    }\n",
       "\n",
       "    // Set the figure to an initial 600x600px, this will subsequently be updated\n",
       "    // upon first draw.\n",
       "    this._resize_canvas(600, 600);\n",
       "\n",
       "    // Disable right mouse context menu.\n",
       "    $(this.rubberband_canvas).bind(\"contextmenu\",function(e){\n",
       "        return false;\n",
       "    });\n",
       "\n",
       "    function set_focus () {\n",
       "        canvas.focus();\n",
       "        canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    window.setTimeout(set_focus, 100);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>')\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items) {\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) {\n",
       "            // put a spacer in here.\n",
       "            continue;\n",
       "        }\n",
       "        var button = $('<button/>');\n",
       "        button.addClass('ui-button ui-widget ui-state-default ui-corner-all ' +\n",
       "                        'ui-button-icon-only');\n",
       "        button.attr('role', 'button');\n",
       "        button.attr('aria-disabled', 'false');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "\n",
       "        var icon_img = $('<span/>');\n",
       "        icon_img.addClass('ui-button-icon-primary ui-icon');\n",
       "        icon_img.addClass(image);\n",
       "        icon_img.addClass('ui-corner-all');\n",
       "\n",
       "        var tooltip_span = $('<span/>');\n",
       "        tooltip_span.addClass('ui-button-text');\n",
       "        tooltip_span.html(tooltip);\n",
       "\n",
       "        button.append(icon_img);\n",
       "        button.append(tooltip_span);\n",
       "\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    var fmt_picker_span = $('<span/>');\n",
       "\n",
       "    var fmt_picker = $('<select/>');\n",
       "    fmt_picker.addClass('mpl-toolbar-option ui-widget ui-widget-content');\n",
       "    fmt_picker_span.append(fmt_picker);\n",
       "    nav_element.append(fmt_picker_span);\n",
       "    this.format_dropdown = fmt_picker[0];\n",
       "\n",
       "    for (var ind in mpl.extensions) {\n",
       "        var fmt = mpl.extensions[ind];\n",
       "        var option = $(\n",
       "            '<option/>', {selected: fmt === mpl.default_extension}).html(fmt);\n",
       "        fmt_picker.append(option)\n",
       "    }\n",
       "\n",
       "    // Add hover states to the ui-buttons\n",
       "    $( \".ui-button\" ).hover(\n",
       "        function() { $(this).addClass(\"ui-state-hover\");},\n",
       "        function() { $(this).removeClass(\"ui-state-hover\");}\n",
       "    );\n",
       "\n",
       "    var status_bar = $('<span class=\"mpl-message\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.request_resize = function(x_pixels, y_pixels) {\n",
       "    // Request matplotlib to resize the figure. Matplotlib will then trigger a resize in the client,\n",
       "    // which will in turn request a refresh of the image.\n",
       "    this.send_message('resize', {'width': x_pixels, 'height': y_pixels});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_message = function(type, properties) {\n",
       "    properties['type'] = type;\n",
       "    properties['figure_id'] = this.id;\n",
       "    this.ws.send(JSON.stringify(properties));\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_draw_message = function() {\n",
       "    if (!this.waiting) {\n",
       "        this.waiting = true;\n",
       "        this.ws.send(JSON.stringify({type: \"draw\", figure_id: this.id}));\n",
       "    }\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    var format_dropdown = fig.format_dropdown;\n",
       "    var format = format_dropdown.options[format_dropdown.selectedIndex].value;\n",
       "    fig.ondownload(fig, format);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_resize = function(fig, msg) {\n",
       "    var size = msg['size'];\n",
       "    if (size[0] != fig.canvas.width || size[1] != fig.canvas.height) {\n",
       "        fig._resize_canvas(size[0], size[1]);\n",
       "        fig.send_message(\"refresh\", {});\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_rubberband = function(fig, msg) {\n",
       "    var x0 = msg['x0'];\n",
       "    var y0 = fig.canvas.height - msg['y0'];\n",
       "    var x1 = msg['x1'];\n",
       "    var y1 = fig.canvas.height - msg['y1'];\n",
       "    x0 = Math.floor(x0) + 0.5;\n",
       "    y0 = Math.floor(y0) + 0.5;\n",
       "    x1 = Math.floor(x1) + 0.5;\n",
       "    y1 = Math.floor(y1) + 0.5;\n",
       "    var min_x = Math.min(x0, x1);\n",
       "    var min_y = Math.min(y0, y1);\n",
       "    var width = Math.abs(x1 - x0);\n",
       "    var height = Math.abs(y1 - y0);\n",
       "\n",
       "    fig.rubberband_context.clearRect(\n",
       "        0, 0, fig.canvas.width, fig.canvas.height);\n",
       "\n",
       "    fig.rubberband_context.strokeRect(min_x, min_y, width, height);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_figure_label = function(fig, msg) {\n",
       "    // Updates the figure title.\n",
       "    fig.header.textContent = msg['label'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_cursor = function(fig, msg) {\n",
       "    var cursor = msg['cursor'];\n",
       "    switch(cursor)\n",
       "    {\n",
       "    case 0:\n",
       "        cursor = 'pointer';\n",
       "        break;\n",
       "    case 1:\n",
       "        cursor = 'default';\n",
       "        break;\n",
       "    case 2:\n",
       "        cursor = 'crosshair';\n",
       "        break;\n",
       "    case 3:\n",
       "        cursor = 'move';\n",
       "        break;\n",
       "    }\n",
       "    fig.rubberband_canvas.style.cursor = cursor;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_message = function(fig, msg) {\n",
       "    fig.message.textContent = msg['message'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_draw = function(fig, msg) {\n",
       "    // Request the server to send over a new figure.\n",
       "    fig.send_draw_message();\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_image_mode = function(fig, msg) {\n",
       "    fig.image_mode = msg['mode'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Called whenever the canvas gets updated.\n",
       "    this.send_message(\"ack\", {});\n",
       "}\n",
       "\n",
       "// A function to construct a web socket function for onmessage handling.\n",
       "// Called in the figure constructor.\n",
       "mpl.figure.prototype._make_on_message_function = function(fig) {\n",
       "    return function socket_on_message(evt) {\n",
       "        if (evt.data instanceof Blob) {\n",
       "            /* FIXME: We get \"Resource interpreted as Image but\n",
       "             * transferred with MIME type text/plain:\" errors on\n",
       "             * Chrome.  But how to set the MIME type?  It doesn't seem\n",
       "             * to be part of the websocket stream */\n",
       "            evt.data.type = \"image/png\";\n",
       "\n",
       "            /* Free the memory for the previous frames */\n",
       "            if (fig.imageObj.src) {\n",
       "                (window.URL || window.webkitURL).revokeObjectURL(\n",
       "                    fig.imageObj.src);\n",
       "            }\n",
       "\n",
       "            fig.imageObj.src = (window.URL || window.webkitURL).createObjectURL(\n",
       "                evt.data);\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "        else if (typeof evt.data === 'string' && evt.data.slice(0, 21) == \"data:image/png;base64\") {\n",
       "            fig.imageObj.src = evt.data;\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        var msg = JSON.parse(evt.data);\n",
       "        var msg_type = msg['type'];\n",
       "\n",
       "        // Call the  \"handle_{type}\" callback, which takes\n",
       "        // the figure and JSON message as its only arguments.\n",
       "        try {\n",
       "            var callback = fig[\"handle_\" + msg_type];\n",
       "        } catch (e) {\n",
       "            console.log(\"No handler for the '\" + msg_type + \"' message type: \", msg);\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        if (callback) {\n",
       "            try {\n",
       "                // console.log(\"Handling '\" + msg_type + \"' message: \", msg);\n",
       "                callback(fig, msg);\n",
       "            } catch (e) {\n",
       "                console.log(\"Exception inside the 'handler_\" + msg_type + \"' callback:\", e, e.stack, msg);\n",
       "            }\n",
       "        }\n",
       "    };\n",
       "}\n",
       "\n",
       "// from http://stackoverflow.com/questions/1114465/getting-mouse-location-in-canvas\n",
       "mpl.findpos = function(e) {\n",
       "    //this section is from http://www.quirksmode.org/js/events_properties.html\n",
       "    var targ;\n",
       "    if (!e)\n",
       "        e = window.event;\n",
       "    if (e.target)\n",
       "        targ = e.target;\n",
       "    else if (e.srcElement)\n",
       "        targ = e.srcElement;\n",
       "    if (targ.nodeType == 3) // defeat Safari bug\n",
       "        targ = targ.parentNode;\n",
       "\n",
       "    // jQuery normalizes the pageX and pageY\n",
       "    // pageX,Y are the mouse positions relative to the document\n",
       "    // offset() returns the position of the element relative to the document\n",
       "    var x = e.pageX - $(targ).offset().left;\n",
       "    var y = e.pageY - $(targ).offset().top;\n",
       "\n",
       "    return {\"x\": x, \"y\": y};\n",
       "};\n",
       "\n",
       "/*\n",
       " * return a copy of an object with only non-object keys\n",
       " * we need this to avoid circular references\n",
       " * http://stackoverflow.com/a/24161582/3208463\n",
       " */\n",
       "function simpleKeys (original) {\n",
       "  return Object.keys(original).reduce(function (obj, key) {\n",
       "    if (typeof original[key] !== 'object')\n",
       "        obj[key] = original[key]\n",
       "    return obj;\n",
       "  }, {});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.mouse_event = function(event, name) {\n",
       "    var canvas_pos = mpl.findpos(event)\n",
       "\n",
       "    if (name === 'button_press')\n",
       "    {\n",
       "        this.canvas.focus();\n",
       "        this.canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    var x = canvas_pos.x;\n",
       "    var y = canvas_pos.y;\n",
       "\n",
       "    this.send_message(name, {x: x, y: y, button: event.button,\n",
       "                             step: event.step,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "\n",
       "    /* This prevents the web browser from automatically changing to\n",
       "     * the text insertion cursor when the button is pressed.  We want\n",
       "     * to control all of the cursor setting manually through the\n",
       "     * 'cursor' event from matplotlib */\n",
       "    event.preventDefault();\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    // Handle any extra behaviour associated with a key event\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.key_event = function(event, name) {\n",
       "\n",
       "    // Prevent repeat events\n",
       "    if (name == 'key_press')\n",
       "    {\n",
       "        if (event.which === this._key)\n",
       "            return;\n",
       "        else\n",
       "            this._key = event.which;\n",
       "    }\n",
       "    if (name == 'key_release')\n",
       "        this._key = null;\n",
       "\n",
       "    var value = '';\n",
       "    if (event.ctrlKey && event.which != 17)\n",
       "        value += \"ctrl+\";\n",
       "    if (event.altKey && event.which != 18)\n",
       "        value += \"alt+\";\n",
       "    if (event.shiftKey && event.which != 16)\n",
       "        value += \"shift+\";\n",
       "\n",
       "    value += 'k';\n",
       "    value += event.which.toString();\n",
       "\n",
       "    this._key_event_extra(event, name);\n",
       "\n",
       "    this.send_message(name, {key: value,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onclick = function(name) {\n",
       "    if (name == 'download') {\n",
       "        this.handle_save(this, null);\n",
       "    } else {\n",
       "        this.send_message(\"toolbar_button\", {name: name});\n",
       "    }\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onmouseover = function(tooltip) {\n",
       "    this.message.textContent = tooltip;\n",
       "};\n",
       "mpl.toolbar_items = [[\"Home\", \"Reset original view\", \"fa fa-home icon-home\", \"home\"], [\"Back\", \"Back to  previous view\", \"fa fa-arrow-left icon-arrow-left\", \"back\"], [\"Forward\", \"Forward to next view\", \"fa fa-arrow-right icon-arrow-right\", \"forward\"], [\"\", \"\", \"\", \"\"], [\"Pan\", \"Pan axes with left mouse, zoom with right\", \"fa fa-arrows icon-move\", \"pan\"], [\"Zoom\", \"Zoom to rectangle\", \"fa fa-square-o icon-check-empty\", \"zoom\"], [\"\", \"\", \"\", \"\"], [\"Download\", \"Download plot\", \"fa fa-floppy-o icon-save\", \"download\"]];\n",
       "\n",
       "mpl.extensions = [\"eps\", \"jpeg\", \"pdf\", \"png\", \"ps\", \"raw\", \"svg\", \"tif\"];\n",
       "\n",
       "mpl.default_extension = \"png\";var comm_websocket_adapter = function(comm) {\n",
       "    // Create a \"websocket\"-like object which calls the given IPython comm\n",
       "    // object with the appropriate methods. Currently this is a non binary\n",
       "    // socket, so there is still some room for performance tuning.\n",
       "    var ws = {};\n",
       "\n",
       "    ws.close = function() {\n",
       "        comm.close()\n",
       "    };\n",
       "    ws.send = function(m) {\n",
       "        //console.log('sending', m);\n",
       "        comm.send(m);\n",
       "    };\n",
       "    // Register the callback with on_msg.\n",
       "    comm.on_msg(function(msg) {\n",
       "        //console.log('receiving', msg['content']['data'], msg);\n",
       "        // Pass the mpl event to the overriden (by mpl) onmessage function.\n",
       "        ws.onmessage(msg['content']['data'])\n",
       "    });\n",
       "    return ws;\n",
       "}\n",
       "\n",
       "mpl.mpl_figure_comm = function(comm, msg) {\n",
       "    // This is the function which gets called when the mpl process\n",
       "    // starts-up an IPython Comm through the \"matplotlib\" channel.\n",
       "\n",
       "    var id = msg.content.data.id;\n",
       "    // Get hold of the div created by the display call when the Comm\n",
       "    // socket was opened in Python.\n",
       "    var element = $(\"#\" + id);\n",
       "    var ws_proxy = comm_websocket_adapter(comm)\n",
       "\n",
       "    function ondownload(figure, format) {\n",
       "        window.open(figure.imageObj.src);\n",
       "    }\n",
       "\n",
       "    var fig = new mpl.figure(id, ws_proxy,\n",
       "                           ondownload,\n",
       "                           element.get(0));\n",
       "\n",
       "    // Call onopen now - mpl needs it, as it is assuming we've passed it a real\n",
       "    // web socket which is closed, not our websocket->open comm proxy.\n",
       "    ws_proxy.onopen();\n",
       "\n",
       "    fig.parent_element = element.get(0);\n",
       "    fig.cell_info = mpl.find_output_cell(\"<div id='\" + id + \"'></div>\");\n",
       "    if (!fig.cell_info) {\n",
       "        console.error(\"Failed to find cell for figure\", id, fig);\n",
       "        return;\n",
       "    }\n",
       "\n",
       "    var output_index = fig.cell_info[2]\n",
       "    var cell = fig.cell_info[0];\n",
       "\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.handle_close = function(fig, msg) {\n",
       "    fig.root.unbind('remove')\n",
       "\n",
       "    // Update the output cell to use the data from the current canvas.\n",
       "    fig.push_to_output();\n",
       "    var dataURL = fig.canvas.toDataURL();\n",
       "    // Re-enable the keyboard manager in IPython - without this line, in FF,\n",
       "    // the notebook keyboard shortcuts fail.\n",
       "    IPython.keyboard_manager.enable()\n",
       "    $(fig.parent_element).html('<img src=\"' + dataURL + '\">');\n",
       "    fig.close_ws(fig, msg);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.close_ws = function(fig, msg){\n",
       "    fig.send_message('closing', msg);\n",
       "    // fig.ws.close()\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.push_to_output = function(remove_interactive) {\n",
       "    // Turn the data on the canvas into data in the output cell.\n",
       "    var dataURL = this.canvas.toDataURL();\n",
       "    this.cell_info[1]['text/html'] = '<img src=\"' + dataURL + '\">';\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Tell IPython that the notebook contents must change.\n",
       "    IPython.notebook.set_dirty(true);\n",
       "    this.send_message(\"ack\", {});\n",
       "    var fig = this;\n",
       "    // Wait a second, then push the new image to the DOM so\n",
       "    // that it is saved nicely (might be nice to debounce this).\n",
       "    setTimeout(function () { fig.push_to_output() }, 1000);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>')\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items){\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) { continue; };\n",
       "\n",
       "        var button = $('<button class=\"btn btn-default\" href=\"#\" title=\"' + name + '\"><i class=\"fa ' + image + ' fa-lg\"></i></button>');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    // Add the status bar.\n",
       "    var status_bar = $('<span class=\"mpl-message\" style=\"text-align:right; float: right;\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "\n",
       "    // Add the close button to the window.\n",
       "    var buttongrp = $('<div class=\"btn-group inline pull-right\"></div>');\n",
       "    var button = $('<button class=\"btn btn-mini btn-primary\" href=\"#\" title=\"Stop Interaction\"><i class=\"fa fa-power-off icon-remove icon-large\"></i></button>');\n",
       "    button.click(function (evt) { fig.handle_close(fig, {}); } );\n",
       "    button.mouseover('Stop Interaction', toolbar_mouse_event);\n",
       "    buttongrp.append(button);\n",
       "    var titlebar = this.root.find($('.ui-dialog-titlebar'));\n",
       "    titlebar.prepend(buttongrp);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(el){\n",
       "    var fig = this\n",
       "    el.on(\"remove\", function(){\n",
       "\tfig.close_ws(fig, {});\n",
       "    });\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(el){\n",
       "    // this is important to make the div 'focusable\n",
       "    el.attr('tabindex', 0)\n",
       "    // reach out to IPython and tell the keyboard manager to turn it's self\n",
       "    // off when our div gets focus\n",
       "\n",
       "    // location in version 3\n",
       "    if (IPython.notebook.keyboard_manager) {\n",
       "        IPython.notebook.keyboard_manager.register_events(el);\n",
       "    }\n",
       "    else {\n",
       "        // location in version 2\n",
       "        IPython.keyboard_manager.register_events(el);\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    var manager = IPython.notebook.keyboard_manager;\n",
       "    if (!manager)\n",
       "        manager = IPython.keyboard_manager;\n",
       "\n",
       "    // Check for shift+enter\n",
       "    if (event.shiftKey && event.which == 13) {\n",
       "        this.canvas_div.blur();\n",
       "        event.shiftKey = false;\n",
       "        // Send a \"J\" for go to next cell\n",
       "        event.which = 74;\n",
       "        event.keyCode = 74;\n",
       "        manager.command_mode();\n",
       "        manager.handle_keydown(event);\n",
       "    }\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    fig.ondownload(fig, null);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.find_output_cell = function(html_output) {\n",
       "    // Return the cell and output element which can be found *uniquely* in the notebook.\n",
       "    // Note - this is a bit hacky, but it is done because the \"notebook_saving.Notebook\"\n",
       "    // IPython event is triggered only after the cells have been serialised, which for\n",
       "    // our purposes (turning an active figure into a static one), is too late.\n",
       "    var cells = IPython.notebook.get_cells();\n",
       "    var ncells = cells.length;\n",
       "    for (var i=0; i<ncells; i++) {\n",
       "        var cell = cells[i];\n",
       "        if (cell.cell_type === 'code'){\n",
       "            for (var j=0; j<cell.output_area.outputs.length; j++) {\n",
       "                var data = cell.output_area.outputs[j];\n",
       "                if (data.data) {\n",
       "                    // IPython >= 3 moved mimebundle to data attribute of output\n",
       "                    data = data.data;\n",
       "                }\n",
       "                if (data['text/html'] == html_output) {\n",
       "                    return [cell, data, j];\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "}\n",
       "\n",
       "// Register the function which deals with the matplotlib target/channel.\n",
       "// The kernel may be null if the page has been refreshed.\n",
       "if (IPython.notebook.kernel != null) {\n",
       "    IPython.notebook.kernel.comm_manager.register_target('matplotlib', mpl.mpl_figure_comm);\n",
       "}\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div id='08fcafcd-4a40-491f-8520-4001c81eb2b7'></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##################### \n",
    "#You can either choose to keep the parameters specified before, or change them.\n",
    "# To keep the same parameters, comment lines below\n",
    "nbecotype2 = 60\n",
    "min_size2 =0\n",
    "max_size2 = 1000000\n",
    "window2 = 1000\n",
    "numchr2 = [1]\n",
    "########################################\n",
    "\n",
    "\n",
    "if numchr2 == 'all' :\n",
    "    for i in range (1, nbchr+1):\n",
    "        exec('plot%d =m.getplot(onlymethyl%d,nbecotype2,min_size2,max_size2,window2)' % (i,i))\n",
    "else :\n",
    "    for i in numchr2 :\n",
    "        exec('plot%d =m.getplot(onlymethyl%d,nbecotype2,min_size2,max_size2,window2)' % (i,i))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## Converting the data onto sequence ##\n",
    "\n",
    "If a site is methylated it will be a C and if not a A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Reading chromosome', 1)\n",
      "I read ecotype 0/59 : Col_0\n",
      "I wrote the fasta file for the ecotype 0/59 : Col_0\n",
      "I read ecotype 1/59 : Fr_2\n",
      "I wrote the fasta file for the ecotype 1/59 : Fr_2\n",
      "I read ecotype 2/59 : Kro_0\n",
      "I wrote the fasta file for the ecotype 2/59 : Kro_0\n",
      "I read ecotype 3/59 : Bsch_0\n",
      "I wrote the fasta file for the ecotype 3/59 : Bsch_0\n",
      "I read ecotype 4/59 : En_D\n",
      "I wrote the fasta file for the ecotype 4/59 : En_D\n",
      "I read ecotype 5/59 : Kelsterbach_4\n",
      "I wrote the fasta file for the ecotype 5/59 : Kelsterbach_4\n",
      "I read ecotype 6/59 : Aa_0\n",
      "I wrote the fasta file for the ecotype 6/59 : Aa_0\n",
      "I read ecotype 7/59 : Abd_0\n",
      "I wrote the fasta file for the ecotype 7/59 : Abd_0\n",
      "I read ecotype 8/59 : Ag_0\n",
      "I wrote the fasta file for the ecotype 8/59 : Ag_0\n",
      "I read ecotype 9/59 : Ak_1\n",
      "I wrote the fasta file for the ecotype 9/59 : Ak_1\n",
      "I read ecotype 10/59 : Altai_5\n",
      "I wrote the fasta file for the ecotype 10/59 : Altai_5\n",
      "I read ecotype 11/59 : Amel_1\n",
      "I wrote the fasta file for the ecotype 11/59 : Amel_1\n",
      "I read ecotype 12/59 : An_1\n",
      "I wrote the fasta file for the ecotype 12/59 : An_1\n",
      "I read ecotype 13/59 : Anholt_1\n",
      "I wrote the fasta file for the ecotype 13/59 : Anholt_1\n",
      "I read ecotype 14/59 : Ann_1\n",
      "I wrote the fasta file for the ecotype 14/59 : Ann_1\n",
      "I read ecotype 15/59 : Anz_0\n",
      "I wrote the fasta file for the ecotype 15/59 : Anz_0\n",
      "I read ecotype 16/59 : Appt_1\n",
      "I wrote the fasta file for the ecotype 16/59 : Appt_1\n",
      "I read ecotype 17/59 : Ba_1\n",
      "I wrote the fasta file for the ecotype 17/59 : Ba_1\n",
      "I read ecotype 18/59 : Baa_1\n",
      "I wrote the fasta file for the ecotype 18/59 : Baa_1\n",
      "I read ecotype 19/59 : Benk_1\n",
      "I wrote the fasta file for the ecotype 19/59 : Benk_1\n",
      "I read ecotype 20/59 : Bik_1\n",
      "I wrote the fasta file for the ecotype 20/59 : Bik_1\n",
      "I read ecotype 21/59 : Bl_1\n",
      "I wrote the fasta file for the ecotype 21/59 : Bl_1\n",
      "I read ecotype 22/59 : Bla_1\n",
      "I wrote the fasta file for the ecotype 22/59 : Bla_1\n",
      "I read ecotype 23/59 : Boot_1\n",
      "I wrote the fasta file for the ecotype 23/59 : Boot_1\n",
      "I read ecotype 24/59 : Bor_4\n",
      "I wrote the fasta file for the ecotype 24/59 : Bor_4\n",
      "I read ecotype 25/59 : Bs_1\n",
      "I wrote the fasta file for the ecotype 25/59 : Bs_1\n",
      "I read ecotype 26/59 : Bu_0\n",
      "I wrote the fasta file for the ecotype 26/59 : Bu_0\n",
      "I read ecotype 27/59 : Ca_0\n",
      "I wrote the fasta file for the ecotype 27/59 : Ca_0\n",
      "I read ecotype 28/59 : Cal_0\n",
      "I wrote the fasta file for the ecotype 28/59 : Cal_0\n",
      "I read ecotype 29/59 : Chat_1\n",
      "I wrote the fasta file for the ecotype 29/59 : Chat_1\n",
      "I read ecotype 30/59 : Chi_0\n",
      "I wrote the fasta file for the ecotype 30/59 : Chi_0\n",
      "I read ecotype 31/59 : Cnt_1\n",
      "I wrote the fasta file for the ecotype 31/59 : Cnt_1\n",
      "I read ecotype 32/59 : Com_1\n",
      "I wrote the fasta file for the ecotype 32/59 : Com_1\n",
      "I read ecotype 33/59 : Cvi_0\n",
      "I wrote the fasta file for the ecotype 33/59 : Cvi_0\n",
      "I read ecotype 34/59 : Da_1_12\n",
      "I wrote the fasta file for the ecotype 34/59 : Da_1_12\n",
      "I read ecotype 35/59 : Db_1\n",
      "I wrote the fasta file for the ecotype 35/59 : Db_1\n",
      "I read ecotype 36/59 : Di_G\n",
      "I wrote the fasta file for the ecotype 36/59 : Di_G\n",
      "I read ecotype 37/59 : Dja_1\n",
      "I wrote the fasta file for the ecotype 37/59 : Dja_1\n",
      "I read ecotype 38/59 : Ei_2\n",
      "I wrote the fasta file for the ecotype 38/59 : Ei_2\n",
      "I read ecotype 39/59 : El_0\n",
      "I wrote the fasta file for the ecotype 39/59 : El_0\n",
      "I read ecotype 40/59 : Ema_1\n",
      "I wrote the fasta file for the ecotype 40/59 : Ema_1\n",
      "I read ecotype 41/59 : Es_0\n",
      "I wrote the fasta file for the ecotype 41/59 : Es_0\n",
      "I read ecotype 42/59 : Est\n",
      "I wrote the fasta file for the ecotype 42/59 : Est\n",
      "I read ecotype 43/59 : Et_0\n",
      "I wrote the fasta file for the ecotype 43/59 : Et_0\n",
      "I read ecotype 44/59 : Fi_0\n",
      "I wrote the fasta file for the ecotype 44/59 : Fi_0\n",
      "I read ecotype 45/59 : Ga_0\n",
      "I wrote the fasta file for the ecotype 45/59 : Ga_0\n",
      "I read ecotype 46/59 : Gel_1\n",
      "I wrote the fasta file for the ecotype 46/59 : Gel_1\n",
      "I read ecotype 47/59 : Gie_0\n",
      "I wrote the fasta file for the ecotype 47/59 : Gie_0\n",
      "I read ecotype 48/59 : Gifu_2\n",
      "I wrote the fasta file for the ecotype 48/59 : Gifu_2\n",
      "I read ecotype 49/59 : Got_7\n",
      "I wrote the fasta file for the ecotype 49/59 : Got_7\n",
      "I read ecotype 50/59 : Gr_1\n",
      "I wrote the fasta file for the ecotype 50/59 : Gr_1\n",
      "I read ecotype 51/59 : Gre_0\n",
      "I wrote the fasta file for the ecotype 51/59 : Gre_0\n",
      "I read ecotype 52/59 : Gu_0\n",
      "I wrote the fasta file for the ecotype 52/59 : Gu_0\n",
      "I read ecotype 53/59 : Gy_0\n",
      "I wrote the fasta file for the ecotype 53/59 : Gy_0\n",
      "I read ecotype 54/59 : Ha_0\n",
      "I wrote the fasta file for the ecotype 54/59 : Ha_0\n",
      "I read ecotype 55/59 : Hey_1\n",
      "I wrote the fasta file for the ecotype 55/59 : Hey_1\n",
      "I read ecotype 56/59 : Hs_0\n",
      "I wrote the fasta file for the ecotype 56/59 : Hs_0\n",
      "I read ecotype 57/59 : In_0\n",
      "I wrote the fasta file for the ecotype 57/59 : In_0\n",
      "I read ecotype 58/59 : Is_0\n",
      "I wrote the fasta file for the ecotype 58/59 : Is_0\n",
      "I read ecotype 59/59 : Je_0\n",
      "I wrote the fasta file for the ecotype 59/59 : Je_0\n",
      "('Reading chromosome', 2)\n",
      "I read ecotype 0/59 : Col_0\n",
      "I wrote the fasta file for the ecotype 0/59 : Col_0\n",
      "I read ecotype 1/59 : Fr_2\n",
      "I wrote the fasta file for the ecotype 1/59 : Fr_2\n",
      "I read ecotype 2/59 : Kro_0\n",
      "I wrote the fasta file for the ecotype 2/59 : Kro_0\n",
      "I read ecotype 3/59 : Bsch_0\n",
      "I wrote the fasta file for the ecotype 3/59 : Bsch_0\n",
      "I read ecotype 4/59 : En_D\n",
      "I wrote the fasta file for the ecotype 4/59 : En_D\n",
      "I read ecotype 5/59 : Kelsterbach_4\n",
      "I wrote the fasta file for the ecotype 5/59 : Kelsterbach_4\n",
      "I read ecotype 6/59 : Aa_0\n",
      "I wrote the fasta file for the ecotype 6/59 : Aa_0\n",
      "I read ecotype 7/59 : Abd_0\n",
      "I wrote the fasta file for the ecotype 7/59 : Abd_0\n",
      "I read ecotype 8/59 : Ag_0\n",
      "I wrote the fasta file for the ecotype 8/59 : Ag_0\n",
      "I read ecotype 9/59 : Ak_1\n",
      "I wrote the fasta file for the ecotype 9/59 : Ak_1\n",
      "I read ecotype 10/59 : Altai_5\n",
      "I wrote the fasta file for the ecotype 10/59 : Altai_5\n",
      "I read ecotype 11/59 : Amel_1\n",
      "I wrote the fasta file for the ecotype 11/59 : Amel_1\n",
      "I read ecotype 12/59 : An_1\n",
      "I wrote the fasta file for the ecotype 12/59 : An_1\n",
      "I read ecotype 13/59 : Anholt_1\n",
      "I wrote the fasta file for the ecotype 13/59 : Anholt_1\n",
      "I read ecotype 14/59 : Ann_1\n",
      "I wrote the fasta file for the ecotype 14/59 : Ann_1\n",
      "I read ecotype 15/59 : Anz_0\n",
      "I wrote the fasta file for the ecotype 15/59 : Anz_0\n",
      "I read ecotype 16/59 : Appt_1\n",
      "I wrote the fasta file for the ecotype 16/59 : Appt_1\n",
      "I read ecotype 17/59 : Ba_1\n",
      "I wrote the fasta file for the ecotype 17/59 : Ba_1\n",
      "I read ecotype 18/59 : Baa_1\n",
      "I wrote the fasta file for the ecotype 18/59 : Baa_1\n",
      "I read ecotype 19/59 : Benk_1\n",
      "I wrote the fasta file for the ecotype 19/59 : Benk_1\n",
      "I read ecotype 20/59 : Bik_1\n",
      "I wrote the fasta file for the ecotype 20/59 : Bik_1\n",
      "I read ecotype 21/59 : Bl_1\n",
      "I wrote the fasta file for the ecotype 21/59 : Bl_1\n",
      "I read ecotype 22/59 : Bla_1\n",
      "I wrote the fasta file for the ecotype 22/59 : Bla_1\n",
      "I read ecotype 23/59 : Boot_1\n",
      "I wrote the fasta file for the ecotype 23/59 : Boot_1\n",
      "I read ecotype 24/59 : Bor_4\n",
      "I wrote the fasta file for the ecotype 24/59 : Bor_4\n",
      "I read ecotype 25/59 : Bs_1\n",
      "I wrote the fasta file for the ecotype 25/59 : Bs_1\n",
      "I read ecotype 26/59 : Bu_0\n",
      "I wrote the fasta file for the ecotype 26/59 : Bu_0\n",
      "I read ecotype 27/59 : Ca_0\n",
      "I wrote the fasta file for the ecotype 27/59 : Ca_0\n",
      "I read ecotype 28/59 : Cal_0\n",
      "I wrote the fasta file for the ecotype 28/59 : Cal_0\n",
      "I read ecotype 29/59 : Chat_1\n",
      "I wrote the fasta file for the ecotype 29/59 : Chat_1\n",
      "I read ecotype 30/59 : Chi_0\n",
      "I wrote the fasta file for the ecotype 30/59 : Chi_0\n",
      "I read ecotype 31/59 : Cnt_1\n",
      "I wrote the fasta file for the ecotype 31/59 : Cnt_1\n",
      "I read ecotype 32/59 : Com_1\n",
      "I wrote the fasta file for the ecotype 32/59 : Com_1\n",
      "I read ecotype 33/59 : Cvi_0\n",
      "I wrote the fasta file for the ecotype 33/59 : Cvi_0\n",
      "I read ecotype 34/59 : Da_1_12\n",
      "I wrote the fasta file for the ecotype 34/59 : Da_1_12\n",
      "I read ecotype 35/59 : Db_1\n",
      "I wrote the fasta file for the ecotype 35/59 : Db_1\n",
      "I read ecotype 36/59 : Di_G\n",
      "I wrote the fasta file for the ecotype 36/59 : Di_G\n",
      "I read ecotype 37/59 : Dja_1\n",
      "I wrote the fasta file for the ecotype 37/59 : Dja_1\n",
      "I read ecotype 38/59 : Ei_2\n",
      "I wrote the fasta file for the ecotype 38/59 : Ei_2\n",
      "I read ecotype 39/59 : El_0\n",
      "I wrote the fasta file for the ecotype 39/59 : El_0\n",
      "I read ecotype 40/59 : Ema_1\n",
      "I wrote the fasta file for the ecotype 40/59 : Ema_1\n",
      "I read ecotype 41/59 : Es_0\n",
      "I wrote the fasta file for the ecotype 41/59 : Es_0\n",
      "I read ecotype 42/59 : Est\n",
      "I wrote the fasta file for the ecotype 42/59 : Est\n",
      "I read ecotype 43/59 : Et_0\n",
      "I wrote the fasta file for the ecotype 43/59 : Et_0\n",
      "I read ecotype 44/59 : Fi_0\n",
      "I wrote the fasta file for the ecotype 44/59 : Fi_0\n",
      "I read ecotype 45/59 : Ga_0\n",
      "I wrote the fasta file for the ecotype 45/59 : Ga_0\n",
      "I read ecotype 46/59 : Gel_1\n",
      "I wrote the fasta file for the ecotype 46/59 : Gel_1\n",
      "I read ecotype 47/59 : Gie_0\n",
      "I wrote the fasta file for the ecotype 47/59 : Gie_0\n",
      "I read ecotype 48/59 : Gifu_2\n",
      "I wrote the fasta file for the ecotype 48/59 : Gifu_2\n",
      "I read ecotype 49/59 : Got_7\n",
      "I wrote the fasta file for the ecotype 49/59 : Got_7\n",
      "I read ecotype 50/59 : Gr_1\n",
      "I wrote the fasta file for the ecotype 50/59 : Gr_1\n",
      "I read ecotype 51/59 : Gre_0\n",
      "I wrote the fasta file for the ecotype 51/59 : Gre_0\n",
      "I read ecotype 52/59 : Gu_0\n",
      "I wrote the fasta file for the ecotype 52/59 : Gu_0\n",
      "I read ecotype 53/59 : Gy_0\n",
      "I wrote the fasta file for the ecotype 53/59 : Gy_0\n",
      "I read ecotype 54/59 : Ha_0\n",
      "I wrote the fasta file for the ecotype 54/59 : Ha_0\n",
      "I read ecotype 55/59 : Hey_1\n",
      "I wrote the fasta file for the ecotype 55/59 : Hey_1\n",
      "I read ecotype 56/59 : Hs_0\n",
      "I wrote the fasta file for the ecotype 56/59 : Hs_0\n",
      "I read ecotype 57/59 : In_0\n",
      "I wrote the fasta file for the ecotype 57/59 : In_0\n",
      "I read ecotype 58/59 : Is_0\n",
      "I wrote the fasta file for the ecotype 58/59 : Is_0\n",
      "I read ecotype 59/59 : Je_0\n",
      "I wrote the fasta file for the ecotype 59/59 : Je_0\n",
      "('Reading chromosome', 3)\n",
      "I read ecotype 0/59 : Col_0\n",
      "I wrote the fasta file for the ecotype 0/59 : Col_0\n",
      "I read ecotype 1/59 : Fr_2\n",
      "I wrote the fasta file for the ecotype 1/59 : Fr_2\n",
      "I read ecotype 2/59 : Kro_0\n",
      "I wrote the fasta file for the ecotype 2/59 : Kro_0\n",
      "I read ecotype 3/59 : Bsch_0\n",
      "I wrote the fasta file for the ecotype 3/59 : Bsch_0\n",
      "I read ecotype 4/59 : En_D\n",
      "I wrote the fasta file for the ecotype 4/59 : En_D\n",
      "I read ecotype 5/59 : Kelsterbach_4\n",
      "I wrote the fasta file for the ecotype 5/59 : Kelsterbach_4\n",
      "I read ecotype 6/59 : Aa_0\n",
      "I wrote the fasta file for the ecotype 6/59 : Aa_0\n",
      "I read ecotype 7/59 : Abd_0\n",
      "I wrote the fasta file for the ecotype 7/59 : Abd_0\n",
      "I read ecotype 8/59 : Ag_0\n",
      "I wrote the fasta file for the ecotype 8/59 : Ag_0\n",
      "I read ecotype 9/59 : Ak_1\n",
      "I wrote the fasta file for the ecotype 9/59 : Ak_1\n",
      "I read ecotype 10/59 : Altai_5\n",
      "I wrote the fasta file for the ecotype 10/59 : Altai_5\n",
      "I read ecotype 11/59 : Amel_1\n",
      "I wrote the fasta file for the ecotype 11/59 : Amel_1\n",
      "I read ecotype 12/59 : An_1\n",
      "I wrote the fasta file for the ecotype 12/59 : An_1\n",
      "I read ecotype 13/59 : Anholt_1\n",
      "I wrote the fasta file for the ecotype 13/59 : Anholt_1\n",
      "I read ecotype 14/59 : Ann_1\n",
      "I wrote the fasta file for the ecotype 14/59 : Ann_1\n",
      "I read ecotype 15/59 : Anz_0\n",
      "I wrote the fasta file for the ecotype 15/59 : Anz_0\n",
      "I read ecotype 16/59 : Appt_1\n",
      "I wrote the fasta file for the ecotype 16/59 : Appt_1\n",
      "I read ecotype 17/59 : Ba_1\n",
      "I wrote the fasta file for the ecotype 17/59 : Ba_1\n",
      "I read ecotype 18/59 : Baa_1\n",
      "I wrote the fasta file for the ecotype 18/59 : Baa_1\n",
      "I read ecotype 19/59 : Benk_1\n",
      "I wrote the fasta file for the ecotype 19/59 : Benk_1\n",
      "I read ecotype 20/59 : Bik_1\n",
      "I wrote the fasta file for the ecotype 20/59 : Bik_1\n",
      "I read ecotype 21/59 : Bl_1\n",
      "I wrote the fasta file for the ecotype 21/59 : Bl_1\n",
      "I read ecotype 22/59 : Bla_1\n",
      "I wrote the fasta file for the ecotype 22/59 : Bla_1\n",
      "I read ecotype 23/59 : Boot_1\n",
      "I wrote the fasta file for the ecotype 23/59 : Boot_1\n",
      "I read ecotype 24/59 : Bor_4\n",
      "I wrote the fasta file for the ecotype 24/59 : Bor_4\n",
      "I read ecotype 25/59 : Bs_1\n",
      "I wrote the fasta file for the ecotype 25/59 : Bs_1\n",
      "I read ecotype 26/59 : Bu_0\n",
      "I wrote the fasta file for the ecotype 26/59 : Bu_0\n",
      "I read ecotype 27/59 : Ca_0\n",
      "I wrote the fasta file for the ecotype 27/59 : Ca_0\n",
      "I read ecotype 28/59 : Cal_0\n",
      "I wrote the fasta file for the ecotype 28/59 : Cal_0\n",
      "I read ecotype 29/59 : Chat_1\n",
      "I wrote the fasta file for the ecotype 29/59 : Chat_1\n",
      "I read ecotype 30/59 : Chi_0\n",
      "I wrote the fasta file for the ecotype 30/59 : Chi_0\n",
      "I read ecotype 31/59 : Cnt_1\n",
      "I wrote the fasta file for the ecotype 31/59 : Cnt_1\n",
      "I read ecotype 32/59 : Com_1\n",
      "I wrote the fasta file for the ecotype 32/59 : Com_1\n",
      "I read ecotype 33/59 : Cvi_0\n",
      "I wrote the fasta file for the ecotype 33/59 : Cvi_0\n",
      "I read ecotype 34/59 : Da_1_12\n",
      "I wrote the fasta file for the ecotype 34/59 : Da_1_12\n",
      "I read ecotype 35/59 : Db_1\n",
      "I wrote the fasta file for the ecotype 35/59 : Db_1\n",
      "I read ecotype 36/59 : Di_G\n",
      "I wrote the fasta file for the ecotype 36/59 : Di_G\n",
      "I read ecotype 37/59 : Dja_1\n",
      "I wrote the fasta file for the ecotype 37/59 : Dja_1\n",
      "I read ecotype 38/59 : Ei_2\n",
      "I wrote the fasta file for the ecotype 38/59 : Ei_2\n",
      "I read ecotype 39/59 : El_0\n",
      "I wrote the fasta file for the ecotype 39/59 : El_0\n",
      "I read ecotype 40/59 : Ema_1\n",
      "I wrote the fasta file for the ecotype 40/59 : Ema_1\n",
      "I read ecotype 41/59 : Es_0\n",
      "I wrote the fasta file for the ecotype 41/59 : Es_0\n",
      "I read ecotype 42/59 : Est\n",
      "I wrote the fasta file for the ecotype 42/59 : Est\n",
      "I read ecotype 43/59 : Et_0\n",
      "I wrote the fasta file for the ecotype 43/59 : Et_0\n",
      "I read ecotype 44/59 : Fi_0\n",
      "I wrote the fasta file for the ecotype 44/59 : Fi_0\n",
      "I read ecotype 45/59 : Ga_0\n",
      "I wrote the fasta file for the ecotype 45/59 : Ga_0\n",
      "I read ecotype 46/59 : Gel_1\n",
      "I wrote the fasta file for the ecotype 46/59 : Gel_1\n",
      "I read ecotype 47/59 : Gie_0\n",
      "I wrote the fasta file for the ecotype 47/59 : Gie_0\n",
      "I read ecotype 48/59 : Gifu_2\n",
      "I wrote the fasta file for the ecotype 48/59 : Gifu_2\n",
      "I read ecotype 49/59 : Got_7\n",
      "I wrote the fasta file for the ecotype 49/59 : Got_7\n",
      "I read ecotype 50/59 : Gr_1\n",
      "I wrote the fasta file for the ecotype 50/59 : Gr_1\n",
      "I read ecotype 51/59 : Gre_0\n",
      "I wrote the fasta file for the ecotype 51/59 : Gre_0\n",
      "I read ecotype 52/59 : Gu_0\n",
      "I wrote the fasta file for the ecotype 52/59 : Gu_0\n",
      "I read ecotype 53/59 : Gy_0\n",
      "I wrote the fasta file for the ecotype 53/59 : Gy_0\n",
      "I read ecotype 54/59 : Ha_0\n",
      "I wrote the fasta file for the ecotype 54/59 : Ha_0\n",
      "I read ecotype 55/59 : Hey_1\n",
      "I wrote the fasta file for the ecotype 55/59 : Hey_1\n",
      "I read ecotype 56/59 : Hs_0\n",
      "I wrote the fasta file for the ecotype 56/59 : Hs_0\n",
      "I read ecotype 57/59 : In_0\n",
      "I wrote the fasta file for the ecotype 57/59 : In_0\n",
      "I read ecotype 58/59 : Is_0\n",
      "I wrote the fasta file for the ecotype 58/59 : Is_0\n",
      "I read ecotype 59/59 : Je_0\n",
      "I wrote the fasta file for the ecotype 59/59 : Je_0\n",
      "('Reading chromosome', 4)\n",
      "I read ecotype 0/59 : Col_0\n",
      "I wrote the fasta file for the ecotype 0/59 : Col_0\n",
      "I read ecotype 1/59 : Fr_2\n",
      "I wrote the fasta file for the ecotype 1/59 : Fr_2\n",
      "I read ecotype 2/59 : Kro_0\n",
      "I wrote the fasta file for the ecotype 2/59 : Kro_0\n",
      "I read ecotype 3/59 : Bsch_0\n",
      "I wrote the fasta file for the ecotype 3/59 : Bsch_0\n",
      "I read ecotype 4/59 : En_D\n",
      "I wrote the fasta file for the ecotype 4/59 : En_D\n",
      "I read ecotype 5/59 : Kelsterbach_4\n",
      "I wrote the fasta file for the ecotype 5/59 : Kelsterbach_4\n",
      "I read ecotype 6/59 : Aa_0\n",
      "I wrote the fasta file for the ecotype 6/59 : Aa_0\n",
      "I read ecotype 7/59 : Abd_0\n",
      "I wrote the fasta file for the ecotype 7/59 : Abd_0\n",
      "I read ecotype 8/59 : Ag_0\n",
      "I wrote the fasta file for the ecotype 8/59 : Ag_0\n",
      "I read ecotype 9/59 : Ak_1\n",
      "I wrote the fasta file for the ecotype 9/59 : Ak_1\n",
      "I read ecotype 10/59 : Altai_5\n",
      "I wrote the fasta file for the ecotype 10/59 : Altai_5\n",
      "I read ecotype 11/59 : Amel_1\n",
      "I wrote the fasta file for the ecotype 11/59 : Amel_1\n",
      "I read ecotype 12/59 : An_1\n",
      "I wrote the fasta file for the ecotype 12/59 : An_1\n",
      "I read ecotype 13/59 : Anholt_1\n",
      "I wrote the fasta file for the ecotype 13/59 : Anholt_1\n",
      "I read ecotype 14/59 : Ann_1\n",
      "I wrote the fasta file for the ecotype 14/59 : Ann_1\n",
      "I read ecotype 15/59 : Anz_0\n",
      "I wrote the fasta file for the ecotype 15/59 : Anz_0\n",
      "I read ecotype 16/59 : Appt_1\n",
      "I wrote the fasta file for the ecotype 16/59 : Appt_1\n",
      "I read ecotype 17/59 : Ba_1\n",
      "I wrote the fasta file for the ecotype 17/59 : Ba_1\n",
      "I read ecotype 18/59 : Baa_1\n",
      "I wrote the fasta file for the ecotype 18/59 : Baa_1\n",
      "I read ecotype 19/59 : Benk_1\n",
      "I wrote the fasta file for the ecotype 19/59 : Benk_1\n",
      "I read ecotype 20/59 : Bik_1\n",
      "I wrote the fasta file for the ecotype 20/59 : Bik_1\n",
      "I read ecotype 21/59 : Bl_1\n",
      "I wrote the fasta file for the ecotype 21/59 : Bl_1\n",
      "I read ecotype 22/59 : Bla_1\n",
      "I wrote the fasta file for the ecotype 22/59 : Bla_1\n",
      "I read ecotype 23/59 : Boot_1\n",
      "I wrote the fasta file for the ecotype 23/59 : Boot_1\n",
      "I read ecotype 24/59 : Bor_4\n",
      "I wrote the fasta file for the ecotype 24/59 : Bor_4\n",
      "I read ecotype 25/59 : Bs_1\n",
      "I wrote the fasta file for the ecotype 25/59 : Bs_1\n",
      "I read ecotype 26/59 : Bu_0\n",
      "I wrote the fasta file for the ecotype 26/59 : Bu_0\n",
      "I read ecotype 27/59 : Ca_0\n",
      "I wrote the fasta file for the ecotype 27/59 : Ca_0\n",
      "I read ecotype 28/59 : Cal_0\n",
      "I wrote the fasta file for the ecotype 28/59 : Cal_0\n",
      "I read ecotype 29/59 : Chat_1\n",
      "I wrote the fasta file for the ecotype 29/59 : Chat_1\n",
      "I read ecotype 30/59 : Chi_0\n",
      "I wrote the fasta file for the ecotype 30/59 : Chi_0\n",
      "I read ecotype 31/59 : Cnt_1\n",
      "I wrote the fasta file for the ecotype 31/59 : Cnt_1\n",
      "I read ecotype 32/59 : Com_1\n",
      "I wrote the fasta file for the ecotype 32/59 : Com_1\n",
      "I read ecotype 33/59 : Cvi_0\n",
      "I wrote the fasta file for the ecotype 33/59 : Cvi_0\n",
      "I read ecotype 34/59 : Da_1_12\n",
      "I wrote the fasta file for the ecotype 34/59 : Da_1_12\n",
      "I read ecotype 35/59 : Db_1\n",
      "I wrote the fasta file for the ecotype 35/59 : Db_1\n",
      "I read ecotype 36/59 : Di_G\n",
      "I wrote the fasta file for the ecotype 36/59 : Di_G\n",
      "I read ecotype 37/59 : Dja_1\n",
      "I wrote the fasta file for the ecotype 37/59 : Dja_1\n",
      "I read ecotype 38/59 : Ei_2\n",
      "I wrote the fasta file for the ecotype 38/59 : Ei_2\n",
      "I read ecotype 39/59 : El_0\n",
      "I wrote the fasta file for the ecotype 39/59 : El_0\n",
      "I read ecotype 40/59 : Ema_1\n",
      "I wrote the fasta file for the ecotype 40/59 : Ema_1\n",
      "I read ecotype 41/59 : Es_0\n",
      "I wrote the fasta file for the ecotype 41/59 : Es_0\n",
      "I read ecotype 42/59 : Est\n",
      "I wrote the fasta file for the ecotype 42/59 : Est\n",
      "I read ecotype 43/59 : Et_0\n",
      "I wrote the fasta file for the ecotype 43/59 : Et_0\n",
      "I read ecotype 44/59 : Fi_0\n",
      "I wrote the fasta file for the ecotype 44/59 : Fi_0\n",
      "I read ecotype 45/59 : Ga_0\n",
      "I wrote the fasta file for the ecotype 45/59 : Ga_0\n",
      "I read ecotype 46/59 : Gel_1\n",
      "I wrote the fasta file for the ecotype 46/59 : Gel_1\n",
      "I read ecotype 47/59 : Gie_0\n",
      "I wrote the fasta file for the ecotype 47/59 : Gie_0\n",
      "I read ecotype 48/59 : Gifu_2\n",
      "I wrote the fasta file for the ecotype 48/59 : Gifu_2\n",
      "I read ecotype 49/59 : Got_7\n",
      "I wrote the fasta file for the ecotype 49/59 : Got_7\n",
      "I read ecotype 50/59 : Gr_1\n",
      "I wrote the fasta file for the ecotype 50/59 : Gr_1\n",
      "I read ecotype 51/59 : Gre_0\n",
      "I wrote the fasta file for the ecotype 51/59 : Gre_0\n",
      "I read ecotype 52/59 : Gu_0\n",
      "I wrote the fasta file for the ecotype 52/59 : Gu_0\n",
      "I read ecotype 53/59 : Gy_0\n",
      "I wrote the fasta file for the ecotype 53/59 : Gy_0\n",
      "I read ecotype 54/59 : Ha_0\n",
      "I wrote the fasta file for the ecotype 54/59 : Ha_0\n",
      "I read ecotype 55/59 : Hey_1\n",
      "I wrote the fasta file for the ecotype 55/59 : Hey_1\n",
      "I read ecotype 56/59 : Hs_0\n",
      "I wrote the fasta file for the ecotype 56/59 : Hs_0\n",
      "I read ecotype 57/59 : In_0\n",
      "I wrote the fasta file for the ecotype 57/59 : In_0\n",
      "I read ecotype 58/59 : Is_0\n",
      "I wrote the fasta file for the ecotype 58/59 : Is_0\n",
      "I read ecotype 59/59 : Je_0\n",
      "I wrote the fasta file for the ecotype 59/59 : Je_0\n",
      "('Reading chromosome', 5)\n",
      "I read ecotype 0/59 : Col_0\n",
      "I wrote the fasta file for the ecotype 0/59 : Col_0\n",
      "I read ecotype 1/59 : Fr_2\n",
      "I wrote the fasta file for the ecotype 1/59 : Fr_2\n",
      "I read ecotype 2/59 : Kro_0\n",
      "I wrote the fasta file for the ecotype 2/59 : Kro_0\n",
      "I read ecotype 3/59 : Bsch_0\n",
      "I wrote the fasta file for the ecotype 3/59 : Bsch_0\n",
      "I read ecotype 4/59 : En_D\n",
      "I wrote the fasta file for the ecotype 4/59 : En_D\n",
      "I read ecotype 5/59 : Kelsterbach_4\n",
      "I wrote the fasta file for the ecotype 5/59 : Kelsterbach_4\n",
      "I read ecotype 6/59 : Aa_0\n",
      "I wrote the fasta file for the ecotype 6/59 : Aa_0\n",
      "I read ecotype 7/59 : Abd_0\n",
      "I wrote the fasta file for the ecotype 7/59 : Abd_0\n",
      "I read ecotype 8/59 : Ag_0\n",
      "I wrote the fasta file for the ecotype 8/59 : Ag_0\n",
      "I read ecotype 9/59 : Ak_1\n",
      "I wrote the fasta file for the ecotype 9/59 : Ak_1\n",
      "I read ecotype 10/59 : Altai_5\n",
      "I wrote the fasta file for the ecotype 10/59 : Altai_5\n",
      "I read ecotype 11/59 : Amel_1\n",
      "I wrote the fasta file for the ecotype 11/59 : Amel_1\n",
      "I read ecotype 12/59 : An_1\n",
      "I wrote the fasta file for the ecotype 12/59 : An_1\n",
      "I read ecotype 13/59 : Anholt_1\n",
      "I wrote the fasta file for the ecotype 13/59 : Anholt_1\n",
      "I read ecotype 14/59 : Ann_1\n",
      "I wrote the fasta file for the ecotype 14/59 : Ann_1\n",
      "I read ecotype 15/59 : Anz_0\n",
      "I wrote the fasta file for the ecotype 15/59 : Anz_0\n",
      "I read ecotype 16/59 : Appt_1\n",
      "I wrote the fasta file for the ecotype 16/59 : Appt_1\n",
      "I read ecotype 17/59 : Ba_1\n",
      "I wrote the fasta file for the ecotype 17/59 : Ba_1\n",
      "I read ecotype 18/59 : Baa_1\n",
      "I wrote the fasta file for the ecotype 18/59 : Baa_1\n",
      "I read ecotype 19/59 : Benk_1\n",
      "I wrote the fasta file for the ecotype 19/59 : Benk_1\n",
      "I read ecotype 20/59 : Bik_1\n",
      "I wrote the fasta file for the ecotype 20/59 : Bik_1\n",
      "I read ecotype 21/59 : Bl_1\n",
      "I wrote the fasta file for the ecotype 21/59 : Bl_1\n",
      "I read ecotype 22/59 : Bla_1\n",
      "I wrote the fasta file for the ecotype 22/59 : Bla_1\n",
      "I read ecotype 23/59 : Boot_1\n",
      "I wrote the fasta file for the ecotype 23/59 : Boot_1\n",
      "I read ecotype 24/59 : Bor_4\n",
      "I wrote the fasta file for the ecotype 24/59 : Bor_4\n",
      "I read ecotype 25/59 : Bs_1\n",
      "I wrote the fasta file for the ecotype 25/59 : Bs_1\n",
      "I read ecotype 26/59 : Bu_0\n",
      "I wrote the fasta file for the ecotype 26/59 : Bu_0\n",
      "I read ecotype 27/59 : Ca_0\n",
      "I wrote the fasta file for the ecotype 27/59 : Ca_0\n",
      "I read ecotype 28/59 : Cal_0\n",
      "I wrote the fasta file for the ecotype 28/59 : Cal_0\n",
      "I read ecotype 29/59 : Chat_1\n",
      "I wrote the fasta file for the ecotype 29/59 : Chat_1\n",
      "I read ecotype 30/59 : Chi_0\n",
      "I wrote the fasta file for the ecotype 30/59 : Chi_0\n",
      "I read ecotype 31/59 : Cnt_1\n",
      "I wrote the fasta file for the ecotype 31/59 : Cnt_1\n",
      "I read ecotype 32/59 : Com_1\n",
      "I wrote the fasta file for the ecotype 32/59 : Com_1\n",
      "I read ecotype 33/59 : Cvi_0\n",
      "I wrote the fasta file for the ecotype 33/59 : Cvi_0\n",
      "I read ecotype 34/59 : Da_1_12\n",
      "I wrote the fasta file for the ecotype 34/59 : Da_1_12\n",
      "I read ecotype 35/59 : Db_1\n",
      "I wrote the fasta file for the ecotype 35/59 : Db_1\n",
      "I read ecotype 36/59 : Di_G\n",
      "I wrote the fasta file for the ecotype 36/59 : Di_G\n",
      "I read ecotype 37/59 : Dja_1\n",
      "I wrote the fasta file for the ecotype 37/59 : Dja_1\n",
      "I read ecotype 38/59 : Ei_2\n",
      "I wrote the fasta file for the ecotype 38/59 : Ei_2\n",
      "I read ecotype 39/59 : El_0\n",
      "I wrote the fasta file for the ecotype 39/59 : El_0\n",
      "I read ecotype 40/59 : Ema_1\n",
      "I wrote the fasta file for the ecotype 40/59 : Ema_1\n",
      "I read ecotype 41/59 : Es_0\n",
      "I wrote the fasta file for the ecotype 41/59 : Es_0\n",
      "I read ecotype 42/59 : Est\n",
      "I wrote the fasta file for the ecotype 42/59 : Est\n",
      "I read ecotype 43/59 : Et_0\n",
      "I wrote the fasta file for the ecotype 43/59 : Et_0\n",
      "I read ecotype 44/59 : Fi_0\n",
      "I wrote the fasta file for the ecotype 44/59 : Fi_0\n",
      "I read ecotype 45/59 : Ga_0\n",
      "I wrote the fasta file for the ecotype 45/59 : Ga_0\n",
      "I read ecotype 46/59 : Gel_1\n",
      "I wrote the fasta file for the ecotype 46/59 : Gel_1\n",
      "I read ecotype 47/59 : Gie_0\n",
      "I wrote the fasta file for the ecotype 47/59 : Gie_0\n",
      "I read ecotype 48/59 : Gifu_2\n",
      "I wrote the fasta file for the ecotype 48/59 : Gifu_2\n",
      "I read ecotype 49/59 : Got_7\n",
      "I wrote the fasta file for the ecotype 49/59 : Got_7\n",
      "I read ecotype 50/59 : Gr_1\n",
      "I wrote the fasta file for the ecotype 50/59 : Gr_1\n",
      "I read ecotype 51/59 : Gre_0\n",
      "I wrote the fasta file for the ecotype 51/59 : Gre_0\n",
      "I read ecotype 52/59 : Gu_0\n",
      "I wrote the fasta file for the ecotype 52/59 : Gu_0\n",
      "I read ecotype 53/59 : Gy_0\n",
      "I wrote the fasta file for the ecotype 53/59 : Gy_0\n",
      "I read ecotype 54/59 : Ha_0\n",
      "I wrote the fasta file for the ecotype 54/59 : Ha_0\n",
      "I read ecotype 55/59 : Hey_1\n",
      "I wrote the fasta file for the ecotype 55/59 : Hey_1\n",
      "I read ecotype 56/59 : Hs_0\n",
      "I wrote the fasta file for the ecotype 56/59 : Hs_0\n",
      "I read ecotype 57/59 : In_0\n",
      "I wrote the fasta file for the ecotype 57/59 : In_0\n",
      "I read ecotype 58/59 : Is_0\n",
      "I wrote the fasta file for the ecotype 58/59 : Is_0\n",
      "I read ecotype 59/59 : Je_0\n",
      "I wrote the fasta file for the ecotype 59/59 : Je_0\n"
     ]
    }
   ],
   "source": [
    "for k in range (1,nbchr+1):\n",
    "    print('Reading chromosome', k)\n",
    "    exec('fasta_file= open(\"fastafile%d.txt\", \"w\")' % (k))\n",
    "    for j in range (nbecotype):\n",
    "        seq_eco=[]\n",
    "        exec('lenght=len(methylome_file%d[0])' % (k))\n",
    "        for i in range (lenght):\n",
    "            exec('meth=methylome_file%d[%d][\"methylation_call\"].iloc[%d]' % (k,j,i))\n",
    "            if meth ==0:\n",
    "                seq_eco.append('A')\n",
    "            if meth ==1:\n",
    "                seq_eco.append('C')\n",
    "        seq=(\"\".join(seq_eco))\n",
    "        print('I read ecotype %d/%d : %s' %(j+1,nbecotype, names[j]))\n",
    "        fasta_file.write(\">\")\n",
    "        fasta_file.write(list_ecotype[j])\n",
    "        fasta_file.write(\"\\n\")\n",
    "        fasta_file.write(seq)\n",
    "        fasta_file.write(\"\\n\")\n",
    "        print('I wrote the fasta file for the ecotype %d/%d : %s' %(j+1,nbecotype, names[j]))\n",
    "    fasta_file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000\n"
     ]
    }
   ],
   "source": [
    "print(len(methylome_file1[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Calculating Tajima's D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I processed ecotype  0 / 151 for chromosome 1\n",
      "I processed ecotype  1 / 151 for chromosome 1\n",
      "I processed ecotype  2 / 151 for chromosome 1\n",
      "I processed ecotype  3 / 151 for chromosome 1\n",
      "I processed ecotype  4 / 151 for chromosome 1\n",
      "I processed ecotype  5 / 151 for chromosome 1\n",
      "I processed ecotype  6 / 151 for chromosome 1\n",
      "I processed ecotype  7 / 151 for chromosome 1\n",
      "I processed ecotype  8 / 151 for chromosome 1\n",
      "I processed ecotype  9 / 151 for chromosome 1\n",
      "I processed ecotype  10 / 151 for chromosome 1\n",
      "I processed ecotype  11 / 151 for chromosome 1\n",
      "I processed ecotype  12 / 151 for chromosome 1\n",
      "I processed ecotype  13 / 151 for chromosome 1\n",
      "I processed ecotype  14 / 151 for chromosome 1\n",
      "I processed ecotype  15 / 151 for chromosome 1\n",
      "I processed ecotype  16 / 151 for chromosome 1\n",
      "I processed ecotype  17 / 151 for chromosome 1\n",
      "I processed ecotype  18 / 151 for chromosome 1\n",
      "I processed ecotype  19 / 151 for chromosome 1\n",
      "I processed ecotype  20 / 151 for chromosome 1\n",
      "I processed ecotype  21 / 151 for chromosome 1\n",
      "I processed ecotype  22 / 151 for chromosome 1\n",
      "I processed ecotype  23 / 151 for chromosome 1\n",
      "I processed ecotype  24 / 151 for chromosome 1\n",
      "I processed ecotype  25 / 151 for chromosome 1\n",
      "I processed ecotype  26 / 151 for chromosome 1\n",
      "I processed ecotype  27 / 151 for chromosome 1\n",
      "I processed ecotype  28 / 151 for chromosome 1\n",
      "I processed ecotype  29 / 151 for chromosome 1\n",
      "I processed ecotype  30 / 151 for chromosome 1\n",
      "I processed ecotype  31 / 151 for chromosome 1\n",
      "I processed ecotype  32 / 151 for chromosome 1\n",
      "I processed ecotype  33 / 151 for chromosome 1\n",
      "I processed ecotype  34 / 151 for chromosome 1\n",
      "I processed ecotype  35 / 151 for chromosome 1\n",
      "I processed ecotype  36 / 151 for chromosome 1\n",
      "I processed ecotype  37 / 151 for chromosome 1\n",
      "I processed ecotype  38 / 151 for chromosome 1\n",
      "I processed ecotype  39 / 151 for chromosome 1\n",
      "I processed ecotype  40 / 151 for chromosome 1\n",
      "I processed ecotype  41 / 151 for chromosome 1\n",
      "I processed ecotype  42 / 151 for chromosome 1\n",
      "I processed ecotype  43 / 151 for chromosome 1\n",
      "I processed ecotype  44 / 151 for chromosome 1\n",
      "I processed ecotype  45 / 151 for chromosome 1\n",
      "I processed ecotype  46 / 151 for chromosome 1\n",
      "I processed ecotype  47 / 151 for chromosome 1\n",
      "I processed ecotype  48 / 151 for chromosome 1\n",
      "I processed ecotype  49 / 151 for chromosome 1\n",
      "I processed ecotype  50 / 151 for chromosome 1\n",
      "I processed ecotype  51 / 151 for chromosome 1\n",
      "I processed ecotype  52 / 151 for chromosome 1\n",
      "I processed ecotype  53 / 151 for chromosome 1\n",
      "I processed ecotype  54 / 151 for chromosome 1\n",
      "I processed ecotype  55 / 151 for chromosome 1\n",
      "I processed ecotype  56 / 151 for chromosome 1\n",
      "I processed ecotype  57 / 151 for chromosome 1\n",
      "I processed ecotype  58 / 151 for chromosome 1\n",
      "I processed ecotype  59 / 151 for chromosome 1\n",
      "I processed ecotype  60 / 151 for chromosome 1\n",
      "I processed ecotype  61 / 151 for chromosome 1\n",
      "I processed ecotype  62 / 151 for chromosome 1\n",
      "I processed ecotype  63 / 151 for chromosome 1\n",
      "I processed ecotype  64 / 151 for chromosome 1\n",
      "I processed ecotype  65 / 151 for chromosome 1\n",
      "I processed ecotype  66 / 151 for chromosome 1\n",
      "I processed ecotype  67 / 151 for chromosome 1\n",
      "I processed ecotype  68 / 151 for chromosome 1\n",
      "I processed ecotype  69 / 151 for chromosome 1\n",
      "I processed ecotype  70 / 151 for chromosome 1\n",
      "I processed ecotype  71 / 151 for chromosome 1\n",
      "I processed ecotype  72 / 151 for chromosome 1\n",
      "I processed ecotype  73 / 151 for chromosome 1\n",
      "I processed ecotype  74 / 151 for chromosome 1\n",
      "I processed ecotype  75 / 151 for chromosome 1\n",
      "I processed ecotype  76 / 151 for chromosome 1\n",
      "I processed ecotype  77 / 151 for chromosome 1\n",
      "I processed ecotype  78 / 151 for chromosome 1\n",
      "I processed ecotype  79 / 151 for chromosome 1\n",
      "I processed ecotype  80 / 151 for chromosome 1\n",
      "I processed ecotype  81 / 151 for chromosome 1\n",
      "I processed ecotype  82 / 151 for chromosome 1\n",
      "I processed ecotype  83 / 151 for chromosome 1\n",
      "I processed ecotype  84 / 151 for chromosome 1\n",
      "I processed ecotype  85 / 151 for chromosome 1\n",
      "I processed ecotype  86 / 151 for chromosome 1\n",
      "I processed ecotype  87 / 151 for chromosome 1\n",
      "I processed ecotype  88 / 151 for chromosome 1\n",
      "I processed ecotype  89 / 151 for chromosome 1\n",
      "I processed ecotype  90 / 151 for chromosome 1\n",
      "I processed ecotype  91 / 151 for chromosome 1\n",
      "I processed ecotype  92 / 151 for chromosome 1\n",
      "I processed ecotype  93 / 151 for chromosome 1\n",
      "I processed ecotype  94 / 151 for chromosome 1\n",
      "I processed ecotype  95 / 151 for chromosome 1\n",
      "I processed ecotype  96 / 151 for chromosome 1\n",
      "I processed ecotype  97 / 151 for chromosome 1\n",
      "I processed ecotype  98 / 151 for chromosome 1\n",
      "I processed ecotype  99 / 151 for chromosome 1\n",
      "I processed ecotype  100 / 151 for chromosome 1\n",
      "I processed ecotype  101 / 151 for chromosome 1\n",
      "I processed ecotype  102 / 151 for chromosome 1\n",
      "I processed ecotype  103 / 151 for chromosome 1\n",
      "I processed ecotype  104 / 151 for chromosome 1\n",
      "I processed ecotype  105 / 151 for chromosome 1\n",
      "I processed ecotype  106 / 151 for chromosome 1\n",
      "I processed ecotype  107 / 151 for chromosome 1\n",
      "I processed ecotype  108 / 151 for chromosome 1\n",
      "I processed ecotype  109 / 151 for chromosome 1\n",
      "I processed ecotype  110 / 151 for chromosome 1\n",
      "I processed ecotype  111 / 151 for chromosome 1\n",
      "I processed ecotype  112 / 151 for chromosome 1\n",
      "I processed ecotype  113 / 151 for chromosome 1\n",
      "I processed ecotype  114 / 151 for chromosome 1\n",
      "I processed ecotype  115 / 151 for chromosome 1\n",
      "I processed ecotype  116 / 151 for chromosome 1\n",
      "I processed ecotype  117 / 151 for chromosome 1\n",
      "I processed ecotype  118 / 151 for chromosome 1\n",
      "I processed ecotype  119 / 151 for chromosome 1\n",
      "I processed ecotype  120 / 151 for chromosome 1\n",
      "I processed ecotype  121 / 151 for chromosome 1\n",
      "I processed ecotype  122 / 151 for chromosome 1\n",
      "I processed ecotype  123 / 151 for chromosome 1\n",
      "I processed ecotype  124 / 151 for chromosome 1\n",
      "I processed ecotype  125 / 151 for chromosome 1\n",
      "I processed ecotype  126 / 151 for chromosome 1\n",
      "I processed ecotype  127 / 151 for chromosome 1\n",
      "I processed ecotype  128 / 151 for chromosome 1\n",
      "I processed ecotype  129 / 151 for chromosome 1\n",
      "I processed ecotype  130 / 151 for chromosome 1\n",
      "I processed ecotype  131 / 151 for chromosome 1\n",
      "I processed ecotype  132 / 151 for chromosome 1\n",
      "I processed ecotype  133 / 151 for chromosome 1\n",
      "I processed ecotype  134 / 151 for chromosome 1\n",
      "I processed ecotype  135 / 151 for chromosome 1\n",
      "I processed ecotype  136 / 151 for chromosome 1\n",
      "I processed ecotype  137 / 151 for chromosome 1\n",
      "I processed ecotype  138 / 151 for chromosome 1\n",
      "I processed ecotype  139 / 151 for chromosome 1\n",
      "I processed ecotype  140 / 151 for chromosome 1\n",
      "I processed ecotype  141 / 151 for chromosome 1\n",
      "I processed ecotype  142 / 151 for chromosome 1\n",
      "I processed ecotype  143 / 151 for chromosome 1\n",
      "I processed ecotype  144 / 151 for chromosome 1\n",
      "I processed ecotype  145 / 151 for chromosome 1\n",
      "I processed ecotype  146 / 151 for chromosome 1\n",
      "I processed ecotype  147 / 151 for chromosome 1\n",
      "I processed ecotype  148 / 151 for chromosome 1\n",
      "I processed ecotype  149 / 151 for chromosome 1\n",
      "I processed ecotype  150 / 151 for chromosome 1\n",
      "I processed ecotype  151 / 151 for chromosome 1\n"
     ]
    }
   ],
   "source": [
    "# First : we need to have a table containing the positions\n",
    "for n in range (1,nbchr+1):\n",
    "    exec('new%d=[]'%(n))\n",
    "    for i in range (nbecotype) :\n",
    "        #exec('methylome_file%d[i][\"index\"]=methylome_file%d[i].chrom.astype(str).str.cat(methylome_file%d[i].strand.astype(str).str.cat(methylome_file%d[i].pos.astype(str)))' % (n,n,n,n))\n",
    "        exec('methylome_file%d[i][\"Index\"]=methylome_file%d[i].pos' % (n,n))\n",
    "        exec('methylome_file%d[i]=methylome_file%d[i].set_index(methylome_file%d[i][\"Index\"])' % (n,n,n))\n",
    "        exec('new%d.append(pandas.DataFrame(methylome_file%d[i].methylation_call))' % (n,n))\n",
    "        print\"I processed ecotype \",i,'/',nbecotype-1, 'for chromosome', n\n",
    "    exec('tajimas_d%d=pandas.DataFrame(new%d[0].methylation_call)'% (n,n))\n",
    "    exec('tajimas_d%d=tajimas_d%d.rename(columns={\"methylation_call\":\"0\"})' % (n,n))\n",
    "    exec('tajimas_d%d[\"pos\"]=methylome_file%d[0][\"Index\"]'%(n,n))\n",
    "    for i in range (1,nbecotype) :\n",
    "        exec('tajimas_d%d[\"%d\"]=new%d[i]'%(n,i,n))\n",
    "        exec('tajimas_d%d=tajimas_d%d.dropna()'%(n,n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tajimas_d1[(tajimas_d1['pos'] >= 433031 )& (tajimas_d1['pos']<=433819)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pi = [] #meanwise pair difference : looking at each pair of sequence to see how many variation do we have \n",
    "theta_pi =[] # estimate of the mutation rate based on the nucleotide diversity\n",
    "theta_s =[]  #estimate of the mutation rate based on the number of segregating sites \n",
    "s =[] # number of segregating sites : total number of sites that are different\n",
    "#using only the methylable sites\n",
    "tajimas=[] #new list\n",
    "########################TAJIMA'S D PARAMETERS#######################\n",
    "a1=sum(1./i for i in range (1,nbecotype))\n",
    "a2=sum(1./i**2 for i in range (1,nbecotype))\n",
    "n=nbecotype\n",
    "b1=(n+1.)/(3*(n-1))\n",
    "b2=(2*(n**2+n+3.))/(9*n*(n-1.))\n",
    "c1=b1-(1/a1)\n",
    "c2=b2 - ((n+2)/(a1*n))+(a2/a1**2)\n",
    "e1=c1/a1\n",
    "e2=c2/(a1**2 +a2)\n",
    "####################################################################\n",
    "import numpy as np\n",
    "window_list=numpy.arange(min_size,max_size+window,window)\n",
    "for n in range (1,nbchr+1):\n",
    "    for i in range (1, int((max_size+window-min_size)/window)) :\n",
    "        pi_int=[]\n",
    "        exec('truc=tajimas_d%d[(tajimas_d%d.pos> window_list[i-1])&(tajimas_d%d.pos<= window_list[i])]'%(n,n,n))\n",
    "        len1=len(truc)\n",
    "        truc=truc.drop(truc.columns[1], axis=1) #droping the column with position number        \n",
    "        for i in range (nbecotype) :\n",
    "            for j in range (nbecotype) :\n",
    "                if i < j :\n",
    "                    exec('pi_int.append(sum(truc[\"%d\"] != truc[\"%d\"]))' % (i,j))\n",
    "        sum_pi_int=numpy.mean(pi_int)\n",
    "        pi.append(numpy.mean(pi_int))\n",
    "        truc2=truc[truc.apply(pandas.Series.nunique, axis=1) == 1] #keeping only the row where all the columns have the same number\n",
    "        len2=len(truc2)\n",
    "        s_int=len1-len2\n",
    "        s.append(s_int)\n",
    "        tajimas.append((sum_pi_int - s_int/a1)/np.sqrt(e1*s_int +e2*s_int*(s_int-1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Ploting S values\n",
    "s=pandas.DataFrame(s)\n",
    "pi=pandas.DataFrame(pi)\n",
    "tajimas=pandas.DataFrame(tajimas)\n",
    "print(pi)\n",
    "#s.plot()\n",
    "#pi.plot()\n",
    "#tajimas.plot()\n",
    "print(np.mean(tajimas))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "s1 = \"GAT-ACA\"\n",
    "s2 = \"AT-GCGA\"\n",
    "s3 = \"AAAAAAA\"\n",
    "list=[s1,s2,s3]\n",
    "hi=[]\n",
    "for i in list :\n",
    "    for j in list :\n",
    "        if i<j:\n",
    "            hi.append(sum (a != b for (a,b) in zip(i,j)))\n",
    "numpy.mean(hi)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code bit by bit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importing the libraries we will use :\n",
    "\n",
    "- os : allowing us to specify the path to file without having to worry about the / or \\ depending on the operating system used\n",
    "- pandas : usefull for easy data/table manipulation\n",
    "- matplotlib : for plots, using the matplotlib notebook allow us to zoom in plots etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Importing libraries\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "%matplotlib notebook\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('ggplot')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Definition of the parameters that will be used later on :\n",
    "\n",
    "+ the number of chromosoms we want to look at\n",
    "+ the number of rows we want to read if we don't want to read the entire file\n",
    "+ the number of ecotypes we are studying\n",
    "+ the size of the window we want to use for the analysis\n",
    "+ the size of the longest chromosome\n",
    "+ the file were are stored the names of the ecotypes\n",
    "+ python version\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Defining parameters\n",
    "nbchr = 1 #numbre of chromosoms\n",
    "nrows = 140000#number of rows to read\n",
    "nbecotype=15 #number of ecotypes\n",
    "window=1000 #size of the window in basepair\n",
    "max_size= 400000 #size of the longest chromosome\n",
    "list_ecotypes=os.path.join('ecotypes.txt') #list of the ecotypes that will be used. Put the ref FIRST\n",
    "python = 2 #Specify the python version 2 or 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we are creating an array with the list of the position of the begginning of each window."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Creating a list with the different windows\n",
    "window_list=np.arange(0,max_size+window,window)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run to read the entire files and separating the data by chromosomes\n",
    "\n",
    "We are reading the ecotype files. First we transpose the list of ecotypes into an readable array. Then we are reading each entry of this array and reading the associated file to append it to methylome_files.\n",
    "methylome_files[0] will be the file for the ecotype 0 and so on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Reading the ecotype files\n",
    "imp=open(list_ecotypes, 'r')\n",
    "list_ecotype=[]\n",
    "for line in imp:\n",
    "    for i in line.split():\n",
    "        list_ecotype.append(i)\n",
    "methylome_files=[]\n",
    "for i in range (nbecotype) :\n",
    "    methylome_files.append(pd.read_csv(os.path.join(list_ecotype[i]), sep='\\t'))\n",
    "    if python == 2 :\n",
    "        print 'I read ecotype', i , ':', list_ecotype[i]\n",
    "    if python == 3 :\n",
    "        print ('I read ecotype', i , ':', list_ecotype[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run to read only a specified number of rows\n",
    "Same thing as above but if we only want to read a certain number of rows in all the files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Reading the ecotype files\n",
    "imp=open(list_ecotypes, 'r')\n",
    "list_ecotype=[]\n",
    "for line in imp:\n",
    "    for i in line.split():\n",
    "        list_ecotype.append(i)\n",
    "methylome_files=[]\n",
    "for i in range (nbecotype) :\n",
    "    methylome_files.append(pd.read_csv(os.path.join(list_ecotype[i]), sep='\\t',nrows=nrows))\n",
    "    if python == 2 :\n",
    "        print 'I read ecotype', i , ':', list_ecotype[i]\n",
    "    if python == 3 :\n",
    "        print ('I read ecotype', i , ':', list_ecotype[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distinguishing between the different methylation class\n",
    "We are creating a new "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Keep only the rows where we have a methylation\n",
    "#onlymethyl=[]\n",
    "#for i in range (nbecotype) :\n",
    "#    onlymethyl.append(pd.DataFrame(methylome_files[i][methylome_files[i].methylation_call==1]))\n",
    "\n",
    "#Defining the different methylation class depending on the nucleotides\n",
    "CHGs=[]\n",
    "for i in range (nbecotype) :\n",
    "    CHGs.append(onlymethyl[i][onlymethyl[i].mc_class=='CAG'].append(onlymethyl[i][onlymethyl[i].mc_class=='CCG']).append(onlymethyl[i][onlymethyl[i].mc_class=='CTG']))\n",
    "CGs=[]\n",
    "for i in range (nbecotype) :\n",
    "    CGs.append(onlymethyl[i][onlymethyl[i].mc_class=='CGA'].append(onlymethyl[i][onlymethyl[i].mc_class=='CGT']).append(onlymethyl[i][onlymethyl[i].mc_class=='CGC']).append(onlymethyl[i][onlymethyl[i].mc_class=='CGG']))\n",
    "CHHs=[]\n",
    "for i in range (nbecotype) :\n",
    "    CHHs.append(onlymethyl[i][onlymethyl[i].mc_class=='CAA'].append(onlymethyl[i][onlymethyl[i].mc_class=='CAT']).append(onlymethyl[i][onlymethyl[i].mc_class=='CAC']).append(onlymethyl[i][onlymethyl[i].mc_class=='CCA']).append(onlymethyl[i][onlymethyl[i].mc_class=='CCC']).append(onlymethyl[i][onlymethyl[i].mc_class=='CCT']).append(onlymethyl[i][onlymethyl[i].mc_class=='CTA']).append(onlymethyl[i][onlymethyl[i].mc_class=='CTC']).append(onlymethyl[i][onlymethyl[i].mc_class=='CTT']))\n",
    "\n",
    "#npmyfile_ref=pd.DataFrame.as_matrix(myfile_ref)\n",
    "#npmyfile_ref_onlymethyl=pd.DataFrame.as_matrix(myfile_ref_onlymethyl)\n",
    "#npCGs_ref=pd.DataFrame.as_matrix(CGs_ref)\n",
    "#npCHGs_ref=pd.DataFrame.as_matrix(CHGs_ref)\n",
    "#npCHHs_ref=pd.DataFrame.as_matrix(CHHs_ref)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "index=np.arange(0,max_size-window,window)\n",
    "columnss=np.arange(0,nbecotype)\n",
    "nbofmethyl=pd.DataFrame(index=index, columns=columnss)\n",
    "for n in range (0, nbecotype):\n",
    "    line=[]\n",
    "    for i in range (1, int((max_size)/window)) : \n",
    "        line.append(methylome_files[n]['methylation_call'][(methylome_files[n]['pos'] > window_list[i-1]) & (methylome_files[n]['pos'] <= window_list[i])].sum())\n",
    "    nbofmethyl[n]=line\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "nbofmethyl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "nbofmethyl.plot.line()\n",
    "#fig1=plot1.get_figure()\n",
    "plt.legend(loc='center left', bbox_to_anchor=(1.0, 0.5))\n",
    "#fig1.savefig('fig1.png')\n",
    "#df=nbofmethyl.sum(axis=1)\n",
    "#plot2=df.plot()\n",
    "#fig2=plot2.get_figure()\n",
    "#fig2.savefig('fig2.png')\n",
    "#nbofmethyl.plot.bar(stacked=True)\n",
    "#plt.legend(loc='center left', bbox_to_anchor=(1.0, 0.5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Some informations about the dataset\n",
    "mycounts=[len(npmyfile_ref)]#,len(npmyfile1), len(npmyfile2), len(npmyfile3), len(npmyfile4), len(npmyfile5), len(npmyfile6)]\n",
    "mycounts_methyl=[len(npmyfile_ref_onlymethyl)]#,len(npmyfile1_onlymethyl), len(npmyfile2_onlymethyl), len(npmyfile3_onlymethyl), len(npmyfile4_onlymethyl), len(npmyfile5_onlymethyl), len(npmyfile6_onlymethyl)]\n",
    "prop_methyl=[mycounts_methyl[i]*100/mycounts[i] for i in range (0,nbecotype)]\n",
    "#here I need decimals..\n",
    "print (mycounts)\n",
    "print (mycounts_methyl)\n",
    "print(\"Percentage of cytosine that are methylated\", prop_methyl) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Methods \n",
    "### Contained in the methyl.py module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def prnt (statement, python) :\n",
    "    \"This function will print the statement depending on the python version\"\n",
    "    \"statement should be the string you want to print or the data\"\n",
    "    \"python should be the version used, 2 or 3\"\n",
    "    if python==2:\n",
    "        print statement\n",
    "    if python==3:\n",
    "        print(statement)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "def getplot (onlymethyl,nbecotype,min_size,max_size,window) :\n",
    "    window_list=np.arange(min_size,max_size+window,window)\n",
    "    columnss=np.arange(0,nbecotype)\n",
    "    nbofmethyl=pd.DataFrame(index=window_list[:-1], columns=columnss)\n",
    "    for n in range (0, nbecotype):\n",
    "        line=[]\n",
    "        for i in range (1, int((max_size+window-min_size)/window)) : \n",
    "            line.append(onlymethyl[n].methylation_call[(onlymethyl[n].pos > window_list[i-1]) & (onlymethyl[n].pos <= window_list[i])].sum())\n",
    "        nbofmethyl[n]=line\n",
    "    nbofmethyl.plot()\n",
    "    plt.legend(loc='center left', bbox_to_anchor=(1.0, 0.5))\n",
    "    \n",
    "def getplotclass (CGs,CHGs,CHHs,min_size,max_size,window) :\n",
    "    window_list=np.arange(min_size,max_size+window,window)\n",
    "    columnss=['CGs','CHGs','CHHs']\n",
    "    nbofmethylclass=pd.DataFrame(index=window_list[:-1], columns=columnss)\n",
    "    for c in columnss:\n",
    "        line=[]\n",
    "        for i in range (1, int((max_size+window-min_size)/window)) : \n",
    "        line.append(c[n].methylation_call[(c[n].pos > window_list[i-1]) & (c[n].pos <= window_list[i])].sum())\n",
    "    nbofmethylclass[c]=line\n",
    "    nbofmethylclass.plot()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Unfinished"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "####TO BE CONTINUED\n",
    "class methylome(object) :\n",
    "    def __init__(self,x):\n",
    "        self.read(x)\n",
    "    def read(self,x) :\n",
    "        \"\"\n",
    "        imp=open(list_ecotypes, 'r')\n",
    "        list_ecotype=[]\n",
    "        for line in imp:\n",
    "            for i in line.split():\n",
    "                list_ecotype.append(i)\n",
    "        for i in range (1, nbchr+1):\n",
    "            exec('self.methylome_file%d = []' % (i))\n",
    "            exec('self.onlymethyl%d = []' % (i))\n",
    "        self.onlymethyl=[]\n",
    "        if nrows == 'all' :\n",
    "            statement ='No number of rows specified, reading all the file'\n",
    "            prnt(statement, python)\n",
    "            for i in range (nbecotype) :    \n",
    "                methylome_files=(pd.read_csv(os.path.join(list_ecotype[i]), sep='\\t'))\n",
    "                if numchr == 'all' :\n",
    "                    for x in range (1,nbchr+1):\n",
    "                        exec('self.onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (x,x))\n",
    "                        exec('self.methylome_file%d.append(methylome_files[methylome_files.chrom == %d])'% (x,x))\n",
    "                else :\n",
    "                    for c in numchr :\n",
    "                        exec('self.onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (c,c))\n",
    "                statement='I read ecotype %d/%d : %s' %(i,nbecotype-1, list_ecotype[i])\n",
    "                prnt(statement,python)\n",
    "        else :\n",
    "            statement='Only the first %d rows will be read' % (nrows)\n",
    "            prnt(statement, python)\n",
    "            for i in range (nbecotype) :\n",
    "                methylome_files=(pd.read_csv(os.path.join(list_ecotype[i]), sep='\\t', nrows=nrows))\n",
    "                if numchr == 'all' :\n",
    "                    for x in range (1,nbchr+1):\n",
    "                        exec('self.onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (x,x))\n",
    "                        exec('self.methylome_file%d.append(methylome_files[methylome_files.chrom == %d])'% (x,x))\n",
    "                else :\n",
    "                    for c in numchr :\n",
    "                        exec('self.onlymethyl%d.append(methylome_files[(methylome_files.methylation_call==1) & (methylome_files.chrom == %d)])' % (c,c))\n",
    "                \n",
    "                stat='I read ecotype %d/%d : %s' %(i,nbecotype-1, list_ecotype[i])\n",
    "                prnt(stat,python)    \n",
    "        methylome_files=None\n",
    "        print x"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
